// office-ui-fabric-react/lib-commonjs/components/Dropdown/examples/Dropdown.Error.Example.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var Dropdown_1 = require("office-ui-fabric-react/lib/Dropdown");
var Utilities_1 = require("office-ui-fabric-react/lib/Utilities");
require("./Dropdown.Basic.Example.scss");
var DropdownErrorExample = /** @class */ (function (_super) {
    tslib_1.__extends(DropdownErrorExample, _super);
    function DropdownErrorExample(props) {
        return _super.call(this, props) || this;
    }
    DropdownErrorExample.prototype.render = function () {
        return (React.createElement("div", { className: "docs-DropdownExample" },
            React.createElement(Dropdown_1.Dropdown, { placeholder: "Select an Option", label: "Error message example:", ariaLabel: "Error message dropdown example", options: [
                    { key: 'A', text: 'Option a' },
                    { key: 'B', text: 'Option b' },
                    { key: 'C', text: 'Option c' },
                    { key: 'D', text: 'Option d' },
                    { key: 'E', text: 'Option e' }
                ], errorMessage: "Error message" })));
    };
    return DropdownErrorExample;
}(Utilities_1.BaseComponent));
exports.DropdownErrorExample = DropdownErrorExample;
