// office-ui-fabric-react/lib/components/ContextualMenu/examples/ContextualMenu.Persisted.Example.js
import * as tslib_1 from "tslib";
// @codepen
import * as React from 'react';
import { ContextualMenuItemType } from 'office-ui-fabric-react/lib/ContextualMenu';
import { DefaultButton } from 'office-ui-fabric-react/lib/Button';
import './ContextualMenuExample.scss';
var ContextualMenuPersistedExample = /** @class */ (function (_super) {
    tslib_1.__extends(ContextualMenuPersistedExample, _super);
    function ContextualMenuPersistedExample(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {
            showCallout: false
        };
        return _this;
    }
    ContextualMenuPersistedExample.prototype.render = function () {
        return (React.createElement("div", null,
            React.createElement(DefaultButton, { text: "Click for ContextualMenu", persistMenu: true, menuProps: {
                    shouldFocusOnMount: true,
                    shouldFocusOnContainer: true,
                    items: [
                        {
                            key: 'newItem',
                            text: 'New',
                            onClick: function () { return console.log('New clicked'); }
                        },
                        {
                            key: 'divider_1',
                            itemType: ContextualMenuItemType.Divider
                        },
                        {
                            key: 'rename',
                            text: 'Rename',
                            onClick: function () { return console.log('Rename clicked'); }
                        },
                        {
                            key: 'edit',
                            text: 'Edit',
                            onClick: function () { return console.log('Edit clicked'); }
                        },
                        {
                            key: 'properties',
                            text: 'Properties',
                            onClick: function () { return console.log('Properties clicked'); }
                        },
                        {
                            key: 'linkNoTarget',
                            text: 'Link same window',
                            href: 'http://bing.com'
                        },
                        {
                            key: 'linkWithTarget',
                            text: 'Link new window',
                            href: 'http://bing.com',
                            target: '_blank'
                        },
                        {
                            key: 'linkWithOnClick',
                            name: 'Link click',
                            href: 'http://bing.com',
                            onClick: function (ev) {
                                alert('Link clicked');
                                ev.preventDefault();
                            },
                            target: '_blank'
                        },
                        {
                            key: 'disabled',
                            text: 'Disabled item',
                            disabled: true,
                            onClick: function () { return console.error('Disabled item should not be clickable.'); }
                        }
                    ]
                } })));
    };
    return ContextualMenuPersistedExample;
}(React.Component));
export { ContextualMenuPersistedExample };
