// office-ui-fabric-react/lib-commonjs/components/Divider/Divider.doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var VerticalDivider_Basic_Example_1 = require("./examples/VerticalDivider.Basic.Example");
var VerticalDivider_Custom_Example_1 = require("./examples/VerticalDivider.Custom.Example");
var Divider_checklist_1 = require("./Divider.checklist");
var VerticalDividerBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Divider/examples/VerticalDivider.Basic.Example.tsx');
var VerticalDividerCustomExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Divider/examples/VerticalDivider.Custom.Example.tsx');
exports.DividerPageProps = {
    title: 'Divider',
    componentName: 'Divider',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Divider',
    componentStatus: Divider_checklist_1.DividerStatus,
    examples: [
        {
            title: 'Vertical Divider',
            code: VerticalDividerBasicExampleCode,
            view: React.createElement(VerticalDivider_Basic_Example_1.VerticalDividerBasicExample, null)
        },
        {
            title: 'Custom Vertical Divider',
            code: VerticalDividerCustomExampleCode,
            view: React.createElement(VerticalDivider_Custom_Example_1.VerticalDividerCustomExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Divider/VerticalDivider.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Divider/docs/DividerOverview.md'),
    bestPractices: require('!raw-loader!office-ui-fabric-react/src/components/Divider/docs/DividerBestPractices.md'),
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Divider/docs/DividerDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Divider/docs/DividerDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
