// office-ui-fabric-react/lib/components/Spinner/Spinner.doc.js
import * as React from 'react';
import { SpinnerBasicExample } from './examples/Spinner.Basic.Example';
import { SpinnerStatus } from './Spinner.checklist';
var SpinnerBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Spinner/examples/Spinner.Basic.Example.tsx');
var SpinnerBasicExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Spinner/Spinner.Basic.Example.Codepen.txt');
export var SpinnerPageProps = {
    title: 'Spinner',
    componentName: 'Spinner',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Spinner',
    componentStatus: SpinnerStatus,
    examples: [
        {
            title: 'Various Spinner Types',
            code: SpinnerBasicExampleCode,
            view: React.createElement(SpinnerBasicExample, null),
            codepenJS: SpinnerBasicExampleCodepen
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Spinner/Spinner.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Spinner/docs/SpinnerOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Spinner/docs/SpinnerDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Spinner/docs/SpinnerDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
