// office-ui-fabric-react/lib-commonjs/components/Label/Label.doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var Label_Basic_Example_1 = require("./examples/Label.Basic.Example");
var Label_checklist_1 = require("./Label.checklist");
var LabelBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Label/examples/Label.Basic.Example.tsx');
exports.LabelPageProps = {
    title: 'Label',
    componentName: 'Label',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Label',
    componentStatus: Label_checklist_1.LabelStatus,
    examples: [
        {
            title: 'Label',
            code: LabelBasicExampleCode,
            view: React.createElement(Label_Basic_Example_1.LabelBasicExample, null),
            codepenJS: require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Label/examples/Label.Basic.Example.tsx')
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Label/Label.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Label/docs/LabelOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Label/docs/LabelDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Label/docs/LabelDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true,
    allowNativeProps: true
};
