// office-ui-fabric-react/lib/components/SearchBox/SearchBox.doc.js
import * as React from 'react';
import { SearchBoxFullSizeExample } from './examples/SearchBox.FullSize.Example';
import { SearchBoxUnderlinedExample } from './examples/SearchBox.Underlined.Example';
import { SearchBoxDisabledExample } from './examples/SearchBox.Disabled.Example';
import { SearchBoxCustomIconExample } from './examples/SearchBox.CustomIcon.Example';
import { SearchBoxSmallExample } from './examples/SearchBox.Small.Example';
import { SearchBoxStatus } from './SearchBox.checklist';
var SearchBoxFullSizeExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/SearchBox/examples/SearchBox.FullSize.Example.tsx');
var SearchBoxFullSizeExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/SearchBox/SearchBox.FullSize.Example.Codepen.txt');
var SearchBoxUnderlinedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/SearchBox/examples/SearchBox.Underlined.Example.tsx');
var SearchBoxDisabledExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/SearchBox/examples/SearchBox.Disabled.Example.tsx');
var SearchBoxCustomIconExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/SearchBox/examples/SearchBox.CustomIcon.Example.tsx');
var SearchBoxSmallExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/SearchBox/examples/SearchBox.Small.Example.tsx');
export var SearchBoxPageProps = {
    title: 'SearchBox',
    componentName: 'SearchBox',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/SearchBox',
    componentStatus: SearchBoxStatus,
    examples: [
        {
            title: 'Default SearchBox',
            code: SearchBoxFullSizeExampleCode,
            view: React.createElement(SearchBoxFullSizeExample, null),
            codepenJS: SearchBoxFullSizeExampleCodepen
        },
        {
            title: 'Underlined SearchBox',
            code: SearchBoxUnderlinedExampleCode,
            view: React.createElement(SearchBoxUnderlinedExample, null)
        },
        {
            title: 'Disabled SearchBoxes',
            code: SearchBoxDisabledExampleCode,
            view: React.createElement(SearchBoxDisabledExample, null)
        },
        {
            title: 'SearchBox with custom icon',
            code: SearchBoxCustomIconExampleCode,
            view: React.createElement(SearchBoxCustomIconExample, null)
        },
        {
            title: 'SearchBox with fixed width and custom event handling',
            code: SearchBoxSmallExampleCode,
            view: React.createElement(SearchBoxSmallExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/SearchBox/SearchBox.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/SearchBox/docs/SearchBoxOverview.md'),
    bestPractices: require('!raw-loader!office-ui-fabric-react/src/components/SearchBox/docs/SearchBoxBestPractices.md'),
    dos: require('!raw-loader!office-ui-fabric-react/src/components/SearchBox/docs/SearchBoxDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/SearchBox/docs/SearchBoxDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
