// office-ui-fabric-react/lib-amd/components/FocusZone/examples/FocusZone.Photos.Example.js
define(["require", "exports", "tslib", "react", "office-ui-fabric-react/lib/FocusZone", "office-ui-fabric-react/lib/Image", "./FocusZone.Photos.Example.scss"], function (require, exports, tslib_1, React, FocusZone_1, Image_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var MAX_COUNT = 20;
    var _counter = 0;
    var FocusZonePhotosExample = /** @class */ (function (_super) {
        tslib_1.__extends(FocusZonePhotosExample, _super);
        function FocusZonePhotosExample(props) {
            var _this = _super.call(this, props) || this;
            _this.state = {
                items: _this._getInitialItems()
            };
            return _this;
        }
        FocusZonePhotosExample.prototype.render = function () {
            var items = this.state.items;
            return (React.createElement(FocusZone_1.FocusZone, { as: "ul", className: "ms-FocusZoneExamples-photoList" },
                React.createElement("p", null, "Note: clicking on the items below will remove them, illustrating how FocusZone will retain focus even when items are removed."),
                items.map(function (item, index) { return (React.createElement("li", { key: item.id, className: "ms-FocusZoneExamples-photoCell", "aria-posinset": index + 1, "aria-setsize": items.length, "aria-label": "Photo", "data-is-focusable": true, onClick: item.onClick },
                    React.createElement(Image_1.Image, { src: item.url, width: item.width, height: item.height }))); })));
        };
        FocusZonePhotosExample.prototype._getInitialItems = function () {
            var items = [];
            for (var i = 0; i < MAX_COUNT; i++) {
                items.push(this._createItem());
            }
            return items;
        };
        FocusZonePhotosExample.prototype._createItem = function () {
            var _this = this;
            var randomWidth = 50 + Math.floor(Math.random() * 150);
            var id = _counter++;
            return {
                id: id,
                url: "http://placehold.it/" + randomWidth + "x100",
                width: randomWidth,
                height: 100,
                onClick: function (ev) {
                    var items = _this.state.items.filter(function (item) { return item.id !== id; });
                    _this.setState({ items: items });
                    // If we have run out of items, repopulate in a couple seconds.
                    if (items.length === 0) {
                        setTimeout(function () { return _this.setState({ items: _this._getInitialItems() }); }, 2000);
                    }
                    ev.preventDefault();
                    ev.stopPropagation();
                }
            };
        };
        return FocusZonePhotosExample;
    }(React.Component));
    exports.FocusZonePhotosExample = FocusZonePhotosExample;
});
