// office-ui-fabric-react/lib-commonjs/components/TeachingBubble/TeachingBubble.doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var TeachingBubble_Basic_Example_1 = require("./examples/TeachingBubble.Basic.Example");
var TeachingBubble_Condensed_Example_1 = require("./examples/TeachingBubble.Condensed.Example");
var TeachingBubble_Illustration_Example_1 = require("./examples/TeachingBubble.Illustration.Example");
var TeachingBubble_WideIllustration_Example_1 = require("./examples/TeachingBubble.WideIllustration.Example");
var TeachingBubble_SmallHeadline_Example_1 = require("./examples/TeachingBubble.SmallHeadline.Example");
var TeachingBubble_checklist_1 = require("./TeachingBubble.checklist");
var TeachingBubbleBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/examples/TeachingBubble.Basic.Example.tsx');
var TeachingBubbleCondensedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/examples/TeachingBubble.Condensed.Example.tsx');
var TeachingBubbleIllustrationExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/examples/TeachingBubble.Basic.Example.tsx');
var TeachingBubbleWideIllustrationExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/examples/TeachingBubble.WideIllustration.Example.tsx');
var TeachingBubbleSmallHeadlineExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/examples/TeachingBubble.SmallHeadline.Example.tsx');
exports.TeachingBubblePageProps = {
    title: 'TeachingBubble',
    componentName: 'TeachingBubble',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/TeachingBubble',
    componentStatus: TeachingBubble_checklist_1.TeachingBubbleStatus,
    examples: [
        {
            title: 'TeachingBubble',
            code: TeachingBubbleBasicExampleCode,
            view: React.createElement(TeachingBubble_Basic_Example_1.TeachingBubbleBasicExample, null)
        },
        {
            title: 'TeachingBubble Condensed',
            code: TeachingBubbleCondensedExampleCode,
            view: React.createElement(TeachingBubble_Condensed_Example_1.TeachingBubbleCondensedExample, null)
        },
        {
            title: 'TeachingBubble with Illustration',
            code: TeachingBubbleIllustrationExampleCode,
            view: React.createElement(TeachingBubble_Illustration_Example_1.TeachingBubbleIllustrationExample, null)
        },
        {
            title: 'TeachingBubble wide with Illustration',
            code: TeachingBubbleWideIllustrationExampleCode,
            view: React.createElement(TeachingBubble_WideIllustration_Example_1.TeachingBubbleWideIllustrationExample, null)
        },
        {
            title: 'TeachingBubble with small headline',
            code: TeachingBubbleSmallHeadlineExampleCode,
            view: React.createElement(TeachingBubble_SmallHeadline_Example_1.TeachingBubbleSmallHeadlineExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/TeachingBubble.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/docs/TeachingBubbleOverview.md'),
    dos: require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/docs/TeachingBubbleDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/docs/TeachingBubbleDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
