// office-ui-fabric-react/lib-amd/components/ScrollablePane/ScrollablePane.doc.js
define(["require", "exports", "react", "./examples/ScrollablePane.Default.Example", "./examples/ScrollablePane.DetailsList.Example", "./ScrollablePane.checklist"], function (require, exports, React, ScrollablePane_Default_Example_1, ScrollablePane_DetailsList_Example_1, ScrollablePane_checklist_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ScrollablePaneDefaultExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ScrollablePane/examples/ScrollablePane.Default.Example.tsx');
    var ScrollablePaneDetailsListExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ScrollablePane/examples/ScrollablePane.DetailsList.Example.tsx');
    var ScrollablePaneDefaultExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/ScrollablePane/examples/ScrollablePane.Default.Example.tsx');
    var ScrollablePaneDetailsListExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/ScrollablePane/examples/ScrollablePane.DetailsList.Example.tsx');
    exports.ScrollablePanePageProps = {
        title: 'ScrollablePane',
        componentName: 'ScrollablePane',
        componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/ScrollablePane',
        componentStatus: ScrollablePane_checklist_1.ScrollablePaneStatus,
        examples: [
            {
                title: 'Default',
                code: ScrollablePaneDefaultExampleCode,
                codepenJS: ScrollablePaneDefaultExampleCodepen,
                view: React.createElement(ScrollablePane_Default_Example_1.ScrollablePaneDefaultExample, null),
                isScrollable: false
            },
            {
                title: 'DetailsList Locked Header',
                code: ScrollablePaneDetailsListExampleCode,
                codepenJS: ScrollablePaneDetailsListExampleCodepen,
                view: React.createElement(ScrollablePane_DetailsList_Example_1.ScrollablePaneDetailsListExample, null),
                isScrollable: false
            }
        ],
        propertiesTablesSources: [
            require('!raw-loader!office-ui-fabric-react/src/components/ScrollablePane/ScrollablePane.types.ts'),
            require('!raw-loader!office-ui-fabric-react/src/components/Sticky/Sticky.types.ts')
        ],
        overview: require('!raw-loader!office-ui-fabric-react/src/components/ScrollablePane/docs/ScrollablePaneOverview.md'),
        bestPractices: '',
        dos: require('!raw-loader!office-ui-fabric-react/src/components/ScrollablePane/docs/ScrollablePaneDos.md'),
        donts: require('!raw-loader!office-ui-fabric-react/src/components/ScrollablePane/docs/ScrollablePaneDonts.md'),
        isHeaderVisible: true,
        isFeedbackVisible: true,
        allowNativeProps: true,
        nativePropsElement: ['a', 'button']
    };
});
