// office-ui-fabric-react/lib/components/ComboBox/ComboBox.doc.js
import * as React from 'react';
import { ComboBoxBasicExample } from './examples/ComboBox.Basic.Example';
import { ComboBoxTogglesExample } from './examples/ComboBox.Toggles.Example';
import { ComboBoxControlledExample } from './examples/ComboBox.Controlled.Example';
import { ComboBoxVirtualizedExample } from './examples/ComboBox.Virtualized.Example';
import { ComboBoxCustomStyledExample } from './examples/ComboBox.CustomStyled.Example';
import { ComboBoxStatus } from './ComboBox.checklist';
var ComboBoxBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ComboBox/examples/ComboBox.Basic.Example.tsx');
var ComboBoxBasicExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/ComboBox/examples/ComboBox.Basic.Example.tsx');
var ComboBoxTogglesExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ComboBox/examples/ComboBox.Toggles.Example.tsx');
var ComboBoxTogglesExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/ComboBox/examples/ComboBox.Toggles.Example.tsx');
var ComboBoxControlledExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ComboBox/examples/ComboBox.Controlled.Example.tsx');
var ComboBoxControlledExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/ComboBox/examples/ComboBox.Controlled.Example.tsx');
var ComboBoxVirtualizedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ComboBox/examples/ComboBox.Virtualized.Example.tsx');
var ComboBoxVirtualizedExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/ComboBox/examples/ComboBox.Virtualized.Example.tsx');
var ComboBoxCustomStyledExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ComboBox/examples/ComboBox.CustomStyled.Example.tsx');
var ComboBoxCustomStyledExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/ComboBox/examples/ComboBox.CustomStyled.Example.tsx');
export var ComboBoxPageProps = {
    title: 'ComboBox',
    componentName: 'ComboBox',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/ComboBox',
    componentStatus: ComboBoxStatus,
    examples: [
        {
            title: 'Basic uncontrolled ComboBox',
            code: ComboBoxBasicExampleCode,
            view: React.createElement(ComboBoxBasicExample, null),
            codepenJS: ComboBoxBasicExampleCodepen
        },
        {
            title: 'ComboBox with toggleable autoComplete and allowFreeform',
            code: ComboBoxTogglesExampleCode,
            view: React.createElement(ComboBoxTogglesExample, null),
            codepenJS: ComboBoxTogglesExampleCodepen
        },
        {
            title: 'Controlled ComboBox',
            code: ComboBoxControlledExampleCode,
            view: React.createElement(ComboBoxControlledExample, null),
            codepenJS: ComboBoxControlledExampleCodepen
        },
        {
            title: 'VirtualizedComboBox',
            code: ComboBoxVirtualizedExampleCode,
            view: React.createElement(ComboBoxVirtualizedExample, null),
            codepenJS: ComboBoxVirtualizedExampleCodepen
        },
        {
            title: 'ComboBox with custom styling',
            code: ComboBoxCustomStyledExampleCode,
            view: React.createElement(ComboBoxCustomStyledExample, null),
            codepenJS: ComboBoxCustomStyledExampleCodepen
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/ComboBox/ComboBox.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/ComboBox/docs/ComboBoxOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/ComboBox/docs/ComboBoxDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/ComboBox/docs/ComboBoxDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true,
    allowNativeProps: true
};
