// office-ui-fabric-react/lib/components/FocusZone/FocusZone.doc.js
import * as React from 'react';
import { FocusZonePhotosExample } from './examples/FocusZone.Photos.Example';
import { FocusZoneListExample } from './examples/FocusZone.List.Example';
import { FocusZoneDisabledExample } from './examples/FocusZone.Disabled.Example';
import { FocusZoneTabbableExample } from './examples/FocusZone.Tabbable.Example';
var FocusZonePhotosExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.Photos.Example.tsx');
var FocusZonePhotosExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.Photos.Example.tsx');
var FocusZoneListExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.List.Example.tsx');
var FocusZoneListExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.List.Example.tsx');
var FocusZoneDisabledExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.Disabled.Example.tsx');
var FocusZoneDisabledExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.Disabled.Example.tsx');
var FocusZoneTabbableCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.Tabbable.Example.tsx');
var FocusZoneTabbableCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.Tabbable.Example.tsx');
export var FocusZonePageProps = {
    title: 'FocusZone',
    componentName: 'FocusZone',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/FocusZone',
    examples: [
        {
            title: 'Non-uniform photos within bidirectional FocusZone',
            code: FocusZonePhotosExampleCode,
            codepenJS: FocusZonePhotosExampleCodepen,
            view: React.createElement(FocusZonePhotosExample, null)
        },
        {
            title: 'Nesting FocusZones in list rows',
            code: FocusZoneListExampleCode,
            codepenJS: FocusZoneListExampleCodepen,
            view: React.createElement(FocusZoneListExample, null)
        },
        {
            title: 'Disabled FocusZone',
            code: FocusZoneDisabledExampleCode,
            codepenJS: FocusZoneDisabledExampleCodepen,
            view: React.createElement(FocusZoneDisabledExample, null)
        },
        {
            title: 'Tabbable FocusZone',
            code: FocusZoneTabbableCode,
            codepenJS: FocusZoneTabbableCodepen,
            view: React.createElement(FocusZoneTabbableExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/FocusZone/FocusZone.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/FocusZone/docs/FocusZoneOverview.md'),
    bestPractices: '',
    dos: '',
    donts: '',
    isHeaderVisible: true,
    isFeedbackVisible: true,
    allowNativeProps: true
};
