// office-ui-fabric-react/lib-amd/components/ChoiceGroup/ChoiceGroup.doc.js
define(["require", "exports", "react", "./examples/ChoiceGroup.Basic.Example", "./examples/ChoiceGroup.Label.Example", "./examples/ChoiceGroup.Custom.Example", "./examples/ChoiceGroup.Image.Example", "./examples/ChoiceGroup.Icon.Example", "./ChoiceGroup.checklist"], function (require, exports, React, ChoiceGroup_Basic_Example_1, ChoiceGroup_Label_Example_1, ChoiceGroup_Custom_Example_1, ChoiceGroup_Image_Example_1, ChoiceGroup_Icon_Example_1, ChoiceGroup_checklist_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ChoiceGroupBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ChoiceGroup/examples/ChoiceGroup.Basic.Example.tsx');
    var ChoiceGroupBasicExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/ChoiceGroup/examples/ChoiceGroup.Basic.Example.tsx');
    var ChoiceGroupLabelExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ChoiceGroup/examples/ChoiceGroup.Label.Example.tsx');
    var ChoiceGroupLabelExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/ChoiceGroup/examples/ChoiceGroup.Label.Example.tsx');
    var ChoiceGroupCustomExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ChoiceGroup/examples/ChoiceGroup.Custom.Example.tsx');
    var ChoiceGroupImageExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ChoiceGroup/examples/ChoiceGroup.Image.Example.tsx');
    var ChoiceGroupIconExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ChoiceGroup/examples/ChoiceGroup.Icon.Example.tsx');
    var ChoiceGroupIconExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/ChoiceGroup/examples/ChoiceGroup.Icon.Example.tsx');
    exports.ChoiceGroupPageProps = {
        title: 'ChoiceGroup',
        componentName: 'ChoiceGroup',
        componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/ChoiceGroup',
        componentStatus: ChoiceGroup_checklist_1.ChoiceGroupStatus,
        examples: [
            {
                title: 'Default ChoiceGroup',
                code: ChoiceGroupBasicExampleCode,
                view: React.createElement(ChoiceGroup_Basic_Example_1.ChoiceGroupBasicExample, null),
                codepenJS: ChoiceGroupBasicExampleCodepen
            },
            {
                title: 'ChoiceGroup with a custom label',
                code: ChoiceGroupLabelExampleCode,
                view: React.createElement(ChoiceGroup_Label_Example_1.ChoiceGroupLabelExample, null),
                codepenJS: ChoiceGroupLabelExampleCodepen
            },
            {
                title: 'ChoiceGroup with dropdown',
                code: ChoiceGroupCustomExampleCode,
                view: React.createElement(ChoiceGroup_Custom_Example_1.ChoiceGroupCustomExample, null)
            },
            {
                title: 'ChoiceGroups with images',
                code: ChoiceGroupImageExampleCode,
                view: React.createElement(ChoiceGroup_Image_Example_1.ChoiceGroupImageExample, null)
            },
            {
                title: 'ChoiceGroup with icons',
                code: ChoiceGroupIconExampleCode,
                view: React.createElement(ChoiceGroup_Icon_Example_1.ChoiceGroupIconExample, null),
                codepenJS: ChoiceGroupIconExampleCodepen
            }
        ],
        propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/ChoiceGroup/ChoiceGroup.types.ts')],
        overview: require('!raw-loader!office-ui-fabric-react/src/components/ChoiceGroup/docs/ChoiceGroupOverview.md'),
        bestPractices: '',
        dos: require('!raw-loader!office-ui-fabric-react/src/components/ChoiceGroup/docs/ChoiceGroupDos.md'),
        donts: require('!raw-loader!office-ui-fabric-react/src/components/ChoiceGroup/docs/ChoiceGroupDonts.md'),
        isHeaderVisible: true,
        isFeedbackVisible: true,
        allowNativeProps: true,
        nativePropsElement: 'input'
    };
});
