// office-ui-fabric-react/lib/common/testUtilities.js
import { mount } from 'enzyme';
import * as ReactDOM from 'react-dom';
import * as ReactTestUtils from 'react-dom/test-utils';
export function findNodes(wrapper, className) {
    return wrapper.find(className).filterWhere(function (node) { return typeof node.type() === 'string'; });
}
export function expectNodes(wrapper, className, n) {
    expect(findNodes(wrapper, className).length).toEqual(n);
}
export function expectOne(wrapper, className) {
    expectNodes(wrapper, className, 1);
}
export function expectMissing(wrapper, className) {
    expectNodes(wrapper, className, 0);
}
export function delay(millisecond) {
    return new Promise(function (resolve) { return setTimeout(resolve, millisecond); });
}
/**
 * Mounts the element attached to a child of document.body. This is primarily for tests involving
 * event handlers (which don't work right unless the element is attached).
 */
export function mountAttached(element) {
    var parent = document.createElement('div');
    document.body.appendChild(parent);
    return mount(element, { attachTo: parent });
}
export function renderIntoDocument(element) {
    var component = ReactTestUtils.renderIntoDocument(element);
    var renderedDOM = ReactDOM.findDOMNode(component);
    return renderedDOM;
}
export function mockEvent(targetValue) {
    if (targetValue === void 0) { targetValue = ''; }
    var target = { value: targetValue };
    return { target: target };
}
