// office-ui-fabric-react/lib-amd/components/Dropdown/Dropdown.doc.js
define(["require", "exports", "react", "./examples/Dropdown.Basic.Example", "./examples/Dropdown.Custom.Example", "./examples/Dropdown.Error.Example", "./examples/Dropdown.Required.Example", "./Dropdown.checklist"], function (require, exports, React, Dropdown_Basic_Example_1, Dropdown_Custom_Example_1, Dropdown_Error_Example_1, Dropdown_Required_Example_1, Dropdown_checklist_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var DropdownBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Basic.Example.tsx');
    var DropdownBasicExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Dropdown/Dropdown.Basic.Example.Codepen.txt');
    var DropdownCustomExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Custom.Example.tsx');
    var DropdownErrorExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Error.Example.tsx');
    var DropdownRequiredExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Required.Example.tsx');
    exports.DropdownPageProps = {
        title: 'Dropdown',
        componentName: 'Dropdown',
        componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Dropdown',
        componentStatus: Dropdown_checklist_1.DropdownStatus,
        examples: [
            {
                title: 'Dropdown',
                code: DropdownBasicExampleCode,
                view: React.createElement(Dropdown_Basic_Example_1.DropdownBasicExample, null),
                codepenJS: DropdownBasicExampleCodepen
            },
            {
                title: 'Customized Dropdown',
                code: DropdownCustomExampleCode,
                view: React.createElement(Dropdown_Custom_Example_1.DropdownCustomExample, null)
            },
            {
                title: 'Dropdown with Error Message',
                code: DropdownErrorExampleCode,
                view: React.createElement(Dropdown_Error_Example_1.DropdownErrorExample, null)
            },
            {
                title: 'Required Dropdown',
                code: DropdownRequiredExampleCode,
                view: React.createElement(Dropdown_Required_Example_1.DropdownRequiredExample, null)
            }
        ],
        propertiesTablesSources: [
            require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/Dropdown.types.ts'),
            require('!raw-loader!office-ui-fabric-react/src/utilities/selectableOption/SelectableDroppableText.types.ts')
        ],
        overview: require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/docs/DropdownOverview.md'),
        bestPractices: '',
        dos: require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/docs/DropdownDos.md'),
        donts: require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/docs/DropdownDonts.md'),
        isHeaderVisible: true,
        isFeedbackVisible: true,
        allowNativeProps: true
    };
});
