// @uifabric/styling/lib-amd/styles/CommonStyles.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.HighContrastSelector = '@media screen and (-ms-high-contrast: active)';
    exports.HighContrastSelectorWhite = '@media screen and (-ms-high-contrast: black-on-white)';
    exports.HighContrastSelectorBlack = '@media screen and (-ms-high-contrast: white-on-black)';
    exports.ScreenWidthMinSmall = 320;
    exports.ScreenWidthMinMedium = 480;
    exports.ScreenWidthMinLarge = 640;
    exports.ScreenWidthMinXLarge = 1024;
    exports.ScreenWidthMinXXLarge = 1366;
    exports.ScreenWidthMinXXXLarge = 1920;
    exports.ScreenWidthMaxSmall = exports.ScreenWidthMinMedium - 1;
    exports.ScreenWidthMaxMedium = exports.ScreenWidthMinLarge - 1;
    exports.ScreenWidthMaxLarge = exports.ScreenWidthMinXLarge - 1;
    exports.ScreenWidthMaxXLarge = exports.ScreenWidthMinXXLarge - 1;
    exports.ScreenWidthMaxXXLarge = exports.ScreenWidthMinXXXLarge - 1;
    exports.ScreenWidthMinUhfMobile = 768;
    function getScreenSelector(min, max) {
        return "@media only screen and (min-width: " + min + "px) and (max-width: " + max + "px)";
    }
    exports.getScreenSelector = getScreenSelector;
});
