// @uifabric/styling/lib-amd/styles/DefaultFontStyles.js
define(["require", "exports", "@uifabric/merge-styles", "./fonts", "@uifabric/utilities"], function (require, exports, merge_styles_1, fonts_1, utilities_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    // Default urls.
    var DefaultBaseUrl = 'https://static2.sharepointonline.com/files/fabric/assets';
    // Standard font styling.
    exports.DefaultFontStyles = fonts_1.createFontStyles(utilities_1.getLanguage());
    function _registerFontFace(fontFamily, url, fontWeight, localFontName) {
        fontFamily = "'" + fontFamily + "'";
        var localFontSrc = localFontName !== undefined ? "local('" + localFontName + "')," : '';
        merge_styles_1.fontFace({
            fontFamily: fontFamily,
            src: localFontSrc + ("url('" + url + ".woff2') format('woff2'),") + ("url('" + url + ".woff') format('woff')"),
            fontWeight: fontWeight,
            fontStyle: 'normal'
        });
    }
    function _registerFontFaceSet(baseUrl, fontFamily, cdnFolder, cdnFontName, localFontName) {
        if (cdnFontName === void 0) { cdnFontName = 'segoeui'; }
        var urlBase = baseUrl + "/" + cdnFolder + "/" + cdnFontName;
        _registerFontFace(fontFamily, urlBase + '-light', fonts_1.FontWeights.light, localFontName && localFontName + ' Light');
        _registerFontFace(fontFamily, urlBase + '-semilight', fonts_1.FontWeights.semilight, localFontName && localFontName + ' SemiLight');
        _registerFontFace(fontFamily, urlBase + '-regular', fonts_1.FontWeights.regular, localFontName);
        _registerFontFace(fontFamily, urlBase + '-semibold', fonts_1.FontWeights.semibold, localFontName && localFontName + ' SemiBold');
    }
    function registerDefaultFontFaces(baseUrl) {
        if (baseUrl) {
            var fontUrl = baseUrl + "/fonts";
            // Produce @font-face definitions for all supported web fonts.
            _registerFontFaceSet(fontUrl, fonts_1.LocalizedFontNames.Thai, 'leelawadeeui-thai', 'leelawadeeui');
            _registerFontFaceSet(fontUrl, fonts_1.LocalizedFontNames.Arabic, 'segoeui-arabic');
            _registerFontFaceSet(fontUrl, fonts_1.LocalizedFontNames.Cyrillic, 'segoeui-cyrillic');
            _registerFontFaceSet(fontUrl, fonts_1.LocalizedFontNames.EastEuropean, 'segoeui-easteuropean');
            _registerFontFaceSet(fontUrl, fonts_1.LocalizedFontNames.Greek, 'segoeui-greek');
            _registerFontFaceSet(fontUrl, fonts_1.LocalizedFontNames.Hebrew, 'segoeui-hebrew');
            _registerFontFaceSet(fontUrl, fonts_1.LocalizedFontNames.Vietnamese, 'segoeui-vietnamese');
            _registerFontFaceSet(fontUrl, fonts_1.LocalizedFontNames.WestEuropean, 'segoeui-westeuropean', 'segoeui', 'Segoe UI');
            _registerFontFaceSet(fontUrl, fonts_1.LocalizedFontFamilies.Selawik, 'selawik', 'selawik');
            // Leelawadee UI (Thai) does not have a 'light' weight, so we override
            // the font-face generated above to use the 'semilight' weight instead.
            _registerFontFace('Leelawadee UI Web', fontUrl + "/leelawadeeui-thai/leelawadeeui-semilight", fonts_1.FontWeights.light);
            // Leelawadee UI (Thai) does not have a 'semibold' weight, so we override
            // the font-face generated above to use the 'bold' weight instead.
            _registerFontFace('Leelawadee UI Web', fontUrl + "/leelawadeeui-thai/leelawadeeui-bold", fonts_1.FontWeights.semibold);
        }
    }
    exports.registerDefaultFontFaces = registerDefaultFontFaces;
    /**
     * Reads the fontBaseUrl from window.FabricConfig.fontBaseUrl or falls back to a default.
     */
    function _getFontBaseUrl() {
        var win = typeof window !== 'undefined' ? window : undefined;
        // tslint:disable-next-line:no-string-literal no-any
        var fabricConfig = win ? win['FabricConfig'] : undefined;
        return fabricConfig && fabricConfig.fontBaseUrl !== undefined ? fabricConfig.fontBaseUrl : DefaultBaseUrl;
    }
    /**
     * Register the font faces.
     */
    registerDefaultFontFaces(_getFontBaseUrl());
});
