// office-ui-fabric-react/lib-amd/components/MessageBar/examples/MessageBar.Basic.Example.js
define(["require", "exports", "react", "office-ui-fabric-react/lib/Button", "office-ui-fabric-react/lib/Link", "office-ui-fabric-react/lib/Label", "office-ui-fabric-react/lib/MessageBar", "./MessageBar.Basic.Example.scss"], function (require, exports, React, Button_1, Link_1, Label_1, MessageBar_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var log = function (text) { return function () { return console.log(text); }; };
    exports.MessageBarBasicExample = function () { return (React.createElement("div", { className: "ms-BasicMessageBarsExample" },
        React.createElement(Label_1.Label, null, "Info/Default MessageBar"),
        React.createElement(MessageBar_1.MessageBar, null,
            "Info lorem ipsum dolor sit amet, a elit sem interdum consectetur adipiscing elit. ",
            React.createElement(Link_1.Link, { href: "www.bing.com" }, "Visit our website.")),
        React.createElement(Label_1.Label, null, "Error MessageBar - single line, with dismiss button"),
        React.createElement(MessageBar_1.MessageBar, { messageBarType: MessageBar_1.MessageBarType.error, isMultiline: false, onDismiss: log('test'), dismissButtonAriaLabel: "Close" },
            "Error lorem ipsum dolor sit amet, a elit sem interdum consectetur adipiscing elit. ",
            React.createElement(Link_1.Link, { href: "www.bing.com" }, "Visit our website.")),
        React.createElement(Label_1.Label, null, "Blocked MessageBar - single line, with dismiss button and truncated text. Truncation is not available if you use action buttons or multiline and should be used sparingly."),
        React.createElement(MessageBar_1.MessageBar, { messageBarType: MessageBar_1.MessageBarType.blocked, isMultiline: false, onDismiss: log('test'), dismissButtonAriaLabel: "Close", truncated: true, overflowButtonAriaLabel: "See more" },
            "Blocked lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi luctus, purus a lobortis tristique, odio augue pharetra metus, ac placerat nunc mi nec dui. Vestibulum aliquam et nunc semper scelerisque. Curabitur vitae orci nec quam condimentum porttitor et sed lacus. Vivamus ac efficitur leo. Cras faucibus mauris libero, ac placerat erat euismod et. Donec pulvinar commodo odio sit amet faucibus. In hac habitasse platea dictumst. Duis eu ante commodo, condimentum nibh pellentesque, laoreet enim. Fusce massa lorem, ultrices eu mi a, fermentum suscipit magna. Integer porta purus pulvinar, hendrerit felis eget, condimentum mauris.",
            ' ',
            React.createElement(Link_1.Link, { href: "www.bing.com" }, "Visit our website.")),
        React.createElement(Label_1.Label, null, "SevereWarning MessageBar - defaults to multiline, with action buttons"),
        React.createElement(MessageBar_1.MessageBar, { messageBarType: MessageBar_1.MessageBarType.severeWarning, actions: React.createElement("div", null,
                React.createElement(Button_1.MessageBarButton, null, "Yes"),
                React.createElement(Button_1.MessageBarButton, null, "No")) },
            React.createElement("span", null, "Severe warning lorem ipsum dolor sit amet, a elit sem interdum consectetur adipiscing elit."),
            React.createElement(Link_1.Link, { href: "www.bing.com" }, "Visit our website.")),
        React.createElement(Label_1.Label, null, "Success MessageBar - single line, with action buttons"),
        React.createElement(MessageBar_1.MessageBar, { actions: React.createElement("div", null,
                React.createElement(Button_1.MessageBarButton, null, "Yes"),
                React.createElement(Button_1.MessageBarButton, null, "No")), messageBarType: MessageBar_1.MessageBarType.success, isMultiline: false },
            "Success lorem ipsum dolor sit amet. ",
            React.createElement(Link_1.Link, { href: "www.bing.com" }, "Visit our website.")),
        React.createElement(Label_1.Label, null, "Warning MessageBar - single line, with dismiss and action buttons"),
        React.createElement(MessageBar_1.MessageBar, { messageBarType: MessageBar_1.MessageBarType.warning, isMultiline: false, onDismiss: log('test'), dismissButtonAriaLabel: "Close", actions: React.createElement("div", null,
                React.createElement(Button_1.MessageBarButton, null, "Action")) },
            "Warning lorem ipsum dolor sit amet, a elit sem interdum consectetur adipiscing elit.",
            ' ',
            React.createElement(Link_1.Link, { href: "www.bing.com" }, "Visit our website.")),
        React.createElement(Label_1.Label, null, "Warning MessageBar - defaults to multiline, with dismiss and action buttons"),
        React.createElement(MessageBar_1.MessageBar, { onDismiss: log('test'), dismissButtonAriaLabel: "Close", messageBarType: MessageBar_1.MessageBarType.warning, ariaLabel: "Aria help text here", actions: React.createElement("div", null,
                React.createElement(Button_1.MessageBarButton, null, "Yes"),
                React.createElement(Button_1.MessageBarButton, null, "No")) },
            "Warning lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi luctus, purus a lobortis tristique, odio augue pharetra metus, ac placerat nunc mi nec dui. Vestibulum aliquam et nunc semper scelerisque. Curabitur vitae orci nec quam condimentum porttitor et sed lacus. Vivamus ac efficitur leo. Cras faucibus mauris libero, ac placerat erat euismod et. Donec pulvinar commodo odio sit amet faucibus. In hac habitasse platea dictumst. Duis eu ante commodo, condimentum nibh pellentesque, laoreet enim. Fusce massa lorem, ultrices eu mi a, fermentum suscipit magna. Integer porta purus pulvinar, hendrerit felis eget, condimentum mauris.",
            React.createElement(Link_1.Link, { href: "www.bing.com" }, "Visit our website.")),
        React.createElement(Label_1.Label, null, "Info/Default MessageBar - custom styles"),
        React.createElement(MessageBar_1.MessageBar, { styles: {
                root: {
                    background: 'rgba(113, 175, 229, 0.2)',
                    color: '#00188f'
                },
                icon: {
                    color: '#00188f'
                }
            } },
            "Info lorem ipsum dolor sit amet, a elit sem interdum consectetur adipiscing elit. ",
            React.createElement(Link_1.Link, { href: "www.bing.com" }, "Visit our website.")),
        React.createElement(Label_1.Label, null, "Error, single line, with dismiss button - custom styles"),
        React.createElement(MessageBar_1.MessageBar, { styles: {
                content: {
                    background: 'rgba(50, 20, 90, 0.2)'
                },
                text: {
                    background: 'rgba(166, 166, 166, 0.5)'
                },
                dismissal: {
                    selectors: {
                        '& .ms-Button-icon': {
                            color: '#107c10'
                        }
                    }
                }
            }, messageBarType: MessageBar_1.MessageBarType.error, isMultiline: false, onDismiss: log('test'), dismissButtonAriaLabel: "Close" },
            "Lorem ipsum dolor sit amet, a elit sem interdum consectetur adipiscing elit. ",
            React.createElement(Link_1.Link, { href: "www.bing.com" }, "Visit our website.")),
        React.createElement(Label_1.Label, null, "Blocked, single line, with dismiss button and truncated text - custom styles"),
        React.createElement(MessageBar_1.MessageBar, { styles: {
                root: {
                    background: 'rgba(0, 178, 148, 0.2)'
                },
                dismissSingleLine: {
                    border: '1px solid #5c2d91'
                },
                expandSingleLine: {
                    border: '1px solid #e3008c'
                },
                dismissal: {
                    selectors: {
                        '& .ms-Button-icon': {
                            color: '#004b50'
                        }
                    }
                },
                expand: {
                    selectors: {
                        '& .ms-Button-icon': {
                            color: '#004b50'
                        }
                    }
                },
                innerText: {
                    textDecoration: 'underline'
                }
            }, messageBarType: MessageBar_1.MessageBarType.blocked, isMultiline: false, onDismiss: log('test'), dismissButtonAriaLabel: "Close", truncated: true, overflowButtonAriaLabel: "See more" },
            "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi luctus, purus a lobortis tristique, odio augue pharetra metus, ac placerat nunc mi nec dui. Vestibulum aliquam et nunc semper scelerisque. Curabitur vitae orci nec quam condimentum porttitor et sed lacus. Vivamus ac efficitur leo. Cras faucibus mauris libero, ac placerat erat euismod et. Donec pulvinar commodo odio sit amet faucibus. In hac habitasse platea dictumst. Duis eu ante commodo, condimentum nibh pellentesque, laoreet enim. Fusce massa lorem, ultrices eu mi a, fermentum suscipit magna. Integer porta purus pulvinar, hendrerit felis eget, condimentum mauris.",
            ' ',
            React.createElement(Link_1.Link, { href: "www.bing.com" }, "Visit our website.")),
        React.createElement(Label_1.Label, null, "Severe warning, multiline, with action buttons - custom styles"),
        React.createElement(MessageBar_1.MessageBar, { styles: {
                root: {
                    background: 'rgba(227, 0, 140, 0.2)'
                },
                icon: {
                    color: '#e3008c'
                },
                actions: {
                    background: 'rgba(180, 0, 158, 0.4)'
                }
            }, messageBarType: MessageBar_1.MessageBarType.severeWarning, actions: React.createElement("div", null,
                React.createElement(Button_1.MessageBarButton, { styles: { root: { color: '#e3008c' } } }, "Yes"),
                React.createElement(Button_1.MessageBarButton, { styles: { root: { color: '#e3008c' } } }, "No")) },
            React.createElement("span", null, "Lorem ipsum dolor sit amet, a elit sem interdum consectetur adipiscing elit."),
            React.createElement(Link_1.Link, { href: "www.bing.com" }, "Visit our website.")),
        React.createElement(Label_1.Label, null, "Success, single line, with action buttons - custom styles"),
        React.createElement(MessageBar_1.MessageBar, { styles: {
                actions: {
                    background: '#007d84',
                    border: '1px solid #ffb900'
                }
            }, actions: React.createElement("div", null,
                React.createElement(Button_1.MessageBarButton, null, "Yes"),
                React.createElement(Button_1.MessageBarButton, null, "No")), messageBarType: MessageBar_1.MessageBarType.success, isMultiline: false },
            "Lorem ipsum dolor sit amet. ",
            React.createElement(Link_1.Link, { href: "www.bing.com" }, "Visit our website.")),
        React.createElement(Label_1.Label, null, "Warning, single line, with dismiss and action buttons - custom styles"),
        React.createElement(MessageBar_1.MessageBar, { styles: {
                content: {
                    background: 'rgba(234, 67, 0, 0.2)'
                },
                actions: {
                    border: '1px solid #ea4300'
                },
                dismissSingleLine: {
                    border: '1px solid #ea4300'
                }
            }, messageBarType: MessageBar_1.MessageBarType.warning, isMultiline: false, onDismiss: log('test'), dismissButtonAriaLabel: "Close", actions: React.createElement("div", null,
                React.createElement(Button_1.MessageBarButton, null, "Action")) },
            "Lorem ipsum dolor sit amet, a elit sem interdum consectetur adipiscing elit. ",
            React.createElement(Link_1.Link, { href: "www.bing.com" }, "Visit our website.")),
        React.createElement(Label_1.Label, null, "Warning, multiline, with dismiss and action buttons - custom styles"),
        React.createElement(MessageBar_1.MessageBar, { styles: {
                root: {
                    background: 'rgba(0, 188, 242, 0.4)'
                },
                dismissal: {
                    border: '1px solid #00188f'
                },
                actions: {
                    border: '1px solid #0078d4',
                    flexDirection: 'row'
                }
            }, onDismiss: log('test'), dismissButtonAriaLabel: "Close", messageBarType: MessageBar_1.MessageBarType.warning, ariaLabel: "Aria help text here", actions: React.createElement("div", null,
                React.createElement(Button_1.MessageBarButton, null, "Yes"),
                React.createElement(Button_1.MessageBarButton, null, "No")) },
            "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi luctus, purus a lobortis tristique, odio augue pharetra metus, ac placerat nunc mi nec dui. Vestibulum aliquam et nunc semper scelerisque. Curabitur vitae orci nec quam condimentum porttitor et sed lacus. Vivamus ac efficitur leo. Cras faucibus mauris libero, ac placerat erat euismod et. Donec pulvinar commodo odio sit amet faucibus. In hac habitasse platea dictumst. Duis eu ante commodo, condimentum nibh pellentesque, laoreet enim. Fusce massa lorem, ultrices eu mi a, fermentum suscipit magna. Integer porta purus pulvinar, hendrerit felis eget, condimentum mauris.",
            React.createElement(Link_1.Link, { href: "www.bing.com" }, "Visit our website.")))); };
});
