// office-ui-fabric-react/lib/components/Persona/examples/Persona.Alternate.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { Persona, PersonaSize, PersonaPresence } from 'office-ui-fabric-react/lib/Persona';
import { TestImages } from '../../../common/TestImages';
import './PersonaExample.scss';
var examplePersona = {
    imageUrl: TestImages.personaMale,
    imageInitials: 'AR',
    text: 'Annie Reid',
    secondaryText: 'Designer',
    tertiaryText: 'In a meeting',
    optionalText: 'Available at 4:00pm',
    showSecondaryText: true
};
var PersonaAlternateExample = /** @class */ (function (_super) {
    tslib_1.__extends(PersonaAlternateExample, _super);
    function PersonaAlternateExample(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {
            renderPersonaDetails: true
        };
        return _this;
    }
    PersonaAlternateExample.prototype.render = function () {
        var renderPersonaDetails = this.state.renderPersonaDetails;
        return (React.createElement("div", { className: "ms-PersonaExample" },
            React.createElement(Persona, tslib_1.__assign({}, examplePersona, { size: PersonaSize.size24, presence: PersonaPresence.none, hidePersonaDetails: !renderPersonaDetails })),
            React.createElement(Persona, tslib_1.__assign({}, examplePersona, { size: PersonaSize.size28, presence: PersonaPresence.none, hidePersonaDetails: !renderPersonaDetails })),
            React.createElement(Persona, tslib_1.__assign({}, examplePersona, { size: PersonaSize.size32, presence: PersonaPresence.online, hidePersonaDetails: !renderPersonaDetails }))));
    };
    return PersonaAlternateExample;
}(React.Component));
export { PersonaAlternateExample };
