// office-ui-fabric-react/lib-amd/components/FocusTrapZone/FocusTrapZone.doc.js
define(["require", "exports", "react", "./examples/FocusTrapZone.Box.Example", "./examples/FocusTrapZone.Box.FocusOnCustomElement.Example", "./examples/FocusTrapZone.Box.Click.Example", "./examples/FocusTrapZone.Nested.Example", "./examples/FocusTrapZone.NoTabbable.Example", "./examples/FocusTrapZone.FocusZone.Example", "./examples/FocusTrapZone.DialogInPanel.Example"], function (require, exports, React, FocusTrapZone_Box_Example_1, FocusTrapZone_Box_FocusOnCustomElement_Example_1, FocusTrapZone_Box_Click_Example_1, FocusTrapZone_Nested_Example_1, FocusTrapZone_NoTabbable_Example_1, FocusTrapZone_FocusZone_Example_1, FocusTrapZone_DialogInPanel_Example_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var FocusTrapZoneBoxExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.Box.Example.tsx');
    var FocusTrapZoneBoxExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.Box.Example.tsx');
    var FocusTrapZoneBoxCustomElementExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.Box.FocusOnCustomElement.Example.tsx');
    var FocusTrapZoneBoxCustomElementExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.Box.FocusOnCustomElement.Example.tsx');
    var FocusTrapZoneBoxClickExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.Box.Click.Example.tsx');
    var FocusTrapZoneBoxClickExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.Box.Click.Example.tsx');
    var FocusTrapZoneNestedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.Nested.Example.tsx');
    var FocusTrapZoneNestedExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.Nested.Example.tsx');
    var FocusTrapZoneNoTabbableExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.NoTabbable.Example.tsx');
    var FocusTrapZoneNoTabbableExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.NoTabbable.Example.tsx');
    var FocusTrapZoneFocusZoneExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.FocusZone.Example.tsx');
    var FocusTrapZoneFocusZoneExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.FocusZone.Example.tsx');
    var FocusTrapZoneDialogInPanelExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.DialogInPanel.Example.tsx');
    var FocusTrapZoneDialogInPanelExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.DialogInPanel.Example.tsx');
    exports.FocusTrapZonePageProps = {
        title: 'FocusTrapZone',
        componentName: 'FocusTrapZone',
        componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/FocusTrapZone',
        examples: [
            {
                title: 'Simple box',
                code: FocusTrapZoneBoxExampleCode,
                codepenJS: FocusTrapZoneBoxExampleCodepen,
                view: React.createElement(FocusTrapZone_Box_Example_1.FocusTrapZoneBoxExample, null)
            },
            {
                title: 'Simple box with focus on custom focusable element',
                code: FocusTrapZoneBoxCustomElementExampleCode,
                codepenJS: FocusTrapZoneBoxCustomElementExampleCodepen,
                view: React.createElement(FocusTrapZone_Box_FocusOnCustomElement_Example_1.FocusTrapZoneBoxCustomElementExample, null)
            },
            {
                title: 'Simple box with clicking outside trap zone enabled',
                code: FocusTrapZoneBoxClickExampleCode,
                codepenJS: FocusTrapZoneBoxClickExampleCodepen,
                view: React.createElement(FocusTrapZone_Box_Click_Example_1.FocusTrapZoneBoxClickExample, null)
            },
            {
                title: 'Multiple nested FocusTrapZones',
                code: FocusTrapZoneNestedExampleCode,
                codepenJS: FocusTrapZoneNestedExampleCodepen,
                view: React.createElement(FocusTrapZone_Nested_Example_1.FocusTrapZoneNestedExample, null)
            },
            {
                title: 'FocusTrapZone with FocusZones',
                code: FocusTrapZoneFocusZoneExampleCode,
                codepenJS: FocusTrapZoneFocusZoneExampleCodepen,
                view: React.createElement(FocusTrapZone_FocusZone_Example_1.FocusTrapZoneFocusZoneExample, null)
            },
            {
                title: 'FocusTrapZone with no tabbable elements',
                code: FocusTrapZoneNoTabbableExampleCode,
                codepenJS: FocusTrapZoneNoTabbableExampleCodepen,
                view: React.createElement(FocusTrapZone_NoTabbable_Example_1.FocusTrapZoneNoTabbableExample, null)
            },
            {
                title: 'A Dialog nested in a Panel',
                code: FocusTrapZoneDialogInPanelExampleCode,
                codepenJS: FocusTrapZoneDialogInPanelExampleCodepen,
                view: React.createElement(FocusTrapZone_DialogInPanel_Example_1.FocusTrapZoneDialogInPanelExample, null)
            }
        ],
        propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/FocusTrapZone/FocusTrapZone.types.ts')],
        overview: require('!raw-loader!office-ui-fabric-react/src/components/FocusTrapZone/docs/FocusTrapZoneOverview.md'),
        bestPractices: '',
        dos: '',
        donts: '',
        isHeaderVisible: true,
        isFeedbackVisible: true,
        allowNativeProps: true
    };
});
