// office-ui-fabric-react/lib/components/Keytip/Keytips.doc.js
import * as React from 'react';
import { KeytipStatus } from './Keytip.checklist';
import { KeytipsBasicExample } from './examples/Keytips.Basic.Example';
import { KeytipsButtonExample } from './examples/Keytips.Button.Example';
import { KeytipsCommandBarExample } from './examples/Keytips.CommandBar.Example';
import { KeytipsOverflowExample } from './examples/Keytips.Overflow.Example';
var KeytipsBasicCode = require('!raw-loader!office-ui-fabric-react/src/components/Keytip/examples/Keytips.Basic.Example.tsx');
var KeytipsButtonCode = require('!raw-loader!office-ui-fabric-react/src/components/Keytip/examples/Keytips.Button.Example.tsx');
var KeytipsCommandBarCode = require('!raw-loader!office-ui-fabric-react/src/components/Keytip/examples/Keytips.CommandBar.Example.tsx');
var KeytipsOverflowCode = require('!raw-loader!office-ui-fabric-react/src/components/Keytip/examples/Keytips.Overflow.Example.tsx');
export var KeytipsPageProps = {
    title: 'Keytips',
    componentName: 'Keytips',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Keytips',
    componentStatus: KeytipStatus,
    examples: [
        {
            title: 'Keytips on Buttons',
            code: KeytipsButtonCode,
            view: React.createElement(KeytipsButtonExample, null)
        },
        {
            title: 'Keytips in a CommandBar',
            code: KeytipsCommandBarCode,
            view: React.createElement(KeytipsCommandBarExample, null)
        },
        {
            title: 'Keytips in an OverflowWell',
            code: KeytipsOverflowCode,
            view: React.createElement(KeytipsOverflowExample, null)
        },
        {
            title: 'Keytips in Pivots',
            code: KeytipsBasicCode,
            view: React.createElement(KeytipsBasicExample, null)
        }
    ],
    propertiesTablesSources: [
        require('!raw-loader!office-ui-fabric-react/src/components/Keytip/Keytip.types.ts'),
        require('!raw-loader!office-ui-fabric-react/src/components/KeytipLayer/KeytipLayer.types.ts')
    ],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Keytip/docs/KeytipOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Keytip/docs/KeytipDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Keytip/docs/KeytipDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
