// office-ui-fabric-react/lib/components/Checkbox/Checkbox.doc.js
import * as React from 'react';
import { CheckboxBasicExample } from './examples/Checkbox.Basic.Example';
import { CheckboxImplementationExamples } from './examples/Checkbox.ImplementationExamples';
import { CheckboxStatus } from './Checkbox.checklist';
var CheckboxBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Checkbox/examples/Checkbox.Basic.Example.tsx');
var CheckboxBasicExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Checkbox/Checkbox.Basic.Example.Codepen.txt');
var CheckboxImplementationExamplesCode = require('!raw-loader!office-ui-fabric-react/src/components/Checkbox/examples/Checkbox.ImplementationExamples.tsx');
export var CheckboxPageProps = {
    title: 'Checkbox',
    componentName: 'Checkbox',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Checkbox',
    componentStatus: CheckboxStatus,
    examples: [
        {
            title: 'Default Checkbox',
            code: CheckboxBasicExampleCode,
            view: React.createElement(CheckboxBasicExample, null),
            codepenJS: CheckboxBasicExampleCodepen
        },
        {
            title: 'Implementation Examples',
            code: CheckboxImplementationExamplesCode,
            view: React.createElement(CheckboxImplementationExamples, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Checkbox/docs/CheckboxOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Checkbox/docs/CheckboxDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Checkbox/docs/CheckboxDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
