// office-ui-fabric-react/lib-commonjs/components/MarqueeSelection/MarqueeSelection.doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var MarqueeSelection_Basic_Example_1 = require("./examples/MarqueeSelection.Basic.Example");
var MarqueeSelectionBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/MarqueeSelection/examples/MarqueeSelection.Basic.Example.tsx');
exports.MarqueeSelectionPageProps = {
    title: 'MarqueeSelection',
    componentName: 'MarqueeSelection',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/MarqueeSelection',
    examples: [
        {
            title: 'Basic Selection Example',
            code: MarqueeSelectionBasicExampleCode,
            view: React.createElement(MarqueeSelection_Basic_Example_1.MarqueeSelectionBasicExample, null)
        }
    ],
    propertiesTablesSources: [],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/MarqueeSelection/docs/MarqueeSelectionOverview.md'),
    bestPractices: '',
    dos: '',
    donts: '',
    isHeaderVisible: true,
    isFeedbackVisible: true
};
