// office-ui-fabric-react/lib/components/Tooltip/examples/Tooltip.Absolute.Position.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { DefaultButton } from 'office-ui-fabric-react/lib/Button';
import { TooltipHost } from 'office-ui-fabric-react/lib/Tooltip';
import { getId } from 'office-ui-fabric-react/lib/Utilities';
var TooltipAbsolutePositionExample = /** @class */ (function (_super) {
    tslib_1.__extends(TooltipAbsolutePositionExample, _super);
    function TooltipAbsolutePositionExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        // Use getId() to ensure that the ID is unique on the page.
        // (It's also okay to use a plain string without getId() and manually ensure uniqueness.)
        _this._hostId = getId('tooltipHost');
        _this._buttonId = getId('targetButton');
        return _this;
    }
    TooltipAbsolutePositionExample.prototype.render = function () {
        return (React.createElement("div", { style: {
                minHeight: 50
            } },
            React.createElement(TooltipHost, { content: "This is the tooltip", id: this._hostId, calloutProps: {
                    gapSpace: 0,
                    target: "#" + this._buttonId
                } },
                React.createElement(DefaultButton, { id: this._buttonId, "aria-labelledby": this._hostId, style: {
                        position: 'absolute',
                        top: 50,
                        left: 200
                    } }, "Hover Over Me"))));
    };
    return TooltipAbsolutePositionExample;
}(React.Component));
export { TooltipAbsolutePositionExample };
