// office-ui-fabric-react/lib/components/Button/examples/Button.Command.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { CommandButton } from 'office-ui-fabric-react/lib/Button';
import { Label } from 'office-ui-fabric-react/lib/Label';
import { css, classNamesFunction } from 'office-ui-fabric-react/lib/Utilities';
import { getStyles } from './Button.Basic.Example.styles';
var ButtonCommandExample = /** @class */ (function (_super) {
    tslib_1.__extends(ButtonCommandExample, _super);
    function ButtonCommandExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ButtonCommandExample.prototype.render = function () {
        var _a = this.props, disabled = _a.disabled, checked = _a.checked;
        var getClassNames = classNamesFunction();
        var classNames = getClassNames(getStyles, {});
        return (React.createElement("div", { className: css(classNames.twoup) },
            React.createElement("div", null,
                React.createElement(Label, null, "CommandButton that focuses on menu"),
                React.createElement(CommandButton, { "data-automation-id": "test", disabled: disabled, checked: checked, iconProps: { iconName: 'Add' }, text: "Create account", menuProps: {
                        items: [
                            {
                                key: 'emailMessage',
                                text: 'Email message',
                                iconProps: { iconName: 'Mail' }
                            },
                            {
                                key: 'calendarEvent',
                                text: 'Calendar event',
                                iconProps: { iconName: 'Calendar' }
                            }
                        ]
                    } })),
            React.createElement("div", null,
                React.createElement(Label, null, "CommandButton that doesn't focus on menu"),
                React.createElement(CommandButton, { "data-automation-id": "test", disabled: disabled, checked: checked, iconProps: { iconName: 'Add' }, text: "Create account", menuProps: {
                        shouldFocusOnMount: false,
                        items: [
                            {
                                key: 'emailMessage',
                                text: 'Email message',
                                iconProps: { iconName: 'Mail' }
                            },
                            {
                                key: 'calendarEvent',
                                text: 'Calendar event',
                                iconProps: { iconName: 'Calendar' }
                            }
                        ]
                    } }))));
    };
    return ButtonCommandExample;
}(React.Component));
export { ButtonCommandExample };
