// office-ui-fabric-react/lib/components/Button/examples/Button.CommandBar.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { CommandBarButton } from 'office-ui-fabric-react/lib/Button';
var ButtonCommandBarExample = /** @class */ (function (_super) {
    tslib_1.__extends(ButtonCommandBarExample, _super);
    function ButtonCommandBarExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ButtonCommandBarExample.prototype.render = function () {
        var _a = this.props, disabled = _a.disabled, checked = _a.checked;
        var alertClicked = function () {
            alert('Clicked');
        };
        return (React.createElement("div", null,
            React.createElement("div", { style: { display: 'flex', alignItems: 'stretch', height: '40px' } },
                React.createElement(CommandBarButton, { "data-automation-id": "test", disabled: disabled, checked: checked, iconProps: { iconName: 'Add' }, text: "Create account", menuProps: {
                        items: [
                            {
                                key: 'emailMessage',
                                text: 'Email message',
                                iconProps: { iconName: 'Mail' }
                            },
                            {
                                key: 'calendarEvent',
                                text: 'Calendar event',
                                iconProps: { iconName: 'Calendar' }
                            }
                        ]
                    } }),
                React.createElement(CommandBarButton, { "data-automation-id": "test", disabled: disabled, checked: checked, iconProps: { iconName: 'Add' }, text: "Create account", split: true, onClick: alertClicked, menuProps: {
                        items: [
                            {
                                key: 'emailMessage',
                                name: 'Email message',
                                icon: 'Mail'
                            },
                            {
                                key: 'calendarEvent',
                                name: 'Calendar event',
                                icon: 'Calendar'
                            }
                        ]
                    } }),
                React.createElement(CommandBarButton, { "data-automation-id": "test2", disabled: disabled, checked: checked, iconProps: { iconName: 'Mail' }, text: "Send Mail" }))));
    };
    return ButtonCommandBarExample;
}(React.Component));
export { ButtonCommandBarExample };
