// office-ui-fabric-react/lib/components/TextField/TextField.doc.js
import * as React from 'react';
import { TextFieldBasicExample } from './examples/TextField.Basic.Example';
import { TextFieldBorderlessExample } from './examples/TextField.Borderless.Example';
import { TextFieldCustomRenderExample } from './examples/TextField.CustomRender.Example';
import { TextFieldErrorMessageExample } from './examples/TextField.ErrorMessage.Example';
import { TextFieldMultilineExample } from './examples/TextField.Multiline.Example';
import { TextFieldControlledExample } from './examples/TextField.Controlled.Example';
import { TextFieldPrefixAndSuffixExample } from './examples/TextField.PrefixAndSuffix.Example';
import { TextFieldStatus } from './TextField.checklist';
import { TextFieldStyledExample } from './examples/TextField.Styled.Example';
var TextFieldBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Basic.Example.tsx');
var TextFieldBasicExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Basic.Example.tsx');
var TextFieldBorderlessExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Borderless.Example.tsx');
var TextFieldBorderlessExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Borderless.Example.tsx');
var TextFieldCustomRenderExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.CustomRender.Example.tsx');
var TextFieldCustomRenderExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.CustomRender.Example.tsx');
var TextFieldErrorMessageExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.ErrorMessage.Example.tsx');
var TextFieldErrorMessageExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.ErrorMessage.Example.tsx');
var TextFieldMultilineExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Multiline.Example.tsx');
var TextFieldMultilineExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Multiline.Example.tsx');
var TextFieldControlledExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Controlled.Example.tsx');
var TextFieldControlledExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Controlled.Example.tsx');
var TextFieldPrefixAndSuffixExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.PrefixAndSuffix.Example.tsx');
var TextFieldPrefixAndSuffixExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.PrefixAndSuffix.Example.tsx');
var TextFieldStyledExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Styled.Example.tsx');
var TextFieldStyledExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Styled.Example.tsx');
export var TextFieldPageProps = {
    title: 'TextField',
    componentName: 'TextField',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/TextField',
    componentStatus: TextFieldStatus,
    examples: [
        {
            title: 'Basic TextFields',
            code: TextFieldBasicExampleCode,
            view: React.createElement(TextFieldBasicExample, null),
            codepenJS: TextFieldBasicExampleCodepen
        },
        {
            title: 'Controlled TextFields',
            code: TextFieldControlledExampleCode,
            view: React.createElement(TextFieldControlledExample, null),
            codepenJS: TextFieldControlledExampleCodepen
        },
        {
            title: 'Multiline TextField',
            code: TextFieldMultilineExampleCode,
            codepenJS: TextFieldMultilineExampleCodepen,
            view: React.createElement(TextFieldMultilineExample, null)
        },
        {
            title: 'Underlined and borderless TextFields',
            code: TextFieldBorderlessExampleCode,
            view: React.createElement(TextFieldBorderlessExample, null),
            codepenJS: TextFieldBorderlessExampleCodepen
        },
        {
            title: 'TextField with prefix and/or suffix',
            code: TextFieldPrefixAndSuffixExampleCode,
            codepenJS: TextFieldPrefixAndSuffixExampleCodepen,
            view: React.createElement(TextFieldPrefixAndSuffixExample, null)
        },
        {
            title: 'TextFields with custom rendering',
            code: TextFieldCustomRenderExampleCode,
            codepenJS: TextFieldCustomRenderExampleCodepen,
            view: React.createElement(TextFieldCustomRenderExample, null)
        },
        {
            title: 'TextField error message variations',
            code: TextFieldErrorMessageExampleCode,
            codepenJS: TextFieldErrorMessageExampleCodepen,
            view: React.createElement(TextFieldErrorMessageExample, null)
        },
        {
            title: 'TextField subcomponent styling',
            code: TextFieldStyledExampleCode,
            codepenJS: TextFieldStyledExampleCodepen,
            view: React.createElement(TextFieldStyledExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/TextField/TextField.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/TextField/docs/TextFieldOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/TextField/docs/TextFieldDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/TextField/docs/TextFieldDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true,
    allowNativeProps: true,
    nativePropsElement: ['input', 'textarea']
};
