// office-ui-fabric-react/lib/components/FocusTrapZone/FocusTrapZone.doc.js
import * as React from 'react';
import { FocusTrapZoneBoxExample } from './examples/FocusTrapZone.Box.Example';
var FocusTrapZoneBoxExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.Box.Example.tsx');
var FocusTrapZoneBoxExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.Box.Example.tsx');
import { FocusTrapZoneBoxCustomElementExample } from './examples/FocusTrapZone.Box.FocusOnCustomElement.Example';
var FocusTrapZoneBoxCustomElementExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.Box.FocusOnCustomElement.Example.tsx');
var FocusTrapZoneBoxCustomElementExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.Box.FocusOnCustomElement.Example.tsx');
import { FocusTrapZoneBoxClickExample } from './examples/FocusTrapZone.Box.Click.Example';
var FocusTrapZoneBoxClickExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.Box.Click.Example.tsx');
var FocusTrapZoneBoxClickExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.Box.Click.Example.tsx');
import { FocusTrapZoneNestedExample } from './examples/FocusTrapZone.Nested.Example';
var FocusTrapZoneNestedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.Nested.Example.tsx');
var FocusTrapZoneNestedExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.Nested.Example.tsx');
import { FocusTrapZoneNoTabbableExample } from './examples/FocusTrapZone.NoTabbable.Example';
var FocusTrapZoneNoTabbableExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.NoTabbable.Example.tsx');
var FocusTrapZoneNoTabbableExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.NoTabbable.Example.tsx');
import { FocusTrapZoneFocusZoneExample } from './examples/FocusTrapZone.FocusZone.Example';
var FocusTrapZoneFocusZoneExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.FocusZone.Example.tsx');
var FocusTrapZoneFocusZoneExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.FocusZone.Example.tsx');
import { FocusTrapZoneDialogInPanelExample } from './examples/FocusTrapZone.DialogInPanel.Example';
var FocusTrapZoneDialogInPanelExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.DialogInPanel.Example.tsx');
var FocusTrapZoneDialogInPanelExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusTrapZone/examples/FocusTrapZone.DialogInPanel.Example.tsx');
export var FocusTrapZonePageProps = {
    title: 'FocusTrapZone',
    componentName: 'FocusTrapZone',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/FocusTrapZone',
    examples: [
        {
            title: 'Simple box',
            code: FocusTrapZoneBoxExampleCode,
            codepenJS: FocusTrapZoneBoxExampleCodepen,
            view: React.createElement(FocusTrapZoneBoxExample, null)
        },
        {
            title: 'Simple box with focus on custom focusable element',
            code: FocusTrapZoneBoxCustomElementExampleCode,
            codepenJS: FocusTrapZoneBoxCustomElementExampleCodepen,
            view: React.createElement(FocusTrapZoneBoxCustomElementExample, null)
        },
        {
            title: 'Simple box with clicking outside trap zone enabled',
            code: FocusTrapZoneBoxClickExampleCode,
            codepenJS: FocusTrapZoneBoxClickExampleCodepen,
            view: React.createElement(FocusTrapZoneBoxClickExample, null)
        },
        {
            title: 'Multiple nested FocusTrapZones',
            code: FocusTrapZoneNestedExampleCode,
            codepenJS: FocusTrapZoneNestedExampleCodepen,
            view: React.createElement(FocusTrapZoneNestedExample, null)
        },
        {
            title: 'FocusTrapZone with FocusZones',
            code: FocusTrapZoneFocusZoneExampleCode,
            codepenJS: FocusTrapZoneFocusZoneExampleCodepen,
            view: React.createElement(FocusTrapZoneFocusZoneExample, null)
        },
        {
            title: 'FocusTrapZone with no tabbable elements',
            code: FocusTrapZoneNoTabbableExampleCode,
            codepenJS: FocusTrapZoneNoTabbableExampleCodepen,
            view: React.createElement(FocusTrapZoneNoTabbableExample, null)
        },
        {
            title: 'A Dialog nested in a Panel',
            code: FocusTrapZoneDialogInPanelExampleCode,
            codepenJS: FocusTrapZoneDialogInPanelExampleCodepen,
            view: React.createElement(FocusTrapZoneDialogInPanelExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/FocusTrapZone/FocusTrapZone.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/FocusTrapZone/docs/FocusTrapZoneOverview.md'),
    bestPractices: '',
    dos: '',
    donts: '',
    isHeaderVisible: true,
    isFeedbackVisible: true,
    allowNativeProps: true
};
