// office-ui-fabric-react/lib-amd/components/ColorPicker/ColorPicker.styles.js
define(["require", "exports", "tslib"], function (require, exports, tslib_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getStyles = function (props) {
        var className = props.className, theme = props.theme;
        return {
            root: [
                'ms-ColorPicker',
                theme.fonts.medium,
                {
                    position: 'relative',
                    maxWidth: 300
                },
                className
            ],
            panel: [
                'ms-ColorPicker-panel',
                {
                    padding: '16px'
                }
            ],
            table: [
                'ms-ColorPicker-table',
                {
                    tableLayout: 'fixed',
                    width: '100%'
                }
            ],
            tableHeader: [
                tslib_1.__assign({}, theme.fonts.small)
            ],
            tableHexCell: [
                {
                    width: '25%'
                }
            ],
            input: [
                'ms-ColorPicker-input',
                {
                    width: '100%',
                    border: 'none',
                    boxSizing: 'border-box',
                    height: 30,
                    selectors: {
                        '&.ms-TextField': {
                            paddingRight: 2
                        },
                        '& .ms-TextField-field': {
                            minWidth: 'auto',
                            padding: 5,
                            textOverflow: 'clip'
                        }
                    }
                }
            ]
        };
    };
});
