// office-ui-fabric-react/lib-amd/components/Spinner/Spinner.doc.js
define(["require", "exports", "react", "./Spinner.checklist", "./examples/Spinner.Basic.Example", "./examples/Spinner.Labeled.Example"], function (require, exports, React, Spinner_checklist_1, Spinner_Basic_Example_1, Spinner_Labeled_Example_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var SpinnerBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Spinner/examples/Spinner.Basic.Example.tsx');
    var SpinnerBasicExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Spinner/examples/Spinner.Basic.Example.tsx');
    var SpinnerLabeledExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Spinner/examples/Spinner.Labeled.Example.tsx');
    var SpinnerLabeledExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Spinner/examples/Spinner.Labeled.Example.tsx');
    exports.SpinnerPageProps = {
        title: 'Spinner',
        componentName: 'Spinner',
        componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Spinner',
        componentStatus: Spinner_checklist_1.SpinnerStatus,
        examples: [
            {
                title: 'Spinner sizes',
                code: SpinnerBasicExampleCode,
                view: React.createElement(Spinner_Basic_Example_1.SpinnerBasicExample, null),
                codepenJS: SpinnerBasicExampleCodepen
            },
            {
                title: 'Spinner label positioning',
                code: SpinnerLabeledExampleCode,
                view: React.createElement(Spinner_Labeled_Example_1.SpinnerLabeledExample, null),
                codepenJS: SpinnerLabeledExampleCodepen
            }
        ],
        propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Spinner/Spinner.types.ts')],
        overview: require('!raw-loader!office-ui-fabric-react/src/components/Spinner/docs/SpinnerOverview.md'),
        bestPractices: '',
        dos: require('!raw-loader!office-ui-fabric-react/src/components/Spinner/docs/SpinnerDos.md'),
        donts: require('!raw-loader!office-ui-fabric-react/src/components/Spinner/docs/SpinnerDonts.md'),
        isHeaderVisible: true,
        isFeedbackVisible: true
    };
});
