// office-ui-fabric-react/lib-amd/components/Spinner/Spinner.doc.js
define(["require", "exports", "react", "./examples/Spinner.Basic.Example", "./Spinner.checklist"], function (require, exports, React, Spinner_Basic_Example_1, Spinner_checklist_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var SpinnerBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Spinner/examples/Spinner.Basic.Example.tsx');
    var SpinnerBasicExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Spinner/Spinner.Basic.Example.Codepen.txt');
    exports.SpinnerPageProps = {
        title: 'Spinner',
        componentName: 'Spinner',
        componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Spinner',
        componentStatus: Spinner_checklist_1.SpinnerStatus,
        examples: [
            {
                title: 'Various Spinner Types',
                code: SpinnerBasicExampleCode,
                view: React.createElement(Spinner_Basic_Example_1.SpinnerBasicExample, null),
                codepenJS: SpinnerBasicExampleCodepen
            }
        ],
        propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Spinner/Spinner.types.ts')],
        overview: require('!raw-loader!office-ui-fabric-react/src/components/Spinner/docs/SpinnerOverview.md'),
        bestPractices: '',
        dos: require('!raw-loader!office-ui-fabric-react/src/components/Spinner/docs/SpinnerDos.md'),
        donts: require('!raw-loader!office-ui-fabric-react/src/components/Spinner/docs/SpinnerDonts.md'),
        isHeaderVisible: true,
        isFeedbackVisible: true
    };
});
