// office-ui-fabric-react/lib-commonjs/components/ContextualMenu/ContextualMenu.doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var ContextualMenu_checklist_1 = require("./ContextualMenu.checklist");
var ContextualMenu_Basic_Example_1 = require("./examples/ContextualMenu.Basic.Example");
var ContextualMenu_Icon_Example_1 = require("./examples/ContextualMenu.Icon.Example");
var ContextualMenu_Icon_SecondaryText_Example_1 = require("./examples/ContextualMenu.Icon.SecondaryText.Example");
var ContextualMenu_Submenu_Example_1 = require("./examples/ContextualMenu.Submenu.Example");
var ContextualMenu_Section_Example_1 = require("./examples/ContextualMenu.Section.Example");
var ContextualMenu_Checkmarks_Example_1 = require("./examples/ContextualMenu.Checkmarks.Example");
var ContextualMenu_Directional_Example_1 = require("./examples/ContextualMenu.Directional.Example");
var ContextualMenu_Customization_Example_1 = require("./examples/ContextualMenu.Customization.Example");
var ContextualMenu_CustomizationWithNoWrap_Example_1 = require("./examples/ContextualMenu.CustomizationWithNoWrap.Example");
var ContextualMenu_ScrollBar_Example_1 = require("./examples/ContextualMenu.ScrollBar.Example");
var ContextualMenu_CustomMenuItem_Example_1 = require("./examples/ContextualMenu.CustomMenuItem.Example");
var ContextualMenu_CustomMenuList_Example_1 = require("./examples/ContextualMenu.CustomMenuList.Example");
var ContextualMenu_Header_Example_1 = require("./examples/ContextualMenu.Header.Example");
var ContextualMenu_Persisted_Example_1 = require("./examples/ContextualMenu.Persisted.Example");
var ContextualMenuBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/examples/ContextualMenu.Basic.Example.tsx');
var ContextualMenuPersistedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/examples/ContextualMenu.Persisted.Example.tsx');
var ContextualMenuBasicExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/ContextualMenu/examples/ContextualMenu.Basic.Example.tsx');
var ContextualMenuIconExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/examples/ContextualMenu.Icon.Example.tsx');
var ContextualMenuIconSecondaryTextExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/examples/ContextualMenu.Icon.SecondaryText.Example.tsx');
var ContextualMenuSubmenuExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/examples/ContextualMenu.Submenu.Example.tsx');
var ContextualMenuSectionExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/examples/ContextualMenu.Section.Example.tsx');
var ContextualMenuCheckmarksExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/examples/ContextualMenu.Checkmarks.Example.tsx');
var ContextualMenuDirectionalExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/examples/ContextualMenu.Directional.Example.tsx');
var ContextualMenuCustomizationExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/examples/ContextualMenu.Customization.Example.tsx');
var ContextualMenuCustomizationWithNoWrapExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/examples/ContextualMenu.CustomizationWithNoWrap.Example.tsx');
var ContextualMenuWithScrollBarExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/examples/ContextualMenu.ScrollBar.Example.tsx');
var ContextualMenuWithCustomMenuItemExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/examples/ContextualMenu.CustomMenuItem.Example.tsx');
var ContextualMenuCustomMenuListExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/examples/ContextualMenu.CustomMenuList.Example.tsx');
var ContextualMenuHeaderExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/examples/ContextualMenu.Header.Example.tsx');
exports.ContextualMenuPageProps = {
    title: 'ContextualMenu',
    componentName: 'ContextualMenu',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/ContextualMenu',
    componentStatus: ContextualMenu_checklist_1.ContextualMenuStatus,
    examples: [
        {
            title: 'Default ContextualMenu',
            code: ContextualMenuBasicExampleCode,
            view: React.createElement(ContextualMenu_Basic_Example_1.ContextualMenuBasicExample, null),
            codepenJS: ContextualMenuBasicExampleCodepen
        },
        {
            title: 'ContextualMenu which is persisted in the DOM',
            code: ContextualMenuPersistedExampleCode,
            view: React.createElement(ContextualMenu_Persisted_Example_1.ContextualMenuPersistedExample, null)
        },
        {
            title: 'ContextualMenu with icons',
            code: ContextualMenuIconExampleCode,
            view: React.createElement(ContextualMenu_Icon_Example_1.ContextualMenuIconExample, null)
        },
        {
            title: 'ContextualMenu with icons and secondary text',
            code: ContextualMenuIconSecondaryTextExampleCode,
            view: React.createElement(ContextualMenu_Icon_SecondaryText_Example_1.ContextualMenuIconSecondaryTextExample, null)
        },
        {
            title: 'ContextualMenu with submenus',
            code: ContextualMenuSubmenuExampleCode,
            view: React.createElement(ContextualMenu_Submenu_Example_1.ContextualMenuSubmenuExample, null)
        },
        {
            title: 'ContextualMenu with section headers',
            code: ContextualMenuSectionExampleCode,
            view: React.createElement(ContextualMenu_Section_Example_1.ContextualMenuSectionExample, null)
        },
        {
            title: 'ContextualMenu with checkable menu items and toggleable split button',
            code: ContextualMenuCheckmarksExampleCode,
            view: React.createElement(ContextualMenu_Checkmarks_Example_1.ContextualMenuCheckmarksExample, null)
        },
        {
            title: 'ContextualMenu with beak and directional settings',
            code: ContextualMenuDirectionalExampleCode,
            view: React.createElement(ContextualMenu_Directional_Example_1.ContextualMenuDirectionalExample, null)
        },
        {
            title: 'ContextualMenu with customized submenus',
            code: ContextualMenuCustomizationExampleCode,
            view: React.createElement(ContextualMenu_Customization_Example_1.ContextualMenuCustomizationExample, null)
        },
        {
            title: 'ContextualMenu with customized submenus and noWrap attributes',
            code: ContextualMenuCustomizationWithNoWrapExampleCode,
            view: React.createElement(ContextualMenu_CustomizationWithNoWrap_Example_1.ContextualMenuCustomizationWithNoWrapExample, null)
        },
        {
            title: 'ContextualMenu with a scroll bar and fixed direction',
            code: ContextualMenuWithScrollBarExampleCode,
            view: React.createElement(ContextualMenu_ScrollBar_Example_1.ContextualMenuWithScrollBarExample, null)
        },
        {
            title: 'ContextualMenu with custom rendered menu items',
            code: ContextualMenuWithCustomMenuItemExampleCode,
            view: React.createElement(ContextualMenu_CustomMenuItem_Example_1.ContextualMenuWithCustomMenuItemExample, null)
        },
        {
            title: 'ContextualMenu with custom rendered menu list that renders a search box to filter menu items',
            code: ContextualMenuCustomMenuListExampleCode,
            view: React.createElement(ContextualMenu_CustomMenuList_Example_1.ContextualMenuWithCustomMenuListExample, null)
        },
        {
            title: 'ContextualMenu with header',
            code: ContextualMenuHeaderExampleCode,
            view: React.createElement(ContextualMenu_Header_Example_1.ContextualMenuHeaderExample, null)
        }
    ],
    propertiesTablesSources: [
        require('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/ContextualMenu.types.ts'),
        require('!raw-loader!office-ui-fabric-react/src/components/Callout/Callout.types.ts')
    ],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/docs/ContextualMenuOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/docs/ContextualMenuDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/docs/ContextualMenuDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
