// office-ui-fabric-react/lib/components/TeachingBubble/TeachingBubble.doc.js
import * as React from 'react';
import { TeachingBubbleBasicExample } from './examples/TeachingBubble.Basic.Example';
import { TeachingBubbleCondensedExample } from './examples/TeachingBubble.Condensed.Example';
import { TeachingBubbleIllustrationExample } from './examples/TeachingBubble.Illustration.Example';
import { TeachingBubbleWideIllustrationExample } from './examples/TeachingBubble.WideIllustration.Example';
import { TeachingBubbleSmallHeadlineExample } from './examples/TeachingBubble.SmallHeadline.Example';
import { TeachingBubbleStatus } from './TeachingBubble.checklist';
var TeachingBubbleBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/examples/TeachingBubble.Basic.Example.tsx');
var TeachingBubbleCondensedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/examples/TeachingBubble.Condensed.Example.tsx');
var TeachingBubbleIllustrationExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/examples/TeachingBubble.Basic.Example.tsx');
var TeachingBubbleWideIllustrationExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/examples/TeachingBubble.WideIllustration.Example.tsx');
var TeachingBubbleSmallHeadlineExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/examples/TeachingBubble.SmallHeadline.Example.tsx');
export var TeachingBubblePageProps = {
    title: 'TeachingBubble',
    componentName: 'TeachingBubble',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/TeachingBubble',
    componentStatus: TeachingBubbleStatus,
    examples: [
        {
            title: 'TeachingBubble',
            code: TeachingBubbleBasicExampleCode,
            view: React.createElement(TeachingBubbleBasicExample, null)
        },
        {
            title: 'TeachingBubble Condensed',
            code: TeachingBubbleCondensedExampleCode,
            view: React.createElement(TeachingBubbleCondensedExample, null)
        },
        {
            title: 'TeachingBubble with Illustration',
            code: TeachingBubbleIllustrationExampleCode,
            view: React.createElement(TeachingBubbleIllustrationExample, null)
        },
        {
            title: 'TeachingBubble wide with Illustration',
            code: TeachingBubbleWideIllustrationExampleCode,
            view: React.createElement(TeachingBubbleWideIllustrationExample, null)
        },
        {
            title: 'TeachingBubble with small headline',
            code: TeachingBubbleSmallHeadlineExampleCode,
            view: React.createElement(TeachingBubbleSmallHeadlineExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/TeachingBubble.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/docs/TeachingBubbleOverview.md'),
    dos: require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/docs/TeachingBubbleDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/docs/TeachingBubbleDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
