// office-ui-fabric-react/lib/components/Panel/Panel.doc.js
import * as React from 'react';
import { PanelCustomExample } from './examples/Panel.Custom.Example';
import { PanelCustomLeftExample } from './examples/Panel.CustomLeft.Example';
import { PanelExtraLargeExample } from './examples/Panel.ExtraLarge.Example';
import { PanelFooterExample } from './examples/Panel.Footer.Example';
import { PanelNavigationExample } from './examples/Panel.Navigation.Example';
import { PanelHandleDismissTargetExample } from './examples/Panel.HandleDismissTarget.Example';
import { PanelHiddenOnDismissExample } from './examples/Panel.HiddenOnDismiss.Example';
import { PanelLargeExample } from './examples/Panel.Large.Example';
import { PanelLargeFixedExample } from './examples/Panel.LargeFixed.Example';
import { PanelLightDismissExample } from './examples/Panel.LightDismiss.Example';
import { PanelLightDismissCustomExample } from './examples/Panel.LightDismissCustom.Example';
import { PanelMediumExample } from './examples/Panel.Medium.Example';
import { PanelNonModalExample } from './examples/Panel.NonModal.Example';
import { PanelPreventDefaultExample } from './examples/Panel.PreventDefault.Example';
import { PanelScrollExample } from './examples/Panel.Scroll.Example';
import { PanelSmallFluidExample } from './examples/Panel.SmallFluid.Example';
import { PanelSmallLeftExample } from './examples/Panel.SmallLeft.Example';
import { PanelSmallRightExample } from './examples/Panel.SmallRight.Example';
import { PanelStatus } from './Panel.checklist';
var PanelSmallRightExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.SmallRight.Example.tsx');
var PanelSmallRightExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Panel/Panel.SmallRight.Example.Codepen.txt');
var PanelSmallLeftExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.SmallLeft.Example.tsx');
var PanelSmallFluidExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.SmallFluid.Example.tsx');
var PanelMediumExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Medium.Example.tsx');
var PanelLargeExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Large.Example.tsx');
var PanelLargeFixedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.LargeFixed.Example.tsx');
var PanelExtraLargeExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.ExtraLarge.Example.tsx');
var PanelCustomExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Custom.Example.tsx');
var PanelCustomLeftExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.CustomLeft.Example.tsx');
var PanelHiddenOnDismissExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.HiddenOnDismiss.Example.tsx');
var PanelLightDismissExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.LightDismiss.Example.tsx');
var PanelLightDismissCustomExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.LightDismissCustom.Example.tsx');
var PanelNonModalExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.NonModal.Example.tsx');
var PanelFooterExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Footer.Example.tsx');
var PanelNavigationExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Navigation.Example.tsx');
var PanelPreventDefaultExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.PreventDefault.Example.tsx');
var PanelHandleDismissTargetExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.HandleDismissTarget.Example.tsx');
var PanelScrollExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Scroll.Example.tsx');
export var PanelPageProps = {
    title: 'Panel',
    componentName: 'Panel',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Panel',
    componentStatus: PanelStatus,
    examples: [
        {
            title: 'Panel - Small Panel, Anchored Right, Fixed Width',
            code: PanelSmallRightExampleCode,
            view: React.createElement(PanelSmallRightExample, null),
            codepenJS: PanelSmallRightExampleCodepen
        },
        {
            title: 'Panel - Small Panel, Anchored Left, Fixed Width',
            code: PanelSmallLeftExampleCode,
            view: React.createElement(PanelSmallLeftExample, null)
        },
        {
            title: 'Panel - Small Panel, Full Screen, Fluid Width',
            code: PanelSmallFluidExampleCode,
            view: React.createElement(PanelSmallFluidExample, null)
        },
        {
            title: 'Panel - Medium',
            code: PanelMediumExampleCode,
            view: React.createElement(PanelMediumExample, null)
        },
        {
            title: 'Panel - Large',
            code: PanelLargeExampleCode,
            view: React.createElement(PanelLargeExample, null)
        },
        {
            title: 'Panel - LargeFixed',
            code: PanelLargeFixedExampleCode,
            view: React.createElement(PanelLargeFixedExample, null)
        },
        {
            title: 'Panel - Extra Large',
            code: PanelExtraLargeExampleCode,
            view: React.createElement(PanelExtraLargeExample, null)
        },
        {
            title: 'Panel - Custom',
            code: PanelCustomExampleCode,
            view: React.createElement(PanelCustomExample, null)
        },
        {
            title: 'Panel - Custom Width, Anchored Left',
            code: PanelCustomLeftExampleCode,
            view: React.createElement(PanelCustomLeftExample, null)
        },
        {
            title: 'Panel - Hidden on Dismiss',
            code: PanelHiddenOnDismissExampleCode,
            view: React.createElement(PanelHiddenOnDismissExample, null)
        },
        {
            title: 'Panel - Light Dismiss',
            code: PanelLightDismissExampleCode,
            view: React.createElement(PanelLightDismissExample, null)
        },
        {
            title: 'Panel - Custom Light Dismiss',
            code: PanelLightDismissCustomExampleCode,
            view: React.createElement(PanelLightDismissCustomExample, null)
        },
        {
            title: 'Panel - Non-Modal',
            code: PanelNonModalExampleCode,
            view: React.createElement(PanelNonModalExample, null)
        },
        {
            title: 'Panel - Footer',
            code: PanelFooterExampleCode,
            view: React.createElement(PanelFooterExample, null)
        },
        {
            title: 'Panel - Navigation',
            code: PanelNavigationExampleCode,
            view: React.createElement(PanelNavigationExample, null)
        },
        {
            title: 'Panel - Prevent Default Sample',
            code: PanelPreventDefaultExampleCode,
            view: React.createElement(PanelPreventDefaultExample, null)
        },
        {
            title: 'Panel - Handle Dismiss Target Sample',
            code: PanelHandleDismissTargetExampleCode,
            view: React.createElement(PanelHandleDismissTargetExample, null)
        },
        {
            title: 'Panel - Scrolling Content Sample',
            code: PanelScrollExampleCode,
            view: React.createElement(PanelScrollExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Panel/Panel.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Panel/docs/PanelOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Panel/docs/PanelDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Panel/docs/PanelDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true,
    allowNativeProps: true
};
