// office-ui-fabric-react/lib-amd/components/OverflowSet/OverflowSet.doc.js
define(["require", "exports", "react", "./examples/OverflowSet.Custom.Example", "./examples/OverflowSet.Basic.Example", "./examples/OverflowSet.Vertical.Example", "./OverflowSet.checklist"], function (require, exports, React, OverflowSet_Custom_Example_1, OverflowSet_Basic_Example_1, OverflowSet_Vertical_Example_1, OverflowSet_checklist_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var OverflowSetCustomExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/OverflowSet/examples/OverflowSet.Custom.Example.tsx');
    var OverflowSetCustomExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/OverflowSet/examples/OverflowSet.Custom.Example.tsx');
    var OverflowSetBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/OverflowSet/examples/OverflowSet.Basic.Example.tsx');
    var OverflowSetBasicExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/OverflowSet/examples/OverflowSet.Basic.Example.tsx');
    var OverflowSetVerticalExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/OverflowSet/examples/OverflowSet.Vertical.Example.tsx');
    var OverflowSetVerticalExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/OverflowSet/examples/OverflowSet.Vertical.Example.tsx');
    exports.OverflowSetPageProps = {
        title: 'OverflowSet',
        componentName: 'OverflowSet',
        componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/OverflowSet',
        componentStatus: OverflowSet_checklist_1.OverflowSetStatus,
        examples: [
            {
                title: 'OverflowSet Basic Example',
                code: OverflowSetBasicExampleCode,
                codepenJS: OverflowSetBasicExampleCodepen,
                view: React.createElement(OverflowSet_Basic_Example_1.OverflowSetBasicExample, null)
            },
            {
                title: 'OverflowSet Vertical Example',
                code: OverflowSetVerticalExampleCode,
                codepenJS: OverflowSetVerticalExampleCodepen,
                view: React.createElement(OverflowSet_Vertical_Example_1.OverflowSetVerticalExample, null)
            },
            {
                title: 'OverflowSet Custom Example',
                code: OverflowSetCustomExampleCode,
                codepenJS: OverflowSetCustomExampleCodepen,
                view: React.createElement(OverflowSet_Custom_Example_1.OverflowSetCustomExample, null)
            }
        ],
        propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/OverflowSet/OverflowSet.types.ts')],
        overview: require('!raw-loader!office-ui-fabric-react/src/components/OverflowSet/docs/OverflowSetOverview.md'),
        bestPractices: '',
        dos: '',
        donts: '',
        isHeaderVisible: true,
        isFeedbackVisible: true,
        allowNativeProps: true
    };
});
