// office-ui-fabric-react/lib/components/Separator/examples/Separator.Theming.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { Separator } from 'office-ui-fabric-react/lib/Separator';
import { createTheme } from 'office-ui-fabric-react/lib/Styling';
import { Stack } from 'office-ui-fabric-react/lib/Stack';
import { Text } from 'office-ui-fabric-react/lib/Text';
var theme = createTheme({
    fonts: {
        medium: {
            fontFamily: 'Monaco, Menlo, Consolas',
            fontSize: '30px'
        }
    }
});
var SeparatorThemingExample = /** @class */ (function (_super) {
    tslib_1.__extends(SeparatorThemingExample, _super);
    function SeparatorThemingExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SeparatorThemingExample.prototype.render = function () {
        var content = 'Today';
        return (React.createElement(Stack, { gap: 12 },
            React.createElement(Text, null, "Horizontal center aligned with custom theme"),
            React.createElement(Separator, { theme: theme }, content)));
    };
    return SeparatorThemingExample;
}(React.Component));
export { SeparatorThemingExample };
