// office-ui-fabric-react/lib-amd/components/Nav/Nav.doc.js
define(["require", "exports", "react", "./examples/Nav.Basic.Example", "./examples/Nav.FabricDemoApp.Example", "./examples/Nav.Nested.Example", "./Nav.checklist", "office-ui-fabric-react/lib/components/Nav/examples/Nav.CustomGroupHeaders.Example"], function (require, exports, React, Nav_Basic_Example_1, Nav_FabricDemoApp_Example_1, Nav_Nested_Example_1, Nav_checklist_1, Nav_CustomGroupHeaders_Example_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var NavBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.Basic.Example.tsx');
    var NavBasicExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.Basic.Example.tsx');
    var NavFabricDemoAppExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.FabricDemoApp.Example.tsx');
    var NavFabricDemoAppExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.FabricDemoApp.Example.tsx');
    var NavNestedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.Nested.Example.tsx');
    var NavNestedExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.Nested.Example.tsx');
    var NavCustomGroupHeadersExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.CustomGroupHeaders.Example.tsx');
    var NavCustomGroupHeadersExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.CustomGroupHeaders.Example.tsx');
    exports.NavPageProps = {
        title: 'Nav',
        componentName: 'Nav',
        componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Nav',
        componentStatus: Nav_checklist_1.NavStatus,
        examples: [
            {
                title: 'Basic nav with sample links',
                code: NavBasicExampleCode,
                codepenJS: NavBasicExampleCodepen,
                view: React.createElement(Nav_Basic_Example_1.NavBasicExample, null)
            },
            {
                title: 'Nav similar to the one in this demo app',
                code: NavFabricDemoAppExampleCode,
                codepenJS: NavFabricDemoAppExampleCodepen,
                view: React.createElement(Nav_FabricDemoApp_Example_1.NavFabricDemoAppExample, null)
            },
            {
                title: 'Nav with nested links',
                code: NavNestedExampleCode,
                codepenJS: NavNestedExampleCodepen,
                view: React.createElement(Nav_Nested_Example_1.NavNestedExample, null)
            },
            {
                title: 'Nav with custom group header',
                code: NavCustomGroupHeadersExampleCode,
                codepenJS: NavCustomGroupHeadersExampleCodepen,
                view: React.createElement(Nav_CustomGroupHeaders_Example_1.NavCustomGroupHeadersExample, null)
            }
        ],
        propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Nav/Nav.types.ts')],
        overview: require('!raw-loader!office-ui-fabric-react/src/components/Nav/docs/NavOverview.md'),
        bestPractices: '',
        dos: require('!raw-loader!office-ui-fabric-react/src/components/Nav/docs/NavDos.md'),
        donts: require('!raw-loader!office-ui-fabric-react/src/components/Nav/docs/NavDonts.md'),
        isHeaderVisible: true,
        isFeedbackVisible: true
    };
});
