// office-ui-fabric-react/lib/components/Modal/Modal.doc.js
import * as React from 'react';
import { ModalBasicExample } from './examples/Modal.Basic.Example';
import { ModalModelessExample } from './examples/Modal.Modeless.Example';
import { ModalStatus } from './Modal.checklist';
var ModalBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Modal/examples/Modal.Basic.Example.tsx');
var ModalBasicExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Modal/Modal.Basic.Example.Codepen.txt');
var ModalModelessExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Modal/examples/Modal.Modeless.Example.tsx');
export var ModalPageProps = {
    title: 'Modal',
    componentName: 'Modal',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Modal',
    componentStatus: ModalStatus,
    examples: [
        {
            title: 'Modal',
            code: ModalBasicExampleCode,
            view: React.createElement(ModalBasicExample, null),
            codepenJS: ModalBasicExampleCodepen
        },
        {
            title: 'Modeless Modal',
            code: ModalModelessExampleCode,
            view: React.createElement(ModalModelessExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Modal/Modal.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Modal/docs/ModalOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Modal/docs/ModalDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Modal/docs/ModalDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
