// office-ui-fabric-react/lib/components/Stack/Stack.doc.js
import * as React from 'react';
import { StackStatus } from './Stack.checklist';
// Vertical Stack Examples
import { VerticalStackBasicExample } from './examples/Stack.Vertical.Basic.Example';
var VerticalStackBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Vertical.Basic.Example.tsx');
var VerticalStackBasicExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Vertical.Basic.Example.Codepen.txt');
import { VerticalStackReversedExample } from './examples/Stack.Vertical.Reversed.Example';
var VerticalStackReversedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Vertical.Reversed.Example.tsx');
var VerticalStackReversedExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Vertical.Reversed.Example.Codepen.txt');
import { VerticalStackSpacingExample } from './examples/Stack.Vertical.Spacing.Example';
var VerticalStackSpacingExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Vertical.Spacing.Example.tsx');
var VerticalStackSpacingExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Vertical.Spacing.Example.Codepen.txt');
import { VerticalStackGrowExample } from './examples/Stack.Vertical.Grow.Example';
var VerticalStackGrowExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Vertical.Grow.Example.tsx');
var VerticalStackGrowExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Vertical.Grow.Example.Codepen.txt');
import { VerticalStackShrinkExample } from './examples/Stack.Vertical.Shrink.Example';
var VerticalStackShrinkExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Vertical.Shrink.Example.tsx');
var VerticalStackShrinkExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Vertical.Shrink.Example.Codepen.txt');
import { VerticalStackWrapExample } from './examples/Stack.Vertical.Wrap.Example';
var VerticalStackWrapExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Vertical.Wrap.Example.tsx');
var VerticalStackWrapExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Vertical.Wrap.Example.Codepen.txt');
import { VerticalStackWrapAdvancedExample } from './examples/Stack.Vertical.WrapAdvanced.Example';
var VerticalStackWrapAdvancedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Vertical.WrapAdvanced.Example.tsx');
var VerticalStackWrapAdvancedExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Vertical.WrapAdvanced.Example.Codepen.txt');
import { VerticalStackWrapNestedExample } from './examples/Stack.Vertical.WrapNested.Example';
var VerticalStackWrapNestedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Vertical.WrapNested.Example.tsx');
var VerticalStackWrapNestedExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Vertical.WrapNested.Example.Codepen.txt');
import { VerticalStackVerticalAlignExample } from './examples/Stack.Vertical.VerticalAlign.Example';
var VerticalStackVerticalAlignExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Vertical.VerticalAlign.Example.tsx');
var VerticalStackVerticalAlignExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Vertical.VerticalAlign.Example.Codepen.txt');
import { VerticalStackHorizontalAlignExample } from './examples/Stack.Vertical.HorizontalAlign.Example';
var VerticalStackHorizontalAlignExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Vertical.HorizontalAlign.Example.tsx');
var VerticalStackHorizontalAlignExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Vertical.HorizontalAlign.Example.Codepen.txt');
import { VerticalStackConfigureExample } from './examples/Stack.Vertical.Configure.Example';
var VerticalStackConfigureExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Vertical.Configure.Example.tsx');
var VerticalStackConfigureExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Vertical.Configure.Example.Codepen.txt');
// Horizontal Stack Examples
import { HorizontalStackBasicExample } from './examples/Stack.Horizontal.Basic.Example';
var HorizontalStackBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Horizontal.Basic.Example.tsx');
var HorizontalStackBasicExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Horizontal.Basic.Example.Codepen.txt');
import { HorizontalStackReversedExample } from './examples/Stack.Horizontal.Reversed.Example';
var HorizontalStackReversedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Horizontal.Reversed.Example.tsx');
var HorizontalStackReversedExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Horizontal.Reversed.Example.Codepen.txt');
import { HorizontalStackSpacingExample } from './examples/Stack.Horizontal.Spacing.Example';
var HorizontalStackSpacingExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Horizontal.Spacing.Example.tsx');
var HorizontalStackSpacingExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Horizontal.Spacing.Example.Codepen.txt');
import { HorizontalStackGrowExample } from './examples/Stack.Horizontal.Grow.Example';
var HorizontalStackGrowExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Horizontal.Grow.Example.tsx');
var HorizontalStackGrowExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Horizontal.Grow.Example.Codepen.txt');
import { HorizontalStackShrinkExample } from './examples/Stack.Horizontal.Shrink.Example';
var HorizontalStackShrinkExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Horizontal.Shrink.Example.tsx');
var HorizontalStackShrinkExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Horizontal.Shrink.Example.Codepen.txt');
import { HorizontalStackWrapExample } from './examples/Stack.Horizontal.Wrap.Example';
var HorizontalStackWrapExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Horizontal.Wrap.Example.tsx');
var HorizontalStackWrapExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Horizontal.Wrap.Example.Codepen.txt');
import { HorizontalStackWrapAdvancedExample } from './examples/Stack.Horizontal.WrapAdvanced.Example';
var HorizontalStackWrapAdvancedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Horizontal.WrapAdvanced.Example.tsx');
var HorizontalStackWrapAdvancedExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Horizontal.WrapAdvanced.Example.Codepen.txt');
import { HorizontalStackWrapNestedExample } from './examples/Stack.Horizontal.WrapNested.Example';
var HorizontalStackWrapNestedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Horizontal.WrapNested.Example.tsx');
var HorizontalStackWrapNestedExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Horizontal.WrapNested.Example.Codepen.txt');
import { HorizontalStackHorizontalAlignExample } from './examples/Stack.Horizontal.HorizontalAlign.Example';
var HorizontalStackHorizontalAlignExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Horizontal.HorizontalAlign.Example.tsx');
var HorizontalStackHorizontalAlignExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Horizontal.HorizontalAlign.Example.Codepen.txt');
import { HorizontalStackVerticalAlignExample } from './examples/Stack.Horizontal.VerticalAlign.Example';
var HorizontalStackVerticalAlignExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Horizontal.VerticalAlign.Example.tsx');
var HorizontalStackVerticalAlignExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Horizontal.VerticalAlign.Example.Codepen.txt');
import { HorizontalStackConfigureExample } from './examples/Stack.Horizontal.Configure.Example';
var HorizontalStackConfigureExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Stack/examples/Stack.Horizontal.Configure.Example.tsx');
var HorizontalStackConfigureExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Stack/Stack.Horizontal.Configure.Example.Codepen.txt');
export var StackPageProps = {
    title: 'Stack',
    componentName: 'Stack',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Stack',
    componentStatus: StackStatus,
    examples: [
        {
            title: 'Basic Vertical Stack',
            code: VerticalStackBasicExampleCode,
            view: React.createElement(VerticalStackBasicExample, null),
            codepenJS: VerticalStackBasicExampleCodepen
        },
        {
            title: 'Reversed Basic Vertical Stack',
            code: VerticalStackReversedExampleCode,
            view: React.createElement(VerticalStackReversedExample, null),
            codepenJS: VerticalStackReversedExampleCodepen
        },
        {
            title: 'Vertical Stack - Gap and Padding Sizes',
            code: VerticalStackSpacingExampleCode,
            view: React.createElement(VerticalStackSpacingExample, null),
            codepenJS: VerticalStackSpacingExampleCodepen
        },
        {
            title: 'Vertical Stack - Growing Items',
            code: VerticalStackGrowExampleCode,
            view: React.createElement(VerticalStackGrowExample, null),
            codepenJS: VerticalStackGrowExampleCodepen
        },
        {
            title: 'Vertical Stack - Shrinking Items',
            code: VerticalStackShrinkExampleCode,
            view: React.createElement(VerticalStackShrinkExample, null),
            codepenJS: VerticalStackShrinkExampleCodepen
        },
        {
            title: 'Vertical Stack - Wrapping - Basic',
            code: VerticalStackWrapExampleCode,
            view: React.createElement(VerticalStackWrapExample, null),
            codepenJS: VerticalStackWrapExampleCodepen
        },
        {
            title: 'Vertical Stack - Wrapping - Advanced',
            code: VerticalStackWrapAdvancedExampleCode,
            view: React.createElement(VerticalStackWrapAdvancedExample, null),
            codepenJS: VerticalStackWrapAdvancedExampleCodepen
        },
        {
            title: 'Vertical Stack - Wrapping - Nested',
            code: VerticalStackWrapNestedExampleCode,
            view: React.createElement(VerticalStackWrapNestedExample, null),
            codepenJS: VerticalStackWrapNestedExampleCodepen
        },
        {
            title: 'Vertical Stack - Vertical Alignments',
            code: VerticalStackVerticalAlignExampleCode,
            view: React.createElement(VerticalStackVerticalAlignExample, null),
            codepenJS: VerticalStackVerticalAlignExampleCodepen
        },
        {
            title: 'Vertical Stack - Horizontal Alignments',
            code: VerticalStackHorizontalAlignExampleCode,
            view: React.createElement(VerticalStackHorizontalAlignExample, null),
            codepenJS: VerticalStackHorizontalAlignExampleCodepen
        },
        {
            title: 'Vertical Stack - Configure Properties',
            code: VerticalStackConfigureExampleCode,
            view: React.createElement(VerticalStackConfigureExample, null),
            codepenJS: VerticalStackConfigureExampleCodepen
        },
        {
            title: 'Basic Horizontal Stack',
            code: HorizontalStackBasicExampleCode,
            view: React.createElement(HorizontalStackBasicExample, null),
            codepenJS: HorizontalStackBasicExampleCodepen
        },
        {
            title: 'Reversed Basic Horizontal Stack',
            code: HorizontalStackReversedExampleCode,
            view: React.createElement(HorizontalStackReversedExample, null),
            codepenJS: HorizontalStackReversedExampleCodepen
        },
        {
            title: 'Horizontal Stack - Gap and Padding Sizes',
            code: HorizontalStackSpacingExampleCode,
            view: React.createElement(HorizontalStackSpacingExample, null),
            codepenJS: HorizontalStackSpacingExampleCodepen
        },
        {
            title: 'Horizontal Stack - Growing Items',
            code: HorizontalStackGrowExampleCode,
            view: React.createElement(HorizontalStackGrowExample, null),
            codepenJS: HorizontalStackGrowExampleCodepen
        },
        {
            title: 'Horizontal Stack - Shrinking Items',
            code: HorizontalStackShrinkExampleCode,
            view: React.createElement(HorizontalStackShrinkExample, null),
            codepenJS: HorizontalStackShrinkExampleCodepen
        },
        {
            title: 'Horizontal Stack - Wrapping - Basic',
            code: HorizontalStackWrapExampleCode,
            view: React.createElement(HorizontalStackWrapExample, null),
            codepenJS: HorizontalStackWrapExampleCodepen
        },
        {
            title: 'Horizontal Stack - Wrapping - Advanced',
            code: HorizontalStackWrapAdvancedExampleCode,
            view: React.createElement(HorizontalStackWrapAdvancedExample, null),
            codepenJS: HorizontalStackWrapAdvancedExampleCodepen
        },
        {
            title: 'Horizontal Stack - Wrapping - Nested',
            code: HorizontalStackWrapNestedExampleCode,
            view: React.createElement(HorizontalStackWrapNestedExample, null),
            codepenJS: HorizontalStackWrapNestedExampleCodepen
        },
        {
            title: 'Horizontal Stack - Horizontal Alignments',
            code: HorizontalStackHorizontalAlignExampleCode,
            view: React.createElement(HorizontalStackHorizontalAlignExample, null),
            codepenJS: HorizontalStackHorizontalAlignExampleCodepen
        },
        {
            title: 'Horizontal Stack - Vertical Alignments',
            code: HorizontalStackVerticalAlignExampleCode,
            view: React.createElement(HorizontalStackVerticalAlignExample, null),
            codepenJS: HorizontalStackVerticalAlignExampleCodepen
        },
        {
            title: 'Horizontal Stack - Configure Properties',
            code: HorizontalStackConfigureExampleCode,
            view: React.createElement(HorizontalStackConfigureExample, null),
            codepenJS: HorizontalStackConfigureExampleCodepen
        }
    ],
    propertiesTablesSources: [
        require('!raw-loader!office-ui-fabric-react/src/components/Stack/Stack.types.ts'),
        require('!raw-loader!office-ui-fabric-react/src/components/Stack/StackItem/StackItem.types.ts')
    ],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Stack/docs/StackOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Stack/docs/StackDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Stack/docs/StackDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true,
    allowNativeProps: true
};
