// office-ui-fabric-react/lib-amd/components/Checkbox/Checkbox.doc.js
define(["require", "exports", "react", "./examples/Checkbox.Basic.Example", "./examples/Checkbox.ImplementationExamples", "./Checkbox.checklist"], function (require, exports, React, Checkbox_Basic_Example_1, Checkbox_ImplementationExamples_1, Checkbox_checklist_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var CheckboxBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Checkbox/examples/Checkbox.Basic.Example.tsx');
    var CheckboxBasicExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Checkbox/Checkbox.Basic.Example.Codepen.txt');
    var CheckboxImplementationExamplesCode = require('!raw-loader!office-ui-fabric-react/src/components/Checkbox/examples/Checkbox.ImplementationExamples.tsx');
    exports.CheckboxPageProps = {
        title: 'Checkbox',
        componentName: 'Checkbox',
        componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Checkbox',
        componentStatus: Checkbox_checklist_1.CheckboxStatus,
        examples: [
            {
                title: 'Default Checkbox',
                code: CheckboxBasicExampleCode,
                view: React.createElement(Checkbox_Basic_Example_1.CheckboxBasicExample, null),
                codepenJS: CheckboxBasicExampleCodepen
            },
            {
                title: 'Implementation Examples',
                code: CheckboxImplementationExamplesCode,
                view: React.createElement(Checkbox_ImplementationExamples_1.CheckboxImplementationExamples, null)
            }
        ],
        propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Checkbox/Checkbox.types.ts')],
        overview: require('!raw-loader!office-ui-fabric-react/src/components/Checkbox/docs/CheckboxOverview.md'),
        bestPractices: '',
        dos: require('!raw-loader!office-ui-fabric-react/src/components/Checkbox/docs/CheckboxDos.md'),
        donts: require('!raw-loader!office-ui-fabric-react/src/components/Checkbox/docs/CheckboxDonts.md'),
        isHeaderVisible: true,
        isFeedbackVisible: true
    };
});
