// office-ui-fabric-react/lib-amd/components/Coachmark/PositioningContainer/PositioningContainer.styles.js
define(["require", "exports", "tslib", "../../../Utilities", "../../../Styling"], function (require, exports, tslib_1, Utilities_1, Styling_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    // @TODO Remove this tslint disable statement after the styles are converted
    // to the updated mergeStyles method.
    /* tslint:disable */
    function highContrastActive(styles) {
        return {
            '@media screen and (-ms-high-contrast: active)': styles
        };
    }
    exports.highContrastActive = highContrastActive;
    function focusClear() {
        return {
            '&::-moz-focus-inner': {
                border: 0
            },
            '&': {
                outline: 'transparent'
            }
        };
    }
    exports.focusClear = focusClear;
    /* tslint:enable */
    exports.getClassNames = Utilities_1.memoizeFunction(function () {
        return Styling_1.mergeStyleSets({
            root: {
                position: 'absolute',
                boxSizing: 'border-box',
                border: '1px solid ${}',
                selectors: tslib_1.__assign({}, highContrastActive({
                    border: '1px solid WindowText'
                }), focusClear())
            },
            container: {
                position: 'relative'
            },
            main: {
                backgroundColor: '#ffffff',
                overflowX: 'hidden',
                overflowY: 'hidden',
                position: 'relative'
            },
            overFlowYHidden: {
                overflowY: 'hidden'
            }
        });
    });
});
