// office-ui-fabric-react/lib-commonjs/components/Facepile/Facepile.doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var Facepile_AddFace_Example_1 = require("./examples/Facepile.AddFace.Example");
var Facepile_Basic_Example_1 = require("./examples/Facepile.Basic.Example");
var Facepile_Overflow_Example_1 = require("./examples/Facepile.Overflow.Example");
var Facepile_checklist_1 = require("./Facepile.checklist");
var FacepileAddFaceExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Facepile/examples/Facepile.AddFace.Example.tsx');
var FacepileBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Facepile/examples/Facepile.Basic.Example.tsx');
var FacepileOverflowExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Facepile/examples/Facepile.Overflow.Example.tsx');
exports.FacepilePageProps = {
    title: 'Facepile',
    componentName: 'Facepile',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Facepile',
    componentStatus: Facepile_checklist_1.FacepileStatus,
    examples: [
        {
            title: 'Facepile with size and fade in options',
            code: FacepileBasicExampleCode,
            view: React.createElement(Facepile_Basic_Example_1.FacepileBasicExample, null)
        },
        {
            title: 'Facepile with overflow buttons',
            code: FacepileOverflowExampleCode,
            view: React.createElement(Facepile_Overflow_Example_1.FacepileOverflowExample, null)
        },
        {
            title: 'Facepile with face adding functionality',
            code: FacepileAddFaceExampleCode,
            view: React.createElement(Facepile_AddFace_Example_1.FacepileAddFaceExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Facepile/Facepile.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Facepile/docs/FacepileOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Facepile/docs/FacepileDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Facepile/docs/FacepileDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
