// office-ui-fabric-react/lib-commonjs/components/ScrollablePane/ScrollablePane.doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var ScrollablePane_Default_Example_1 = require("./examples/ScrollablePane.Default.Example");
var ScrollablePane_DetailsList_Example_1 = require("./examples/ScrollablePane.DetailsList.Example");
var ScrollablePane_checklist_1 = require("./ScrollablePane.checklist");
var ScrollablePaneDefaultExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ScrollablePane/examples/ScrollablePane.Default.Example.tsx');
var ScrollablePaneDetailsListExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ScrollablePane/examples/ScrollablePane.DetailsList.Example.tsx');
var ScrollablePaneDefaultExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/ScrollablePane/ScrollablePane.Default.Example.Codepen.txt');
var ScrollablePaneDetailsListExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/ScrollablePane/ScrollablePane.DetailsList.Example.Codepen.txt');
exports.ScrollablePanePageProps = {
    title: 'ScrollablePane',
    componentName: 'ScrollablePane',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/ScrollablePane',
    componentStatus: ScrollablePane_checklist_1.ScrollablePaneStatus,
    examples: [
        {
            title: 'Default',
            code: ScrollablePaneDefaultExampleCode,
            codepenJS: ScrollablePaneDefaultExampleCodepen,
            view: React.createElement(ScrollablePane_Default_Example_1.ScrollablePaneDefaultExample, null),
            isScrollable: false
        },
        {
            title: 'DetailsList Locked Header',
            code: ScrollablePaneDetailsListExampleCode,
            codepenJS: ScrollablePaneDetailsListExampleCodepen,
            view: React.createElement(ScrollablePane_DetailsList_Example_1.ScrollablePaneDetailsListExample, null),
            isScrollable: false
        }
    ],
    propertiesTablesSources: [
        require('!raw-loader!office-ui-fabric-react/src/components/ScrollablePane/ScrollablePane.types.ts'),
        require('!raw-loader!office-ui-fabric-react/src/components/Sticky/Sticky.types.ts')
    ],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/ScrollablePane/docs/ScrollablePaneOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/ScrollablePane/docs/ScrollablePaneDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/ScrollablePane/docs/ScrollablePaneDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true,
    allowNativeProps: true,
    nativePropsElement: ['a', 'button']
};
