// office-ui-fabric-react/lib-commonjs/components/Tooltip/Tooltip.doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var Tooltip_Custom_Example_1 = require("./examples/Tooltip.Custom.Example");
var Tooltip_Basic_Example_1 = require("./examples/Tooltip.Basic.Example");
var Tooltip_Display_Example_1 = require("./examples/Tooltip.Display.Example");
var Tooltip_Interactive_Example_1 = require("./examples/Tooltip.Interactive.Example");
var Tooltip_Overflow_Example_1 = require("./examples/Tooltip.Overflow.Example");
var Tooltip_NoScroll_Example_1 = require("./examples/Tooltip.NoScroll.Example");
var Tooltip_Absolute_Position_Example_1 = require("./examples/Tooltip.Absolute.Position.Example");
var Tooltip_checklist_1 = require("./Tooltip.checklist");
var TooltipBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Tooltip/examples/Tooltip.Basic.Example.tsx');
var TooltipDisplayExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Tooltip/examples/Tooltip.Display.Example.tsx');
var TooltipBasicExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Tooltip/examples/Tooltip.Basic.Example.tsx');
var TooltipCustomExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Tooltip/examples/Tooltip.Custom.Example.tsx');
var TooltipInteractiveExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Tooltip/examples/Tooltip.Interactive.Example.tsx');
var TooltipOverflowExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Tooltip/examples/Tooltip.Overflow.Example.tsx');
var TooltipNoScrollExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Tooltip/examples/Tooltip.NoScroll.Example.tsx');
var TooltipAbsolutePositionExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Tooltip/examples/Tooltip.Absolute.Position.Example.tsx');
exports.TooltipPageProps = {
    title: 'Tooltip',
    componentName: 'Tooltip',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Tooltip',
    componentStatus: Tooltip_checklist_1.TooltipStatus,
    examples: [
        {
            title: 'Default Tooltip',
            code: TooltipBasicExampleCode,
            view: React.createElement(Tooltip_Basic_Example_1.TooltipBasicExample, null),
            codepenJS: TooltipBasicExampleCodepen
        },
        {
            title: 'Tooltip wrapping inline or inline-block elements',
            code: TooltipDisplayExampleCode,
            view: React.createElement(Tooltip_Display_Example_1.TooltipDisplayExample, null)
        },
        {
            title: 'Tooltip with list',
            code: TooltipCustomExampleCode,
            view: React.createElement(Tooltip_Custom_Example_1.TooltipCustomExample, null)
        },
        {
            title: 'Tooltip with a closing delay',
            code: TooltipInteractiveExampleCode,
            view: React.createElement(Tooltip_Interactive_Example_1.TooltipInteractiveExample, null)
        },
        {
            title: 'Tooltip only on overflow',
            code: TooltipOverflowExampleCode,
            view: React.createElement(Tooltip_Overflow_Example_1.TooltipOverflowExample, null)
        },
        {
            title: 'Tooltip without scroll (improves performance)',
            code: TooltipNoScrollExampleCode,
            view: React.createElement(Tooltip_NoScroll_Example_1.TooltipNoScrollExample, null)
        },
        {
            title: 'Tooltip with position absolute',
            code: TooltipAbsolutePositionExampleCode,
            view: React.createElement(Tooltip_Absolute_Position_Example_1.TooltipAbsolutePositionExample, null)
        }
    ],
    propertiesTablesSources: [
        require('!raw-loader!office-ui-fabric-react/src/components/Tooltip/Tooltip.types.ts'),
        require('!raw-loader!office-ui-fabric-react/src/components/Tooltip/TooltipHost.types.ts')
    ],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Tooltip/docs/TooltipOverview.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true,
    allowNativeProps: true
};
