// office-ui-fabric-react/lib/components/FocusZone/examples/FocusZone.Photos.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { FocusZone } from 'office-ui-fabric-react/lib/FocusZone';
import { Image } from 'office-ui-fabric-react/lib/Image';
import { getId } from 'office-ui-fabric-react/lib/Utilities';
import { mergeStyleSets, getTheme } from 'office-ui-fabric-react/lib/Styling';
var theme = getTheme();
var classNames = mergeStyleSets({
    photoList: {
        display: 'inline-block',
        border: '1px solid ' + theme.palette.neutralTertiary,
        padding: 10,
        lineHeight: 0,
        overflow: 'hidden'
    },
    photoCell: {
        position: 'relative',
        display: 'inline-block',
        padding: 2,
        boxSizing: 'border-box',
        selectors: {
            '&:focus': {
                outline: 'none',
                selectors: {
                    '&:after': {
                        content: '',
                        position: 'absolute',
                        right: 4,
                        left: 4,
                        top: 4,
                        bottom: 4,
                        border: '1px solid ' + theme.palette.white,
                        outline: '2px solid ' + theme.palette.themePrimary
                    }
                }
            }
        }
    }
});
var MAX_COUNT = 20;
var FocusZonePhotosExample = /** @class */ (function (_super) {
    tslib_1.__extends(FocusZonePhotosExample, _super);
    function FocusZonePhotosExample(props) {
        var _this = _super.call(this, props) || this;
        _this._items = _this._getItems();
        return _this;
    }
    FocusZonePhotosExample.prototype.render = function () {
        var _this = this;
        return (React.createElement(FocusZone, { as: "ul", className: classNames.photoList }, this._items.map(function (item, index) { return (React.createElement("li", { key: item.id, className: classNames.photoCell, "aria-posinset": index + 1, "aria-setsize": _this._items.length, "aria-label": "Photo", "data-is-focusable": true },
            React.createElement(Image, { src: item.url, width: item.width, height: item.height }))); })));
    };
    FocusZonePhotosExample.prototype._getItems = function () {
        var items = [];
        for (var i = 0; i < MAX_COUNT; i++) {
            var randomWidth = 50 + Math.floor(Math.random() * 150);
            items.push({
                id: getId('photo'),
                url: "http://placehold.it/" + randomWidth + "x100",
                width: randomWidth,
                height: 100
            });
        }
        return items;
    };
    return FocusZonePhotosExample;
}(React.PureComponent));
export { FocusZonePhotosExample };
