// office-ui-fabric-react/lib-commonjs/components/Nav/examples/Nav.ByKeys.Example.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var Nav_1 = require("office-ui-fabric-react/lib/Nav");
var NavByKeysExample = /** @class */ (function (_super) {
    tslib_1.__extends(NavByKeysExample, _super);
    function NavByKeysExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NavByKeysExample.prototype.render = function () {
        return (React.createElement(Nav_1.Nav, { groups: [
                {
                    links: [
                        { name: 'Home', key: 'Home', url: '' },
                        { name: 'Activity', key: 'Activity', url: '' },
                        { name: 'News', key: 'News', url: '' },
                        { name: 'Documents', key: 'Documents', url: '' }
                    ]
                }
            ] }));
    };
    return NavByKeysExample;
}(React.Component));
exports.NavByKeysExample = NavByKeysExample;
