// office-ui-fabric-react/lib-amd/components/Panel/Panel.doc.js
define(["require", "exports", "react", "./examples/Panel.Custom.Example", "./examples/Panel.CustomLeft.Example", "./examples/Panel.ExtraLarge.Example", "./examples/Panel.Footer.Example", "./examples/Panel.Navigation.Example", "./examples/Panel.HandleDismissTarget.Example", "./examples/Panel.HiddenOnDismiss.Example", "./examples/Panel.Large.Example", "./examples/Panel.LargeFixed.Example", "./examples/Panel.LightDismiss.Example", "./examples/Panel.LightDismissCustom.Example", "./examples/Panel.Medium.Example", "./examples/Panel.NonModal.Example", "./examples/Panel.PreventDefault.Example", "./examples/Panel.Scroll.Example", "./examples/Panel.SmallFluid.Example", "./examples/Panel.SmallLeft.Example", "./examples/Panel.SmallRight.Example", "./Panel.checklist"], function (require, exports, React, Panel_Custom_Example_1, Panel_CustomLeft_Example_1, Panel_ExtraLarge_Example_1, Panel_Footer_Example_1, Panel_Navigation_Example_1, Panel_HandleDismissTarget_Example_1, Panel_HiddenOnDismiss_Example_1, Panel_Large_Example_1, Panel_LargeFixed_Example_1, Panel_LightDismiss_Example_1, Panel_LightDismissCustom_Example_1, Panel_Medium_Example_1, Panel_NonModal_Example_1, Panel_PreventDefault_Example_1, Panel_Scroll_Example_1, Panel_SmallFluid_Example_1, Panel_SmallLeft_Example_1, Panel_SmallRight_Example_1, Panel_checklist_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var PanelSmallRightExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.SmallRight.Example.tsx');
    var PanelSmallRightExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Panel/Panel.SmallRight.Example.Codepen.txt');
    var PanelSmallLeftExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.SmallLeft.Example.tsx');
    var PanelSmallFluidExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.SmallFluid.Example.tsx');
    var PanelMediumExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Medium.Example.tsx');
    var PanelLargeExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Large.Example.tsx');
    var PanelLargeFixedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.LargeFixed.Example.tsx');
    var PanelExtraLargeExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.ExtraLarge.Example.tsx');
    var PanelCustomExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Custom.Example.tsx');
    var PanelCustomLeftExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.CustomLeft.Example.tsx');
    var PanelHiddenOnDismissExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.HiddenOnDismiss.Example.tsx');
    var PanelLightDismissExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.LightDismiss.Example.tsx');
    var PanelLightDismissCustomExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.LightDismissCustom.Example.tsx');
    var PanelNonModalExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.NonModal.Example.tsx');
    var PanelFooterExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Footer.Example.tsx');
    var PanelNavigationExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Navigation.Example.tsx');
    var PanelPreventDefaultExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.PreventDefault.Example.tsx');
    var PanelHandleDismissTargetExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.HandleDismissTarget.Example.tsx');
    var PanelScrollExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Panel/examples/Panel.Scroll.Example.tsx');
    exports.PanelPageProps = {
        title: 'Panel',
        componentName: 'Panel',
        componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Panel',
        componentStatus: Panel_checklist_1.PanelStatus,
        examples: [
            {
                title: 'Panel - Small Panel, Anchored Right, Fixed Width',
                code: PanelSmallRightExampleCode,
                view: React.createElement(Panel_SmallRight_Example_1.PanelSmallRightExample, null),
                codepenJS: PanelSmallRightExampleCodepen
            },
            {
                title: 'Panel - Small Panel, Anchored Left, Fixed Width',
                code: PanelSmallLeftExampleCode,
                view: React.createElement(Panel_SmallLeft_Example_1.PanelSmallLeftExample, null)
            },
            {
                title: 'Panel - Small Panel, Full Screen, Fluid Width',
                code: PanelSmallFluidExampleCode,
                view: React.createElement(Panel_SmallFluid_Example_1.PanelSmallFluidExample, null)
            },
            {
                title: 'Panel - Medium',
                code: PanelMediumExampleCode,
                view: React.createElement(Panel_Medium_Example_1.PanelMediumExample, null)
            },
            {
                title: 'Panel - Large',
                code: PanelLargeExampleCode,
                view: React.createElement(Panel_Large_Example_1.PanelLargeExample, null)
            },
            {
                title: 'Panel - LargeFixed',
                code: PanelLargeFixedExampleCode,
                view: React.createElement(Panel_LargeFixed_Example_1.PanelLargeFixedExample, null)
            },
            {
                title: 'Panel - Extra Large',
                code: PanelExtraLargeExampleCode,
                view: React.createElement(Panel_ExtraLarge_Example_1.PanelExtraLargeExample, null)
            },
            {
                title: 'Panel - Custom',
                code: PanelCustomExampleCode,
                view: React.createElement(Panel_Custom_Example_1.PanelCustomExample, null)
            },
            {
                title: 'Panel - Custom Width, Anchored Left',
                code: PanelCustomLeftExampleCode,
                view: React.createElement(Panel_CustomLeft_Example_1.PanelCustomLeftExample, null)
            },
            {
                title: 'Panel - Hidden on Dismiss',
                code: PanelHiddenOnDismissExampleCode,
                view: React.createElement(Panel_HiddenOnDismiss_Example_1.PanelHiddenOnDismissExample, null)
            },
            {
                title: 'Panel - Light Dismiss',
                code: PanelLightDismissExampleCode,
                view: React.createElement(Panel_LightDismiss_Example_1.PanelLightDismissExample, null)
            },
            {
                title: 'Panel - Custom Light Dismiss',
                code: PanelLightDismissCustomExampleCode,
                view: React.createElement(Panel_LightDismissCustom_Example_1.PanelLightDismissCustomExample, null)
            },
            {
                title: 'Panel - Non-Modal',
                code: PanelNonModalExampleCode,
                view: React.createElement(Panel_NonModal_Example_1.PanelNonModalExample, null)
            },
            {
                title: 'Panel - Footer',
                code: PanelFooterExampleCode,
                view: React.createElement(Panel_Footer_Example_1.PanelFooterExample, null)
            },
            {
                title: 'Panel - Navigation',
                code: PanelNavigationExampleCode,
                view: React.createElement(Panel_Navigation_Example_1.PanelNavigationExample, null)
            },
            {
                title: 'Panel - Prevent Default Sample',
                code: PanelPreventDefaultExampleCode,
                view: React.createElement(Panel_PreventDefault_Example_1.PanelPreventDefaultExample, null)
            },
            {
                title: 'Panel - Handle Dismiss Target Sample',
                code: PanelHandleDismissTargetExampleCode,
                view: React.createElement(Panel_HandleDismissTarget_Example_1.PanelHandleDismissTargetExample, null)
            },
            {
                title: 'Panel - Scrolling Content Sample',
                code: PanelScrollExampleCode,
                view: React.createElement(Panel_Scroll_Example_1.PanelScrollExample, null)
            }
        ],
        propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Panel/Panel.types.ts')],
        overview: require('!raw-loader!office-ui-fabric-react/src/components/Panel/docs/PanelOverview.md'),
        bestPractices: '',
        dos: require('!raw-loader!office-ui-fabric-react/src/components/Panel/docs/PanelDos.md'),
        donts: require('!raw-loader!office-ui-fabric-react/src/components/Panel/docs/PanelDonts.md'),
        isHeaderVisible: true,
        isFeedbackVisible: true,
        allowNativeProps: true
    };
});
