// office-ui-fabric-react/lib-amd/components/DocumentCard/DocumentCardStatus.base.js
define(["require", "exports", "tslib", "react", "../../Utilities", "../../Icon"], function (require, exports, tslib_1, React, Utilities_1, Icon_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var getClassNames = Utilities_1.classNamesFunction();
    var DocumentCardStatusBase = /** @class */ (function (_super) {
        tslib_1.__extends(DocumentCardStatusBase, _super);
        function DocumentCardStatusBase() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        DocumentCardStatusBase.prototype.render = function () {
            var _a = this.props, statusIcon = _a.statusIcon, status = _a.status, styles = _a.styles, theme = _a.theme, className = _a.className;
            var iconProps = {
                iconName: statusIcon,
                styles: {
                    root: { padding: '8px' }
                }
            };
            this._classNames = getClassNames(styles, {
                theme: theme,
                className: className
            });
            return (React.createElement("div", { className: this._classNames.root },
                statusIcon && React.createElement(Icon_1.Icon, tslib_1.__assign({}, iconProps)),
                status));
        };
        return DocumentCardStatusBase;
    }(Utilities_1.BaseComponent));
    exports.DocumentCardStatusBase = DocumentCardStatusBase;
});
