// office-ui-fabric-react/lib/components/Stack/examples/Stack.Horizontal.HorizontalAlign.Example.js
import * as tslib_1 from "tslib";
// @codepen
import * as React from 'react';
import { Stack } from '../Stack';
import { mergeStyleSets, DefaultPalette } from 'office-ui-fabric-react/lib/Styling';
var HorizontalStackHorizontalAlignExample = /** @class */ (function (_super) {
    tslib_1.__extends(HorizontalStackHorizontalAlignExample, _super);
    function HorizontalStackHorizontalAlignExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HorizontalStackHorizontalAlignExample.prototype.render = function () {
        var styles = mergeStyleSets({
            root: {
                background: DefaultPalette.themeTertiary,
                selectors: {
                    '> *': {
                        width: 50,
                        height: 50,
                        display: 'flex',
                        alignItems: 'center',
                        justifyContent: 'center',
                        background: DefaultPalette.themePrimary,
                        color: DefaultPalette.white
                    }
                }
            }
        });
        return (React.createElement(Stack, { gap: 5 },
            React.createElement("span", null, "Left-aligned"),
            React.createElement(Stack, { horizontal: true, horizontalAlign: "start", className: styles.root },
                React.createElement("span", null, "1"),
                React.createElement("span", null, "2"),
                React.createElement("span", null, "3")),
            React.createElement("span", null, "Horizontally centered"),
            React.createElement(Stack, { horizontal: true, horizontalAlign: "center", className: styles.root },
                React.createElement("span", null, "1"),
                React.createElement("span", null, "2"),
                React.createElement("span", null, "3")),
            React.createElement("span", null, "Right-aligned"),
            React.createElement(Stack, { horizontal: true, horizontalAlign: "end", className: styles.root },
                React.createElement("span", null, "1"),
                React.createElement("span", null, "2"),
                React.createElement("span", null, "3")),
            React.createElement("span", null, "Horizontal space around items"),
            React.createElement(Stack, { horizontal: true, horizontalAlign: "space-around", className: styles.root },
                React.createElement("span", null, "1"),
                React.createElement("span", null, "2"),
                React.createElement("span", null, "3")),
            React.createElement("span", null, "Horizontal space between items"),
            React.createElement(Stack, { horizontal: true, horizontalAlign: "space-between", className: styles.root },
                React.createElement("span", null, "1"),
                React.createElement("span", null, "2"),
                React.createElement("span", null, "3")),
            React.createElement("span", null, "Items horizontally evenly spaced"),
            React.createElement(Stack, { horizontal: true, horizontalAlign: "space-evenly", className: styles.root },
                React.createElement("span", null, "1"),
                React.createElement("span", null, "2"),
                React.createElement("span", null, "3"))));
    };
    return HorizontalStackHorizontalAlignExample;
}(React.Component));
export { HorizontalStackHorizontalAlignExample };
