// office-ui-fabric-react/lib/components/Stack/examples/Stack.Horizontal.WrapNested.Example.js
import * as tslib_1 from "tslib";
// @codepen
import * as React from 'react';
import { Slider } from 'office-ui-fabric-react/lib/Slider';
import { Stack } from '../Stack';
import { mergeStyleSets, DefaultPalette } from 'office-ui-fabric-react/lib/Styling';
var HorizontalStackWrapNestedExample = /** @class */ (function (_super) {
    tslib_1.__extends(HorizontalStackWrapNestedExample, _super);
    function HorizontalStackWrapNestedExample(props) {
        var _this = _super.call(this, props) || this;
        _this._onWidthChange = function (value) {
            _this.setState({ stackWidth: value });
        };
        _this.state = {
            stackWidth: 100
        };
        return _this;
    }
    HorizontalStackWrapNestedExample.prototype.render = function () {
        var textStyles = {
            width: 50,
            height: 50,
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
            color: DefaultPalette.white
        };
        var styles = mergeStyleSets({
            root: {
                background: DefaultPalette.themeTertiary,
                width: this.state.stackWidth + "%"
            },
            stackOne: {
                background: DefaultPalette.neutralTertiary,
                selectors: {
                    '& span': tslib_1.__assign({}, textStyles, { background: DefaultPalette.themePrimary })
                }
            },
            stackTwo: {
                background: DefaultPalette.neutralSecondary,
                selectors: {
                    '& span': tslib_1.__assign({}, textStyles, { background: DefaultPalette.themeDark })
                }
            },
            stackThree: {
                background: DefaultPalette.neutralPrimary,
                selectors: {
                    '& span': tslib_1.__assign({}, textStyles, { background: DefaultPalette.themeDarker })
                }
            }
        });
        return (React.createElement(Stack, { gap: 10 },
            React.createElement(Slider, { label: "Change the stack width to see how child items wrap onto multiple rows:", min: 1, max: 100, step: 1, defaultValue: 100, showValue: true, onChange: this._onWidthChange }),
            React.createElement(Stack, { horizontal: true, wrap: true, gap: '30 40', className: styles.root },
                React.createElement(Stack, { horizontal: true, wrap: true, gap: '10 30', className: styles.stackOne },
                    React.createElement("span", null, "1"),
                    React.createElement("span", null, "2"),
                    React.createElement("span", null, "3"),
                    React.createElement("span", null, "4"),
                    React.createElement("span", null, "5"),
                    React.createElement("span", null, "6"),
                    React.createElement("span", null, "7")),
                React.createElement(Stack, { horizontal: true, wrap: true, gap: '20 50', className: styles.stackTwo },
                    React.createElement("span", null, "1"),
                    React.createElement("span", null, "2"),
                    React.createElement("span", null, "3")),
                React.createElement(Stack, { horizontal: true, wrap: true, className: styles.stackThree },
                    React.createElement("span", null, "1"),
                    React.createElement("span", null, "2"),
                    React.createElement("span", null, "3"),
                    React.createElement("span", null, "4"),
                    React.createElement("span", null, "5"),
                    React.createElement("span", null, "6"),
                    React.createElement("span", null, "7"),
                    React.createElement("span", null, "8"),
                    React.createElement("span", null, "9"),
                    React.createElement("span", null, "10")))));
    };
    return HorizontalStackWrapNestedExample;
}(React.Component));
export { HorizontalStackWrapNestedExample };
