// office-ui-fabric-react/lib/components/Stack/examples/Stack.Horizontal.Spacing.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { Stack } from '../Stack';
import { mergeStyleSets, DefaultPalette } from 'office-ui-fabric-react/lib/Styling';
var HorizontalStackSpacingExample = /** @class */ (function (_super) {
    tslib_1.__extends(HorizontalStackSpacingExample, _super);
    function HorizontalStackSpacingExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HorizontalStackSpacingExample.prototype.render = function () {
        var styles = mergeStyleSets({
            root: {
                background: DefaultPalette.themeTertiary,
                width: 300,
                selectors: {
                    '> *': {
                        width: 50,
                        height: 50,
                        display: 'flex',
                        alignItems: 'center',
                        justifyContent: 'center',
                        background: DefaultPalette.themePrimary,
                        color: DefaultPalette.white
                    }
                }
            },
            item: {
                width: 50,
                height: 50,
                display: 'flex',
                alignItems: 'center',
                justifyContent: 'center',
                background: DefaultPalette.themePrimary,
                color: DefaultPalette.white
            }
        });
        var tokens = {
            numericalSpacing: {
                childrenGap: 10
            },
            customSpacing: {
                childrenGap: '10%'
            },
            themedExtraSmall: {
                childrenGap: 's2'
            },
            themedSmall: {
                childrenGap: 's1'
            },
            themedMedium: {
                childrenGap: 'm'
            },
            themedLarge: {
                childrenGap: 'l1'
            },
            themedExtraLarge: {
                childrenGap: 'l2'
            }
        };
        return (React.createElement(Stack, { tokens: tokens.numericalSpacing },
            React.createElement(Stack, { horizontal: true, disableShrink: true, horizontalAlign: "space-between" },
                React.createElement(Stack, null,
                    React.createElement("span", null, "Numerical spacing"),
                    React.createElement(Stack, { horizontal: true, className: styles.root, tokens: tokens.numericalSpacing, padding: 10 },
                        React.createElement("span", null, "1"),
                        React.createElement("span", null, "2"),
                        React.createElement("span", null, "3"))),
                React.createElement(Stack, null,
                    React.createElement("span", null, "Custom spacing"),
                    React.createElement(Stack, { horizontal: true, className: styles.root, tokens: tokens.customSpacing, padding: "s1 15%" },
                        React.createElement("span", null, "1"),
                        React.createElement("span", null, "2"),
                        React.createElement("span", null, "3")))),
            React.createElement(Stack, { horizontal: true, disableShrink: true, horizontalAlign: "space-between" },
                React.createElement(Stack, null,
                    React.createElement("span", null, "Themed spacing (extra small)"),
                    React.createElement(Stack, { horizontal: true, className: styles.root, tokens: tokens.themedExtraSmall, padding: "s2" },
                        React.createElement("span", null, "1"),
                        React.createElement("span", null, "2"),
                        React.createElement("span", null, "3"))),
                React.createElement(Stack, null,
                    React.createElement("span", null, "Themed spacing (small)"),
                    React.createElement(Stack, { horizontal: true, className: styles.root, tokens: tokens.themedSmall, padding: "s1" },
                        React.createElement("span", null, "1"),
                        React.createElement("span", null, "2"),
                        React.createElement("span", null, "3"))),
                React.createElement(Stack, null,
                    React.createElement("span", null, "Themed spacing (medium)"),
                    React.createElement(Stack, { horizontal: true, className: styles.root, tokens: tokens.themedMedium, padding: "m" },
                        React.createElement("span", null, "1"),
                        React.createElement("span", null, "2"),
                        React.createElement("span", null, "3")))),
            React.createElement(Stack, { horizontal: true, horizontalAlign: "space-between" },
                React.createElement(Stack, null,
                    React.createElement("span", null, "Themed spacing (large)"),
                    React.createElement(Stack, { horizontal: true, className: styles.root, tokens: tokens.themedLarge, padding: "l1" },
                        React.createElement("span", null, "1"),
                        React.createElement("span", null, "2"),
                        React.createElement("span", null, "3"))),
                React.createElement(Stack, null,
                    React.createElement("span", null, "Themed spacing (extra large)"),
                    React.createElement(Stack, { horizontal: true, className: styles.root, tokens: tokens.themedExtraLarge, padding: "l2" },
                        React.createElement("span", null, "1"),
                        React.createElement("span", null, "2"),
                        React.createElement("span", null, "3"))))));
    };
    return HorizontalStackSpacingExample;
}(React.Component));
export { HorizontalStackSpacingExample };
