// office-ui-fabric-react/lib/components/Stack/examples/Stack.Vertical.Reversed.Example.js
import * as tslib_1 from "tslib";
// @codepen
import * as React from 'react';
import { Stack } from '../Stack';
import { mergeStyleSets, DefaultPalette } from 'office-ui-fabric-react/lib/Styling';
var VerticalStackReversedExample = /** @class */ (function (_super) {
    tslib_1.__extends(VerticalStackReversedExample, _super);
    function VerticalStackReversedExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._onClick = function () {
            alert('Clicked VerticalStack');
        };
        return _this;
    }
    VerticalStackReversedExample.prototype.render = function () {
        var styles = mergeStyleSets({
            root: {
                background: DefaultPalette.themeTertiary
            },
            item: {
                color: DefaultPalette.white,
                background: DefaultPalette.themePrimary,
                padding: 5
            }
        });
        return (React.createElement(Stack, { gap: 5 },
            React.createElement("span", null, "Default vertical stack"),
            React.createElement(Stack, { reversed: true, className: styles.root },
                React.createElement("span", null, "Item One"),
                React.createElement("span", null, "Item Two"),
                React.createElement("span", null, "Item Three")),
            React.createElement("span", null, "Vertical gap between items"),
            React.createElement(Stack, { reversed: true, gap: 10, padding: 10, className: styles.root },
                React.createElement("span", null, "Item One"),
                React.createElement("span", null, "Item Two"),
                React.createElement("span", null, "Item Three")),
            React.createElement("span", null, "Item alignments"),
            React.createElement(Stack, { reversed: true, gap: 5, padding: 10, className: styles.root },
                React.createElement(Stack.Item, { align: "auto", className: styles.item },
                    React.createElement("span", null, "Auto-aligned item")),
                React.createElement(Stack.Item, { align: "stretch", className: styles.item },
                    React.createElement("span", null, "Stretch-aligned item")),
                React.createElement(Stack.Item, { align: "baseline", className: styles.item },
                    React.createElement("span", null, "Baseline-aligned item")),
                React.createElement(Stack.Item, { align: "start", className: styles.item },
                    React.createElement("span", null, "Start-aligned item")),
                React.createElement(Stack.Item, { align: "center", className: styles.item },
                    React.createElement("span", null, "Center-aligned item")),
                React.createElement(Stack.Item, { align: "end", className: styles.item },
                    React.createElement("span", null, "End-aligned item"))),
            React.createElement("span", null, "Clickable vertical stack"),
            React.createElement(Stack, { onClick: this._onClick, padding: 10, className: styles.root },
                React.createElement("span", null, "Click inside this box"))));
    };
    return VerticalStackReversedExample;
}(React.Component));
export { VerticalStackReversedExample };
