// office-ui-fabric-react/lib-amd/components/Slider/Slider.doc.js
define(["require", "exports", "react", "./examples/Slider.Basic.Example", "./Slider.checklist"], function (require, exports, React, Slider_Basic_Example_1, Slider_checklist_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var SliderBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Slider/examples/Slider.Basic.Example.tsx');
    exports.SliderPageProps = {
        title: 'Slider',
        componentName: 'Slider',
        componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/undefined',
        componentStatus: Slider_checklist_1.SliderStatus,
        examples: [
            {
                title: 'Slider',
                code: SliderBasicExampleCode,
                view: React.createElement(Slider_Basic_Example_1.SliderBasicExample, null)
            }
        ],
        propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Slider/Slider.types.ts')],
        overview: require('!raw-loader!office-ui-fabric-react/src/components/Slider/docs/SliderOverview.md'),
        bestPractices: '',
        dos: require('!raw-loader!office-ui-fabric-react/src/components/Slider/docs/SliderDos.md'),
        donts: require('!raw-loader!office-ui-fabric-react/src/components/Slider/docs/SliderDonts.md'),
        isHeaderVisible: true,
        isFeedbackVisible: true
    };
});
