// office-ui-fabric-react/lib-commonjs/components/Callout/Callout.doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var data_1 = require("../CommandBar/examples/data");
var Callout_Basic_Example_1 = require("./examples/Callout.Basic.Example");
var Callout_Nested_Example_1 = require("./examples/Callout.Nested.Example");
var Callout_FocusTrap_Example_1 = require("./examples/Callout.FocusTrap.Example");
var Callout_Directional_Example_1 = require("./examples/Callout.Directional.Example");
var Callout_Cover_Example_1 = require("./examples/Callout.Cover.Example");
var Callout_checklist_1 = require("./Callout.checklist");
var CalloutBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Callout/examples/Callout.Basic.Example.tsx');
var CalloutNestedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Callout/examples/Callout.Nested.Example.tsx');
var CalloutFocusTrapExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Callout/examples/Callout.FocusTrap.Example.tsx');
var CalloutDirectionalExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Callout/examples/Callout.Directional.Example.tsx');
var CalloutCoverExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Callout/examples/Callout.Cover.Example.tsx');
var cmdBarParamsTextAndIcons = { items: data_1.items, farItems: null };
exports.CalloutPageProps = {
    title: 'Callout',
    componentName: 'Callout',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Callout',
    componentStatus: Callout_checklist_1.CalloutStatus,
    examples: [
        {
            title: 'Default Callout',
            code: CalloutBasicExampleCode,
            view: React.createElement(Callout_Basic_Example_1.CalloutBasicExample, null)
        },
        {
            title: 'Nested Callout... Callout with a commandbar with a sub menu',
            code: CalloutNestedExampleCode,
            view: React.createElement(Callout_Nested_Example_1.CalloutNestedExample, tslib_1.__assign({}, cmdBarParamsTextAndIcons))
        },
        {
            title: 'Focus Trap Callout',
            code: CalloutFocusTrapExampleCode,
            view: React.createElement(Callout_FocusTrap_Example_1.CalloutFocusTrapExample, tslib_1.__assign({}, cmdBarParamsTextAndIcons))
        },
        {
            title: 'Callout with directional hint',
            code: CalloutDirectionalExampleCode,
            view: React.createElement(Callout_Directional_Example_1.CalloutDirectionalExample, null)
        },
        {
            title: 'Callout with cover',
            code: CalloutCoverExampleCode,
            view: React.createElement(Callout_Cover_Example_1.CalloutCoverExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Callout/Callout.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Callout/docs/CalloutOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Callout/docs/CalloutDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Callout/docs/CalloutDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
