// office-ui-fabric-react/lib/components/SelectedItemsList/SelectedPeopleList/SelectedPeopleList.doc.js
import * as React from 'react';
import { PeopleSelectedItemsListExample } from '../examples/SelectedPeopleList.Basic.Example';
var PeopleSelectedItemsListExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/SelectedItemsList/examples/SelectedPeopleList.Basic.Example.tsx');
export var SelectedPeopleListPageProps = {
    title: 'SelectedPeopleList',
    componentName: 'SelectedPeopleList',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/SelectedPeopleList',
    examples: [
        {
            title: 'Selected People List',
            code: PeopleSelectedItemsListExampleCode,
            view: React.createElement(PeopleSelectedItemsListExample, null)
        }
    ],
    propertiesTablesSources: [
        require('!raw-loader!office-ui-fabric-react/src/components/SelectedItemsList/BaseSelectedItemsList.types.ts')
    ],
    overview: '',
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/SelectedItemsList/docs/SelectedPeopleListDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/SelectedItemsList/docs/SelectedPeopleListDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
