// office-ui-fabric-react/lib/components/MessageBar/MessageBar.doc.js
import * as React from 'react';
import { MessageBarBasicExample } from './examples/MessageBar.Basic.Example';
import { MessageBarStyledExample } from './examples/MessageBar.Styled.Example';
import { MessageBarStatus } from './MessageBar.checklist';
var MessageBarBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/MessageBar/examples/MessageBar.Basic.Example.tsx');
var MessageBarBasicExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/MessageBar/examples/MessageBar.Basic.Example.tsx');
var MessageBarStyledExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/MessageBar/examples/MessageBar.Styled.Example.tsx');
var MessageBarStyledExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/MessageBar/examples/MessageBar.Styled.Example.tsx');
export var MessageBarPageProps = {
    title: 'MessageBar',
    componentName: 'MessageBar',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/MessageBar',
    componentStatus: MessageBarStatus,
    examples: [
        {
            title: 'Various MessageBar types',
            code: MessageBarBasicExampleCode,
            view: React.createElement(MessageBarBasicExample, null),
            codepenJS: MessageBarBasicExampleCodepen
        },
        {
            title: 'Custom styled MessageBars',
            code: MessageBarStyledExampleCode,
            view: React.createElement(MessageBarStyledExample, null),
            codepenJS: MessageBarStyledExampleCodepen
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/MessageBar/MessageBar.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/MessageBar/docs/MessageBarOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/MessageBar/docs/MessageBarDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/MessageBar/docs/MessageBarDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
