// office-ui-fabric-react/lib/components/Label/Label.doc.js
import * as React from 'react';
import { LabelBasicExample } from './examples/Label.Basic.Example';
import { LabelStatus } from './Label.checklist';
var LabelBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Label/examples/Label.Basic.Example.tsx');
export var LabelPageProps = {
    title: 'Label',
    componentName: 'Label',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Label',
    componentStatus: LabelStatus,
    examples: [
        {
            title: 'Label',
            code: LabelBasicExampleCode,
            view: React.createElement(LabelBasicExample, null),
            codepenJS: require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Label/examples/Label.Basic.Example.tsx')
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Label/Label.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Label/docs/LabelOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Label/docs/LabelDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Label/docs/LabelDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true,
    allowNativeProps: true
};
