// office-ui-fabric-react/lib-commonjs/components/Rating/Rating.doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var Rating_Basic_Example_1 = require("./examples/Rating.Basic.Example");
var Rating_ButtonControlled_Example_1 = require("./examples/Rating.ButtonControlled.Example");
var Rating_checklist_1 = require("./Rating.checklist");
var RatingBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Rating/examples/Rating.Basic.Example.tsx');
var RatingBasicExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Rating/examples/Rating.Basic.Example.tsx');
var RatingButtonControlledExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Rating/examples/Rating.ButtonControlled.Example.tsx');
exports.RatingPageProps = {
    title: 'Rating',
    componentName: 'Rating',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Rating',
    componentStatus: Rating_checklist_1.RatingStatus,
    examples: [
        {
            title: 'Rating',
            code: RatingBasicExampleCode,
            view: React.createElement(Rating_Basic_Example_1.RatingBasicExample, null),
            codepenJS: RatingBasicExampleCodepen
        },
        {
            title: 'Button Controlled Rating',
            code: RatingButtonControlledExampleCode,
            view: React.createElement(Rating_ButtonControlled_Example_1.RatingButtonControlledExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Rating/Rating.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Rating/docs/RatingOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Rating/docs/RatingDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Rating/docs/RatingDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
