// office-ui-fabric-react/lib-amd/components/Label/Label.styles.js
define(["require", "exports", "../../Styling"], function (require, exports, Styling_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getStyles = function (props) {
        var theme = props.theme, className = props.className, disabled = props.disabled, required = props.required;
        return {
            root: [
                'ms-Label',
                theme.fonts.medium,
                {
                    color: theme.semanticColors.bodyText,
                    boxSizing: 'border-box',
                    boxShadow: 'none',
                    margin: 0,
                    display: 'block',
                    padding: '5px 0',
                    wordWrap: 'break-word',
                    overflowWrap: 'break-word'
                },
                disabled && {
                    color: theme.semanticColors.disabledBodyText,
                    selectors: (_a = {},
                        _a[Styling_1.HighContrastSelector] = {
                            color: 'GrayText'
                        },
                        _a)
                },
                required && {
                    selectors: {
                        '::after': {
                            content: "' *'",
                            color: theme.semanticColors.errorText,
                            paddingRight: 12
                        }
                    }
                },
                className
            ]
        };
        var _a;
    };
});
