// office-ui-fabric-react/lib-amd/components/Nav/examples/Nav.ByKeys.Example.js
define(["require", "exports", "tslib", "react", "office-ui-fabric-react/lib/Nav"], function (require, exports, tslib_1, React, Nav_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var NavByKeysExample = /** @class */ (function (_super) {
        tslib_1.__extends(NavByKeysExample, _super);
        function NavByKeysExample() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        NavByKeysExample.prototype.render = function () {
            return (React.createElement(Nav_1.Nav, { groups: [
                    {
                        links: [
                            { name: 'Home', key: 'Home', url: '' },
                            { name: 'Activity', key: 'Activity', url: '' },
                            { name: 'News', key: 'News', url: '' },
                            { name: 'Documents', key: 'Documents', url: '' }
                        ]
                    }
                ] }));
        };
        return NavByKeysExample;
    }(React.Component));
    exports.NavByKeysExample = NavByKeysExample;
});
