// office-ui-fabric-react/lib-amd/components/Nav/examples/Nav.CustomGroupHeaders.Example.js
define(["require", "exports", "react", "office-ui-fabric-react/lib/Nav"], function (require, exports, React, Nav_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.NavCustomGroupHeadersExample = function () {
        return (React.createElement(Nav_1.Nav, { onRenderGroupHeader: _onRenderGroupHeader, groups: [
                {
                    name: 'Pages',
                    links: [
                        // prettier-ignore
                        { name: 'Activity', url: 'http://msn.com', key: 'key1' },
                        { name: 'News', url: 'http://msn.com', key: 'key2' }
                    ]
                },
                {
                    name: 'More pages',
                    links: [
                        // prettier-ignore
                        { name: 'Settings', url: 'http://msn.com', key: 'key3' },
                        { name: 'Notes', url: 'http://msn.com', key: 'key4' }
                    ]
                }
            ] }));
    };
    function _onRenderGroupHeader(group) {
        return React.createElement("h3", null, group.name);
    }
});
