// office-ui-fabric-react/lib-commonjs/components/Layer/Layer.doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var Layer_Basic_Example_1 = require("./examples/Layer.Basic.Example");
var Layer_Hosted_Example_1 = require("./examples/Layer.Hosted.Example");
var Layer_Customized_Example_1 = require("./examples/Layer.Customized.Example");
var Layer_NestedLayers_Example_1 = require("./examples/Layer.NestedLayers.Example");
var Layer_checklist_1 = require("./Layer.checklist");
var LayerBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Layer/examples/Layer.Basic.Example.tsx');
var LayerHostedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Layer/examples/Layer.Hosted.Example.tsx');
var LayerCustomizedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Layer/examples/Layer.Customized.Example.tsx');
var LayerNestedLayersExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Layer/examples/Layer.NestedLayers.Example.tsx');
exports.LayerPageProps = {
    title: 'Layer',
    componentName: 'Layer',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Layer',
    componentStatus: Layer_checklist_1.LayerStatus,
    examples: [
        {
            title: 'Basic layered content',
            code: LayerBasicExampleCode,
            view: React.createElement(Layer_Basic_Example_1.LayerBasicExample, null)
        },
        {
            title: 'Using LayerHost to control projection',
            code: LayerHostedExampleCode,
            view: React.createElement(Layer_Hosted_Example_1.LayerHostedExample, null)
        },
        {
            title: 'Using Customizer to control the default layer behavior',
            code: LayerCustomizedExampleCode,
            view: React.createElement(Layer_Customized_Example_1.LayerCustomizedExample, null)
        },
        {
            title: 'Nested Layers Example',
            code: LayerNestedLayersExampleCode,
            view: React.createElement(Layer_NestedLayers_Example_1.LayerNestedLayersExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Layer/Layer.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Layer/docs/LayerOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Layer/docs/LayerDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Layer/docs/LayerDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
