// office-ui-fabric-react/lib-amd/components/TextField/TextField.doc.js
define(["require", "exports", "react", "./examples/TextField.Basic.Example", "./examples/TextField.Borderless.Example", "./examples/TextField.CustomRender.Example", "./examples/TextField.ErrorMessage.Example", "./examples/TextField.Icon.Example", "./examples/TextField.Multiline.Example", "./examples/TextField.Placeholder.Example", "./examples/TextField.Prefix.Example", "./examples/TextField.PrefixAndSuffix.Example", "./TextField.checklist", "./examples/TextField.Styled.Example", "./examples/TextField.Suffix.Example", "./examples/TextField.Underlined.Example", "./examples/TextField.AutoComplete.Example", "./examples/TextField.OnRenderDescription.Example"], function (require, exports, React, TextField_Basic_Example_1, TextField_Borderless_Example_1, TextField_CustomRender_Example_1, TextField_ErrorMessage_Example_1, TextField_Icon_Example_1, TextField_Multiline_Example_1, TextField_Placeholder_Example_1, TextField_Prefix_Example_1, TextField_PrefixAndSuffix_Example_1, TextField_checklist_1, TextField_Styled_Example_1, TextField_Suffix_Example_1, TextField_Underlined_Example_1, TextField_AutoComplete_Example_1, TextField_OnRenderDescription_Example_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var TextFieldBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Basic.Example.tsx');
    var TextFieldBasicExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/TextField/TextField.Basic.Example.Codepen.txt');
    var TextFieldBorderlessExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Borderless.Example.tsx');
    var TextFieldBorderlessExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/TextField/TextField.Borderless.Example.Codepen.txt');
    var TextFieldCustomRenderExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.CustomRender.Example.tsx');
    var TextFieldErrorMessageExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.ErrorMessage.Example.tsx');
    var TextFieldIconExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Icon.Example.tsx');
    var TextFieldMultilineExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Multiline.Example.tsx');
    var TextFieldPlaceholderExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Placeholder.Example.tsx');
    var TextFieldPlaceholderExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/TextField/TextField.Placeholder.Example.Codepen.txt');
    var TextFieldPrefixExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Prefix.Example.tsx');
    var TextFieldPrefixAndSuffixExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.PrefixAndSuffix.Example.tsx');
    var TextFieldStyledExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Styled.Example.tsx');
    var TextFieldSuffixExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Suffix.Example.tsx');
    var TextFieldUnderlinedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Underlined.Example.tsx');
    var TextFieldUnderlinedExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/TextField/TextField.Underlined.Example.Codepen.txt');
    var TextFieldAutoCompleteExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.AutoComplete.Example.tsx');
    var TextFieldOnRenderDescriptionExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.OnRenderDescription.Example.tsx');
    exports.TextFieldPageProps = {
        title: 'TextField',
        componentName: 'TextField',
        componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/TextField',
        componentStatus: TextField_checklist_1.TextFieldStatus,
        examples: [
            {
                title: 'Default TextField with Label',
                code: TextFieldBasicExampleCode,
                view: React.createElement(TextField_Basic_Example_1.TextFieldBasicExample, null),
                codepenJS: TextFieldBasicExampleCodepen
            },
            {
                title: 'TextField with Placeholder',
                code: TextFieldPlaceholderExampleCode,
                view: React.createElement(TextField_Placeholder_Example_1.TextFieldPlaceholderExample, null),
                codepenJS: TextFieldPlaceholderExampleCodepen
            },
            {
                title: 'Multiline TextField',
                code: TextFieldMultilineExampleCode,
                view: React.createElement(TextField_Multiline_Example_1.TextFieldMultilineExample, null)
            },
            {
                title: 'Underlined TextField',
                code: TextFieldUnderlinedExampleCode,
                view: React.createElement(TextField_Underlined_Example_1.TextFieldUnderlinedExample, null),
                codepenJS: TextFieldUnderlinedExampleCodepen
            },
            {
                title: 'Borderless TextField',
                code: TextFieldBorderlessExampleCode,
                view: React.createElement(TextField_Borderless_Example_1.TextFieldBorderlessExample, null),
                codepenJS: TextFieldBorderlessExampleCodepen
            },
            {
                title: 'TextField with browser AutoComplete',
                code: TextFieldAutoCompleteExampleCode,
                view: React.createElement(TextField_AutoComplete_Example_1.TextFieldAutoCompleteExample, null)
            }
        ],
        implementationExamples: [
            {
                title: 'Textfield with a prefix',
                code: TextFieldPrefixExampleCode,
                view: React.createElement(TextField_Prefix_Example_1.TextFieldPrefixExample, null)
            },
            {
                title: 'Textfield with a suffix',
                code: TextFieldSuffixExampleCode,
                view: React.createElement(TextField_Suffix_Example_1.TextFieldSuffixExample, null)
            },
            {
                title: 'Textfield with a prefix and a suffix',
                code: TextFieldPrefixAndSuffixExampleCode,
                view: React.createElement(TextField_PrefixAndSuffix_Example_1.TextFieldPrefixAndSuffixExample, null)
            },
            {
                title: 'TextField with an icon',
                code: TextFieldIconExampleCode,
                view: React.createElement(TextField_Icon_Example_1.TextFieldIconExample, null)
            },
            {
                title: 'TextField with custom Label',
                code: TextFieldCustomRenderExampleCode,
                view: React.createElement(TextField_CustomRender_Example_1.TextFieldCustomRenderExample, null)
            },
            {
                title: 'TextField with custom description',
                code: TextFieldOnRenderDescriptionExampleCode,
                view: React.createElement(TextField_OnRenderDescription_Example_1.TextFieldOnRenderDescriptionExample, null)
            },
            {
                title: 'TextField error message variations',
                code: TextFieldErrorMessageExampleCode,
                view: React.createElement(TextField_ErrorMessage_Example_1.TextFieldErrorMessageExample, null)
            },
            {
                title: 'TextField Subcomponent Styling',
                code: TextFieldStyledExampleCode,
                view: React.createElement(TextField_Styled_Example_1.TextFieldStyledExample, null)
            }
        ],
        propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/TextField/TextField.types.ts')],
        overview: require('!raw-loader!office-ui-fabric-react/src/components/TextField/docs/TextFieldOverview.md'),
        bestPractices: '',
        dos: require('!raw-loader!office-ui-fabric-react/src/components/TextField/docs/TextFieldDos.md'),
        donts: require('!raw-loader!office-ui-fabric-react/src/components/TextField/docs/TextFieldDonts.md'),
        isHeaderVisible: true,
        isFeedbackVisible: true,
        allowNativeProps: true,
        nativePropsElement: ['input', 'textarea']
    };
});
