// office-ui-fabric-react/lib/components/ScrollablePane/ScrollablePane.doc.js
import * as React from 'react';
import { ScrollablePaneDefaultExample } from './examples/ScrollablePane.Default.Example';
import { ScrollablePaneDetailsListExample } from './examples/ScrollablePane.DetailsList.Example';
import { ScrollablePaneStatus } from './ScrollablePane.checklist';
var ScrollablePaneDefaultExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ScrollablePane/examples/ScrollablePane.Default.Example.tsx');
var ScrollablePaneDetailsListExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ScrollablePane/examples/ScrollablePane.DetailsList.Example.tsx');
var ScrollablePaneDefaultExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/ScrollablePane/examples/ScrollablePane.Default.Example.tsx');
var ScrollablePaneDetailsListExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/ScrollablePane/examples/ScrollablePane.DetailsList.Example.tsx');
export var ScrollablePanePageProps = {
    title: 'ScrollablePane',
    componentName: 'ScrollablePane',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/ScrollablePane',
    componentStatus: ScrollablePaneStatus,
    examples: [
        {
            title: 'Default',
            code: ScrollablePaneDefaultExampleCode,
            codepenJS: ScrollablePaneDefaultExampleCodepen,
            view: React.createElement(ScrollablePaneDefaultExample, null),
            isScrollable: false
        },
        {
            title: 'DetailsList Locked Header',
            code: ScrollablePaneDetailsListExampleCode,
            codepenJS: ScrollablePaneDetailsListExampleCodepen,
            view: React.createElement(ScrollablePaneDetailsListExample, null),
            isScrollable: false
        }
    ],
    propertiesTablesSources: [
        require('!raw-loader!office-ui-fabric-react/src/components/ScrollablePane/ScrollablePane.types.ts'),
        require('!raw-loader!office-ui-fabric-react/src/components/Sticky/Sticky.types.ts')
    ],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/ScrollablePane/docs/ScrollablePaneOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/ScrollablePane/docs/ScrollablePaneDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/ScrollablePane/docs/ScrollablePaneDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true,
    allowNativeProps: true,
    nativePropsElement: ['a', 'button']
};
