// office-ui-fabric-react/lib-amd/components/MessageBar/MessageBar.doc.js
define(["require", "exports", "react", "./examples/MessageBar.Basic.Example", "./MessageBar.checklist"], function (require, exports, React, MessageBar_Basic_Example_1, MessageBar_checklist_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var MessageBarBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/MessageBar/examples/MessageBar.Basic.Example.tsx');
    var MessageBarBasicExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/MessageBar/MessageBar.Basic.Example.Codepen.txt');
    exports.MessageBarPageProps = {
        title: 'MessageBar',
        componentName: 'MessageBar',
        componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/MessageBar',
        componentStatus: MessageBar_checklist_1.MessageBarStatus,
        examples: [
            {
                title: 'Various MessageBar types',
                code: MessageBarBasicExampleCode,
                view: React.createElement(MessageBar_Basic_Example_1.MessageBarBasicExample, null),
                codepenJS: MessageBarBasicExampleCodepen
            }
        ],
        propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/MessageBar/MessageBar.types.ts')],
        overview: require('!raw-loader!office-ui-fabric-react/src/components/MessageBar/docs/MessageBarOverview.md'),
        bestPractices: '',
        dos: require('!raw-loader!office-ui-fabric-react/src/components/MessageBar/docs/MessageBarDos.md'),
        donts: require('!raw-loader!office-ui-fabric-react/src/components/MessageBar/docs/MessageBarDonts.md'),
        isHeaderVisible: true,
        isFeedbackVisible: true
    };
});
