// office-ui-fabric-react/lib/components/Layer/examples/Layer.Basic.Example.js
import * as tslib_1 from "tslib";
import { Toggle } from 'office-ui-fabric-react/lib/Toggle';
import 'office-ui-fabric-react/lib/common/_exampleStyles.scss';
import { Layer } from 'office-ui-fabric-react/lib/Layer';
import { AnimationClassNames } from 'office-ui-fabric-react/lib/Styling';
import { BaseComponent, css } from 'office-ui-fabric-react/lib/Utilities';
import * as PropTypes from 'prop-types';
import * as React from 'react';
import * as styles from './Layer.Example.scss';
var LayerContentExample = /** @class */ (function (_super) {
    tslib_1.__extends(LayerContentExample, _super);
    function LayerContentExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            time: new Date().toLocaleTimeString()
        };
        return _this;
    }
    LayerContentExample.prototype.componentDidMount = function () {
        var _this = this;
        this._async.setInterval(function () {
            _this.setState({
                time: new Date().toLocaleTimeString()
            });
        }, 1000);
    };
    LayerContentExample.prototype.render = function () {
        return (React.createElement("div", { className: css(styles.content, AnimationClassNames.scaleUpIn100) },
            React.createElement("div", { className: styles.textContent }, this.context.message),
            React.createElement("div", null, this.state.time)));
    };
    LayerContentExample.contextTypes = {
        message: PropTypes.string
    };
    return LayerContentExample;
}(BaseComponent));
export { LayerContentExample };
var LayerBasicExample = /** @class */ (function (_super) {
    tslib_1.__extends(LayerBasicExample, _super);
    function LayerBasicExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            showLayer: false
        };
        _this._onChange = function (ev, checked) {
            _this.setState({ showLayer: checked });
        };
        return _this;
    }
    LayerBasicExample.prototype.getChildContext = function () {
        return {
            message: 'Hello world.'
        };
    };
    LayerBasicExample.prototype.render = function () {
        var showLayer = this.state.showLayer;
        return (React.createElement("div", null,
            React.createElement(Toggle, { label: "Wrap the content box below in a Layer", inlineLabel: true, checked: showLayer, onChange: this._onChange }),
            showLayer ? (React.createElement(Layer, null,
                React.createElement(LayerContentExample, null))) : (React.createElement(LayerContentExample, null))));
    };
    LayerBasicExample.childContextTypes = {
        message: PropTypes.string
    };
    return LayerBasicExample;
}(BaseComponent));
export { LayerBasicExample };
