// office-ui-fabric-react/lib-amd/components/Shimmer/ShimmerCircle/ShimmerCircle.styles.js
define(["require", "exports", "../../../Styling"], function (require, exports, Styling_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var GlobalClassNames = {
        root: 'ms-ShimmerCircle-root',
        svg: 'ms-ShimmerCircle-svg'
    };
    function getStyles(props) {
        var height = props.height, borderStyle = props.borderStyle, theme = props.theme;
        var palette = theme.palette;
        var globalClassNames = Styling_1.getGlobalClassNames(GlobalClassNames, theme);
        var borderStyles = !!borderStyle ? borderStyle : {};
        return {
            root: [
                globalClassNames.root,
                theme.fonts.medium,
                {
                    width: height + "px",
                    height: height + "px",
                    minWidth: height + "px",
                    boxSizing: 'content-box',
                    borderTopStyle: 'solid',
                    borderBottomStyle: 'solid',
                    borderColor: palette.white,
                    selectors: (_a = {},
                        _a[Styling_1.HighContrastSelector] = {
                            borderColor: 'Window'
                        },
                        _a)
                },
                borderStyles
            ],
            svg: [
                globalClassNames.svg,
                {
                    display: 'block',
                    fill: palette.white,
                    selectors: (_b = {},
                        _b[Styling_1.HighContrastSelector] = {
                            fill: 'Window'
                        },
                        _b)
                }
            ]
        };
        var _a, _b;
    }
    exports.getStyles = getStyles;
});
