// office-ui-fabric-react/lib-commonjs/components/Dropdown/Dropdown.doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var Dropdown_Basic_Example_1 = require("./examples/Dropdown.Basic.Example");
var Dropdown_Controlled_Example_1 = require("./examples/Dropdown.Controlled.Example");
var Dropdown_ControlledMulti_Example_1 = require("./examples/Dropdown.ControlledMulti.Example");
var Dropdown_Custom_Example_1 = require("./examples/Dropdown.Custom.Example");
var Dropdown_Error_Example_1 = require("./examples/Dropdown.Error.Example");
var Dropdown_Required_Example_1 = require("./examples/Dropdown.Required.Example");
var Dropdown_checklist_1 = require("./Dropdown.checklist");
var DropdownBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Basic.Example.tsx');
var DropdownBasicExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Basic.Example.tsx');
var DropdownControlledExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Controlled.Example.tsx');
var DropdownControlledExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Controlled.Example.tsx');
var DropdownControlledMultiExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.ControlledMulti.Example.tsx');
var DropdownControlledMultiExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.ControlledMulti.Example.tsx');
var DropdownCustomExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Custom.Example.tsx');
var DropdownCustomExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Custom.Example.tsx');
var DropdownErrorExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Error.Example.tsx');
var DropdownErrorExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Error.Example.tsx');
var DropdownRequiredExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Required.Example.tsx');
var DropdownRequiredExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Required.Example.tsx');
exports.DropdownPageProps = {
    title: 'Dropdown',
    componentName: 'Dropdown',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Dropdown',
    componentStatus: Dropdown_checklist_1.DropdownStatus,
    examples: [
        {
            title: 'Basic Dropdowns',
            code: DropdownBasicExampleCode,
            view: React.createElement(Dropdown_Basic_Example_1.DropdownBasicExample, null),
            codepenJS: DropdownBasicExampleCodepen
        },
        {
            title: 'Controlled single-select Dropdown',
            code: DropdownControlledExampleCode,
            view: React.createElement(Dropdown_Controlled_Example_1.DropdownControlledExample, null),
            codepenJS: DropdownControlledExampleCodepen
        },
        {
            title: 'Controlled multi-select Dropdown',
            code: DropdownControlledMultiExampleCode,
            view: React.createElement(Dropdown_ControlledMulti_Example_1.DropdownControlledMultiExample, null),
            codepenJS: DropdownControlledMultiExampleCodepen
        },
        {
            title: 'Customized Dropdown',
            code: DropdownCustomExampleCode,
            codepenJS: DropdownCustomExampleCodepen,
            view: React.createElement(Dropdown_Custom_Example_1.DropdownCustomExample, null)
        },
        {
            title: 'Dropdown with error message',
            code: DropdownErrorExampleCode,
            codepenJS: DropdownErrorExampleCodepen,
            view: React.createElement(Dropdown_Error_Example_1.DropdownErrorExample, null)
        },
        {
            title: 'Required Dropdown',
            code: DropdownRequiredExampleCode,
            codepenJS: DropdownRequiredExampleCodepen,
            view: (React.createElement(React.Fragment, null,
                React.createElement("p", null, "This example also demonstrates how to programmatically set focus on and open a Dropdown."),
                React.createElement(Dropdown_Required_Example_1.DropdownRequiredExample, null)))
        }
    ],
    propertiesTablesSources: [
        require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/Dropdown.types.ts'),
        require('!raw-loader!office-ui-fabric-react/src/utilities/selectableOption/SelectableDroppableText.types.ts')
    ],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/docs/DropdownOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/docs/DropdownDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/docs/DropdownDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true,
    allowNativeProps: true
};
