// office-ui-fabric-react/lib-commonjs/components/Image/Image.doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var Image_Center_Example_1 = require("./examples/Image.Center.Example");
var Image_CenterCover_Example_1 = require("./examples/Image.CenterCover.Example");
var Image_Contain_Example_1 = require("./examples/Image.Contain.Example");
var Image_Cover_Example_1 = require("./examples/Image.Cover.Example");
var Image_Default_Example_1 = require("./examples/Image.Default.Example");
var Image_MaximizeFrame_Example_1 = require("./examples/Image.MaximizeFrame.Example");
var Image_None_Example_1 = require("./examples/Image.None.Example");
var Image_checklist_1 = require("./Image.checklist");
require("./ImagePage.global.scss");
var ImageDefaultExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Image/examples/Image.Default.Example.tsx');
var ImageCenterExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Image/examples/Image.Center.Example.tsx');
var ImageContainExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Image/examples/Image.Contain.Example.tsx');
var ImageCoverExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Image/examples/Image.Cover.Example.tsx');
var ImageCenterCoverExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Image/examples/Image.CenterCover.Example.tsx');
var ImageNoneExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Image/examples/Image.None.Example.tsx');
var ImageMaximizeFrameExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Image/examples/Image.MaximizeFrame.Example.tsx');
exports.ImagePageProps = {
    title: 'Image',
    componentName: 'Image',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Image',
    componentStatus: Image_checklist_1.ImageStatus,
    examples: [
        {
            title: 'ImageFit: Not specified',
            code: ImageDefaultExampleCode,
            view: React.createElement(Image_Default_Example_1.ImageDefaultExample, null)
        },
        {
            title: 'ImageFit: None',
            code: ImageNoneExampleCode,
            view: React.createElement(Image_None_Example_1.ImageNoneExample, null)
        },
        {
            title: 'ImageFit: Center',
            code: ImageCenterExampleCode,
            view: React.createElement(Image_Center_Example_1.ImageCenterExample, null)
        },
        {
            title: 'ImageFit: Contain',
            code: ImageContainExampleCode,
            view: React.createElement(Image_Contain_Example_1.ImageContainExample, null)
        },
        {
            title: 'ImageFit: Cover',
            code: ImageCoverExampleCode,
            view: React.createElement(Image_Cover_Example_1.ImageCoverExample, null)
        },
        {
            title: 'ImageFit: CenterCover',
            code: ImageCenterCoverExampleCode,
            view: React.createElement(Image_CenterCover_Example_1.ImageCenterCoverExample, null)
        },
        {
            title: 'Maximizing the image frame',
            code: ImageMaximizeFrameExampleCode,
            view: React.createElement(Image_MaximizeFrame_Example_1.ImageMaximizeFrameExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Image/Image.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Image/docs/ImageOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Image/docs/ImageDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Image/docs/ImageDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true,
    allowNativeProps: true,
    nativePropsElement: 'img'
};
