// office-ui-fabric-react/lib-amd/components/ProgressIndicator/ProgressIndicator.doc.js
define(["require", "exports", "react", "./examples/ProgressIndicator.Basic.Example", "./examples/ProgressIndicator.Indeterminate.Example", "./ProgressIndicator.checklist"], function (require, exports, React, ProgressIndicator_Basic_Example_1, ProgressIndicator_Indeterminate_Example_1, ProgressIndicator_checklist_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ProgressIndicatorBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ProgressIndicator/examples/ProgressIndicator.Basic.Example.tsx');
    var ProgressIndicatorIndeterminateExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ProgressIndicator/examples/ProgressIndicator.Indeterminate.Example.tsx');
    exports.ProgressIndicatorPageProps = {
        title: 'ProgressIndicator',
        componentName: 'ProgressIndicator',
        componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/ProgressIndicator',
        componentStatus: ProgressIndicator_checklist_1.ProgressIndicatorStatus,
        examples: [
            {
                title: 'Default ProgressIndicator',
                code: ProgressIndicatorBasicExampleCode,
                view: React.createElement(ProgressIndicator_Basic_Example_1.ProgressIndicatorBasicExample, null)
            },
            {
                title: 'Indeterminate ProgressIndicator',
                code: ProgressIndicatorIndeterminateExampleCode,
                view: React.createElement(ProgressIndicator_Indeterminate_Example_1.ProgressIndicatorIndeterminateExample, null)
            }
        ],
        propertiesTablesSources: [
            require('!raw-loader!office-ui-fabric-react/src/components/ProgressIndicator/ProgressIndicator.types.ts')
        ],
        overview: require('!raw-loader!office-ui-fabric-react/src/components/ProgressIndicator/docs/ProgressIndicatorOverview.md'),
        bestPractices: '',
        dos: require('!raw-loader!office-ui-fabric-react/src/components/ProgressIndicator/docs/ProgressIndicatorDos.md'),
        donts: require('!raw-loader!office-ui-fabric-react/src/components/ProgressIndicator/docs/ProgressIndicatorDonts.md'),
        isHeaderVisible: true,
        isFeedbackVisible: true
    };
});
