// office-ui-fabric-react/lib-amd/components/ResizeGroup/ResizeGroup.doc.js
define(["require", "exports", "react", "./examples/ResizeGroup.OverflowSet.Example", "./examples/ResizeGroup.FlexBox.Example", "./ResizeGroup.checklist"], function (require, exports, React, ResizeGroup_OverflowSet_Example_1, ResizeGroup_FlexBox_Example_1, ResizeGroup_checklist_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ResizeGroupBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ResizeGroup/examples/ResizeGroup.OverflowSet.Example.tsx');
    var ResizeGroupFlexBoxExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ResizeGroup/examples/ResizeGroup.FlexBox.Example.tsx');
    exports.ResizeGroupPageProps = {
        title: 'ResizeGroup',
        componentName: 'ResizeGroup',
        componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/ResizeGroup',
        componentStatus: ResizeGroup_checklist_1.ResizeGroupStatus,
        examples: [
            {
                title: 'Use ResizeGroup to move commands into an overflow menu',
                code: ResizeGroupBasicExampleCode,
                view: React.createElement(ResizeGroup_OverflowSet_Example_1.ResizeGroupOverflowSetExample, null)
            },
            {
                title: 'Use ResizeGroup to prevent two groups of items from overlapping',
                code: ResizeGroupFlexBoxExampleCode,
                view: React.createElement(ResizeGroup_FlexBox_Example_1.FlexBoxResizeGroupExample, null)
            }
        ],
        propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/ResizeGroup/ResizeGroup.types.ts')],
        overview: require('!raw-loader!office-ui-fabric-react/src/components/ResizeGroup/docs/ResizeGroupOverview.md'),
        bestPractices: '',
        dos: require('!raw-loader!office-ui-fabric-react/src/components/ResizeGroup/docs/ResizeGroupDos.md'),
        donts: require('!raw-loader!office-ui-fabric-react/src/components/ResizeGroup/docs/ResizeGroupDonts.md'),
        isHeaderVisible: true,
        isFeedbackVisible: true,
        allowNativeProps: true
    };
});
