// office-ui-fabric-react/lib/components/ActivityItem/examples/ActivityItem.Persona.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { css, classNamesFunction } from 'office-ui-fabric-react/lib/Utilities';
import { ActivityItem } from '../../../ActivityItem';
import { getStyles } from './ActivityItem.Example.styles';
import { TestImages } from '../../../common/TestImages';
import { Link } from '../../../Link';
var ActivityItemPersonaExample = /** @class */ (function (_super) {
    tslib_1.__extends(ActivityItemPersonaExample, _super);
    function ActivityItemPersonaExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ActivityItemPersonaExample.prototype.render = function () {
        var getClassNames = classNamesFunction();
        var classNames = getClassNames(getStyles, {});
        // tslint:disable:jsx-no-lambda
        var activityItemExamples = [
            {
                key: 1,
                activityDescription: [
                    React.createElement(Link, { key: 1, className: css(classNames.nameText), onClick: function () {
                            alert('A name was clicked.');
                        } }, "Jack Howden"),
                    React.createElement("span", { key: 2 }, " renamed "),
                    React.createElement("span", { key: 3, className: css(classNames.nameText) }, "DocumentTitle.docx")
                ],
                activityPersonas: [
                    {
                        imageUrl: TestImages.personaMale
                    }
                ],
                comments: 'Hello, this is the text of my basic comment!',
                timeStamp: '23m ago'
            },
            {
                key: 2,
                activityDescription: [
                    React.createElement(Link, { key: 1, className: css(classNames.nameText), onClick: function () {
                            alert('A name was clicked.');
                        } }, "Javiera M\u00E1rquez"),
                    React.createElement("span", { key: 2 }, " and "),
                    React.createElement(Link, { key: 3, className: css(classNames.nameText), onClick: function () {
                            alert('A name was clicked.');
                        } }, "Amelia Poval\u0456y"),
                    React.createElement("span", { key: 4 }, " edited "),
                    React.createElement(Link, { key: 5, className: css(classNames.nameText), onClick: function () {
                            alert('A document was clicked.');
                        } }, "SpreadsheetTitle.xlsx")
                ],
                activityPersonas: [
                    {
                        imageInitials: 'JM',
                        text: 'Javiera Márquez'
                    },
                    {
                        imageUrl: TestImages.personaFemale
                    }
                ],
                timeStamp: '9:27 am'
            },
            {
                key: 3,
                activityDescription: [
                    React.createElement(Link, { key: 1, className: css(classNames.nameText), onClick: function () {
                            alert('A name was clicked.');
                        } }, "Robert Larsson"),
                    React.createElement("span", { key: 2 }, " and "),
                    React.createElement(Link, { key: 3, className: css(classNames.nameText), onClick: function () {
                            alert('A name was clicked.');
                        } }, "2 others"),
                    React.createElement("span", { key: 4 }, " commented ")
                ],
                activityPersonas: [
                    {
                        imageInitials: 'RL',
                        text: 'Robert Larsson'
                    },
                    {
                        imageUrl: TestImages.personaMale
                    },
                    {
                        imageUrl: TestImages.personaFemale
                    }
                ],
                timeStamp: '3 days ago'
            },
            {
                key: 4,
                activityDescription: [
                    React.createElement(Link, { key: 1, className: css(classNames.nameText), onClick: function () {
                            alert('A name was clicked.');
                        } }, "Jin Cheng"),
                    React.createElement("span", { key: 2 }, " and "),
                    React.createElement(Link, { key: 3, className: css(classNames.nameText), onClick: function () {
                            alert('A name was clicked.');
                        } }, "5 others"),
                    React.createElement("span", { key: 4 }, " edited this file")
                ],
                activityPersonas: [
                    {
                        imageInitials: 'JC',
                        text: 'Jin Cheng'
                    },
                    {
                        imageUrl: TestImages.personaMale
                    },
                    {
                        imageInitials: 'AL',
                        text: 'Annie Lindqvist'
                    },
                    {
                        imageUrl: TestImages.personaFemale
                    },
                    {
                        imageUrl: TestImages.personaMale
                    },
                    {
                        imageUrl: TestImages.personaMale
                    }
                ],
                timeStamp: 'August 3, 2017'
            }
        ];
        var activityExampleList = [];
        activityItemExamples.forEach(function (item) {
            var props = item;
            activityExampleList.push(React.createElement(ActivityItem, tslib_1.__assign({}, props, { key: item.key, className: css(classNames.exampleRoot) })));
        });
        return React.createElement("div", null, activityExampleList);
    };
    return ActivityItemPersonaExample;
}(React.Component));
export { ActivityItemPersonaExample };
