// office-ui-fabric-react/lib/components/ActivityItem/examples/ActivityItem.Basic.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { css, classNamesFunction } from 'office-ui-fabric-react/lib/Utilities';
import { ActivityItem } from '../ActivityItem';
import { getStyles } from './ActivityItem.Example.styles';
import { Icon } from '../../../Icon';
import { Link } from '../../../Link';
var ActivityItemBasicExample = /** @class */ (function (_super) {
    tslib_1.__extends(ActivityItemBasicExample, _super);
    function ActivityItemBasicExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ActivityItemBasicExample.prototype.render = function () {
        var getClassNames = classNamesFunction();
        var classNames = getClassNames(getStyles, {});
        // tslint:disable:jsx-no-lambda
        var activityItemExamples = [
            {
                key: 1,
                activityDescription: [
                    React.createElement(Link, { key: 1, className: css(classNames.nameText), onClick: function () {
                            alert('A name was clicked.');
                        } }, "Philippe Lampros"),
                    React.createElement("span", { key: 2 }, " commented")
                ],
                activityIcon: React.createElement(Icon, { iconName: 'Message' }),
                comments: [
                    React.createElement("span", { key: 1 }, "Hello! I am making a comment and mentioning "),
                    React.createElement(Link, { key: 2, className: css(classNames.nameText), onClick: function () {
                            alert('An @mentioned name was clicked.');
                        } }, "@An\u0111ela Debeljak"),
                    React.createElement("span", { key: 3 }, " in the text of the comment.")
                ],
                timeStamp: 'Just now'
            },
            {
                key: 2,
                activityDescription: [
                    React.createElement(Link, { key: 1, className: css(classNames.nameText), onClick: function () {
                            alert('A name was clicked.');
                        } }, "Lisha Refai"),
                    React.createElement("span", { key: 2 }, " deleted "),
                    React.createElement("span", { key: 3, className: css(classNames.nameText) }, "DocumentTitle.docx")
                ],
                activityIcon: React.createElement(Icon, { iconName: 'Trash' }),
                timeStamp: '2 hours ago'
            },
            {
                key: 3,
                activityDescription: [
                    React.createElement(Link, { key: 1, className: css(classNames.nameText), onClick: function () {
                            alert('A name was clicked.');
                        } }, "Julian Arvidsson"),
                    React.createElement("span", { key: 2 }, " moved "),
                    React.createElement(Link, { key: 3, className: css(classNames.nameText), onClick: function () {
                            alert('A document was clicked.');
                        } }, "PresentationTitle.pptx"),
                    React.createElement("span", { key: 4 }, " to "),
                    React.createElement(Link, { key: 5, className: css(classNames.nameText), onClick: function () {
                            alert('A folder was clicked.');
                        } }, "Destination Folder")
                ],
                activityIcon: React.createElement(Icon, { iconName: 'FabricMovetoFolder' }),
                timeStamp: 'Yesterday'
            }
        ];
        var activityExampleList = [];
        activityItemExamples.forEach(function (item) {
            var props = item;
            activityExampleList.push(React.createElement(ActivityItem, tslib_1.__assign({}, props, { key: item.key, className: css(classNames.exampleRoot) })));
        });
        return React.createElement("div", null, activityExampleList);
    };
    return ActivityItemBasicExample;
}(React.Component));
export { ActivityItemBasicExample };
