// office-ui-fabric-react/lib/components/Slider/Slider.doc.js
import * as React from 'react';
import { SliderBasicExample } from './examples/Slider.Basic.Example';
import { SliderStatus } from './Slider.checklist';
var SliderBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Slider/examples/Slider.Basic.Example.tsx');
export var SliderPageProps = {
    title: 'Slider',
    componentName: 'Slider',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/undefined',
    componentStatus: SliderStatus,
    examples: [
        {
            title: 'Slider',
            code: SliderBasicExampleCode,
            view: React.createElement(SliderBasicExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Slider/Slider.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Slider/docs/SliderOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Slider/docs/SliderDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Slider/docs/SliderDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
