// office-ui-fabric-react/lib-commonjs/components/ColorPicker/ColorPicker.doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var ColorPicker_Basic_Example_1 = require("./examples/ColorPicker.Basic.Example");
var ColorPicker_checklist_1 = require("./ColorPicker.checklist");
var ColorPickerBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ColorPicker/examples/ColorPicker.Basic.Example.tsx');
exports.ColorPickerPageProps = {
    title: 'ColorPicker',
    componentName: 'ColorPicker',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/ColorPicker',
    componentStatus: ColorPicker_checklist_1.ColorPickerStatus,
    examples: [
        {
            title: 'Default ColorPicker',
            code: ColorPickerBasicExampleCode,
            view: React.createElement(ColorPicker_Basic_Example_1.ColorPickerBasicExample, null)
        }
    ],
    propertiesTablesSources: [
        require('!raw-loader!office-ui-fabric-react/src/components/ColorPicker/ColorPicker.types.ts'),
        require('!raw-loader!office-ui-fabric-react/src/components/ColorPicker/ColorRectangle/ColorRectangle.types.ts'),
        require('!raw-loader!office-ui-fabric-react/src/components/ColorPicker/ColorSlider/ColorSlider.types.ts')
    ],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/ColorPicker/docs/ColorPickerOverview.md'),
    bestPractices: '',
    dos: '',
    donts: '',
    isHeaderVisible: true,
    isFeedbackVisible: true
};
