// office-ui-fabric-react/lib/components/Dropdown/Dropdown.doc.js
import * as React from 'react';
import { DropdownBasicExample } from './examples/Dropdown.Basic.Example';
import { DropdownCustomExample } from './examples/Dropdown.Custom.Example';
import { DropdownErrorExample } from './examples/Dropdown.Error.Example';
import { DropdownRequiredExample } from './examples/Dropdown.Required.Example';
import { DropdownStatus } from './Dropdown.checklist';
var DropdownBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Basic.Example.tsx');
var DropdownBasicExampleCodepen = require('!raw-loader!office-ui-fabric-react/lib/codepen/components/Dropdown/Dropdown.Basic.Example.Codepen.txt');
var DropdownCustomExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Custom.Example.tsx');
var DropdownErrorExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Error.Example.tsx');
var DropdownRequiredExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/examples/Dropdown.Required.Example.tsx');
export var DropdownPageProps = {
    title: 'Dropdown',
    componentName: 'Dropdown',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Dropdown',
    componentStatus: DropdownStatus,
    examples: [
        {
            title: 'Dropdown',
            code: DropdownBasicExampleCode,
            view: React.createElement(DropdownBasicExample, null),
            codepenJS: DropdownBasicExampleCodepen
        },
        {
            title: 'Customized Dropdown',
            code: DropdownCustomExampleCode,
            view: React.createElement(DropdownCustomExample, null)
        },
        {
            title: 'Dropdown with Error Message',
            code: DropdownErrorExampleCode,
            view: React.createElement(DropdownErrorExample, null)
        },
        {
            title: 'Required Dropdown',
            code: DropdownRequiredExampleCode,
            view: React.createElement(DropdownRequiredExample, null)
        }
    ],
    propertiesTablesSources: [
        require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/Dropdown.types.ts'),
        require('!raw-loader!office-ui-fabric-react/src/utilities/selectableOption/SelectableDroppableText.types.ts')
    ],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/docs/DropdownOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/docs/DropdownDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Dropdown/docs/DropdownDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true,
    allowNativeProps: true
};
