// office-ui-fabric-react/lib-commonjs/components/TextField/TextField.doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var TextField_Basic_Example_1 = require("./examples/TextField.Basic.Example");
var TextField_Borderless_Example_1 = require("./examples/TextField.Borderless.Example");
var TextField_CustomRender_Example_1 = require("./examples/TextField.CustomRender.Example");
var TextField_ErrorMessage_Example_1 = require("./examples/TextField.ErrorMessage.Example");
var TextField_Multiline_Example_1 = require("./examples/TextField.Multiline.Example");
var TextField_Controlled_Example_1 = require("./examples/TextField.Controlled.Example");
var TextField_PrefixAndSuffix_Example_1 = require("./examples/TextField.PrefixAndSuffix.Example");
var TextField_checklist_1 = require("./TextField.checklist");
var TextField_Styled_Example_1 = require("./examples/TextField.Styled.Example");
var TextFieldBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Basic.Example.tsx');
var TextFieldBasicExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Basic.Example.tsx');
var TextFieldBorderlessExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Borderless.Example.tsx');
var TextFieldBorderlessExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Borderless.Example.tsx');
var TextFieldCustomRenderExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.CustomRender.Example.tsx');
var TextFieldCustomRenderExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.CustomRender.Example.tsx');
var TextFieldErrorMessageExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.ErrorMessage.Example.tsx');
var TextFieldErrorMessageExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.ErrorMessage.Example.tsx');
var TextFieldMultilineExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Multiline.Example.tsx');
var TextFieldMultilineExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Multiline.Example.tsx');
var TextFieldControlledExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Controlled.Example.tsx');
var TextFieldControlledExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Controlled.Example.tsx');
var TextFieldPrefixAndSuffixExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.PrefixAndSuffix.Example.tsx');
var TextFieldPrefixAndSuffixExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.PrefixAndSuffix.Example.tsx');
var TextFieldStyledExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Styled.Example.tsx');
var TextFieldStyledExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/TextField/examples/TextField.Styled.Example.tsx');
exports.TextFieldPageProps = {
    title: 'TextField',
    componentName: 'TextField',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/TextField',
    componentStatus: TextField_checklist_1.TextFieldStatus,
    examples: [
        {
            title: 'Basic TextFields',
            code: TextFieldBasicExampleCode,
            view: React.createElement(TextField_Basic_Example_1.TextFieldBasicExample, null),
            codepenJS: TextFieldBasicExampleCodepen
        },
        {
            title: 'Controlled TextFields',
            code: TextFieldControlledExampleCode,
            view: React.createElement(TextField_Controlled_Example_1.TextFieldControlledExample, null),
            codepenJS: TextFieldControlledExampleCodepen
        },
        {
            title: 'Multiline TextField',
            code: TextFieldMultilineExampleCode,
            codepenJS: TextFieldMultilineExampleCodepen,
            view: React.createElement(TextField_Multiline_Example_1.TextFieldMultilineExample, null)
        },
        {
            title: 'Underlined and borderless TextFields',
            code: TextFieldBorderlessExampleCode,
            view: React.createElement(TextField_Borderless_Example_1.TextFieldBorderlessExample, null),
            codepenJS: TextFieldBorderlessExampleCodepen
        },
        {
            title: 'TextField with prefix and/or suffix',
            code: TextFieldPrefixAndSuffixExampleCode,
            codepenJS: TextFieldPrefixAndSuffixExampleCodepen,
            view: React.createElement(TextField_PrefixAndSuffix_Example_1.TextFieldPrefixAndSuffixExample, null)
        },
        {
            title: 'TextFields with custom rendering',
            code: TextFieldCustomRenderExampleCode,
            codepenJS: TextFieldCustomRenderExampleCodepen,
            view: React.createElement(TextField_CustomRender_Example_1.TextFieldCustomRenderExample, null)
        },
        {
            title: 'TextField error message variations',
            code: TextFieldErrorMessageExampleCode,
            codepenJS: TextFieldErrorMessageExampleCodepen,
            view: React.createElement(TextField_ErrorMessage_Example_1.TextFieldErrorMessageExample, null)
        },
        {
            title: 'TextField subcomponent styling',
            code: TextFieldStyledExampleCode,
            codepenJS: TextFieldStyledExampleCodepen,
            view: React.createElement(TextField_Styled_Example_1.TextFieldStyledExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/TextField/TextField.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/TextField/docs/TextFieldOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/TextField/docs/TextFieldDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/TextField/docs/TextFieldDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true,
    allowNativeProps: true,
    nativePropsElement: ['input', 'textarea']
};
