// office-ui-fabric-react/lib-amd/components/DetailsList/ShimmeredDetailsList.styles.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getStyles = function (props) {
        var theme = props.theme, className = props.className, enableShimmer = props.enableShimmer;
        var palette = theme.palette;
        return {
            root: [
                theme.fonts.small,
                enableShimmer && {
                    selectors: {
                        ':after': {
                            content: '""',
                            position: 'absolute',
                            top: 0,
                            right: 0,
                            bottom: 0,
                            left: 0,
                            backgroundImage: "linear-gradient(to bottom, transparent 30%, " + palette.whiteTranslucent40 + " 65%," + palette.white + " 100%)"
                        }
                    }
                },
                className
            ]
        };
    };
});
