// office-ui-fabric-react/lib/components/ColorPicker/ColorPicker.doc.js
import * as React from 'react';
import { ColorPickerBasicExample } from './examples/ColorPicker.Basic.Example';
import { ColorPickerStatus } from './ColorPicker.checklist';
var ColorPickerBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ColorPicker/examples/ColorPicker.Basic.Example.tsx');
export var ColorPickerPageProps = {
    title: 'ColorPicker',
    componentName: 'ColorPicker',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/ColorPicker',
    componentStatus: ColorPickerStatus,
    examples: [
        {
            title: 'Default ColorPicker',
            code: ColorPickerBasicExampleCode,
            view: React.createElement(ColorPickerBasicExample, null)
        }
    ],
    propertiesTablesSources: [
        require('!raw-loader!office-ui-fabric-react/src/components/ColorPicker/ColorPicker.types.ts'),
        require('!raw-loader!office-ui-fabric-react/src/components/ColorPicker/ColorRectangle/ColorRectangle.types.ts'),
        require('!raw-loader!office-ui-fabric-react/src/components/ColorPicker/ColorSlider/ColorSlider.types.ts')
    ],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/ColorPicker/docs/ColorPickerOverview.md'),
    bestPractices: '',
    dos: '',
    donts: '',
    isHeaderVisible: true,
    isFeedbackVisible: true
};
