// office-ui-fabric-react/lib/components/Nav/examples/Nav.FabricDemoApp.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { Nav } from 'office-ui-fabric-react/lib/Nav';
var AppDefinition = {
    appTitle: 'Fabric - React',
    examplePages: [
        {
            links: [
                {
                    key: 'ActivityItem',
                    name: 'ActivityItem',
                    url: '#/examples/activityitem'
                },
                {
                    key: 'Breadcrumb',
                    name: 'Breadcrumb',
                    url: '#/examples/breadcrumb'
                },
                {
                    key: 'Button',
                    name: 'Button',
                    url: '#/examples/button'
                }
            ],
            name: 'Basic components'
        },
        {
            links: [
                {
                    key: 'ColorPicker',
                    name: 'ColorPicker',
                    url: '#/examples/colorpicker'
                },
                {
                    key: 'ExtendedPeoplePicker',
                    name: 'ExtendedPeoplePicker',
                    url: '#examples/extendedpeoplepicker'
                },
                {
                    key: 'GroupedList',
                    name: 'GroupedList',
                    url: '#examples/groupedlist'
                }
            ],
            name: 'Extended components'
        },
        {
            links: [
                {
                    key: 'FocusTrapZone',
                    name: 'FocusTrapZone',
                    url: '#examples/focustrapzone'
                },
                {
                    key: 'FocusZone',
                    name: 'FocusZone',
                    url: '#examples/focuszone'
                },
                {
                    key: 'MarqueeSelection',
                    name: 'MarqueeSelection',
                    url: '#examples/marqueeselection'
                }
            ],
            name: 'Utilities'
        }
    ]
};
var NavFabricDemoAppExample = /** @class */ (function (_super) {
    tslib_1.__extends(NavFabricDemoAppExample, _super);
    function NavFabricDemoAppExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._onRenderLink = function (link) {
            return (React.createElement("span", null,
                React.createElement("span", { key: 1, className: "Nav-linkText" }, link.name)));
        };
        return _this;
    }
    NavFabricDemoAppExample.prototype.render = function () {
        return React.createElement(Nav, { groups: AppDefinition.examplePages, onRenderLink: this._onRenderLink, expandButtonAriaLabel: 'Expand or collapse' });
    };
    return NavFabricDemoAppExample;
}(React.Component));
export { NavFabricDemoAppExample };
