// office-ui-fabric-react/lib/components/Divider/Divider.doc.js
import * as React from 'react';
import { VerticalDividerBasicExample } from './examples/VerticalDivider.Basic.Example';
import { VerticalDividerCustomExample } from './examples/VerticalDivider.Custom.Example';
import { DividerStatus } from './Divider.checklist';
var VerticalDividerBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Divider/examples/VerticalDivider.Basic.Example.tsx');
var VerticalDividerCustomExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Divider/examples/VerticalDivider.Custom.Example.tsx');
export var DividerPageProps = {
    title: 'Divider',
    componentName: 'Divider',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Divider',
    componentStatus: DividerStatus,
    examples: [
        {
            title: 'Vertical Divider',
            code: VerticalDividerBasicExampleCode,
            view: React.createElement(VerticalDividerBasicExample, null)
        },
        {
            title: 'Custom Vertical Divider',
            code: VerticalDividerCustomExampleCode,
            view: React.createElement(VerticalDividerCustomExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Divider/VerticalDivider.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Divider/docs/DividerOverview.md'),
    bestPractices: require('!raw-loader!office-ui-fabric-react/src/components/Divider/docs/DividerBestPractices.md'),
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Divider/docs/DividerDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Divider/docs/DividerDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
