// office-ui-fabric-react/lib-commonjs/components/FocusZone/FocusZone.doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var FocusZone_Photos_Example_1 = require("./examples/FocusZone.Photos.Example");
var FocusZone_List_Example_1 = require("./examples/FocusZone.List.Example");
var FocusZone_Disabled_Example_1 = require("./examples/FocusZone.Disabled.Example");
var FocusZone_Tabbable_Example_1 = require("./examples/FocusZone.Tabbable.Example");
var FocusZonePhotosExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.Photos.Example.tsx');
var FocusZonePhotosExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.Photos.Example.tsx');
var FocusZoneListExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.List.Example.tsx');
var FocusZoneListExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.List.Example.tsx');
var FocusZoneDisabledExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.Disabled.Example.tsx');
var FocusZoneDisabledExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.Disabled.Example.tsx');
var FocusZoneTabbableCode = require('!raw-loader!office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.Tabbable.Example.tsx');
var FocusZoneTabbableCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/FocusZone/examples/FocusZone.Tabbable.Example.tsx');
exports.FocusZonePageProps = {
    title: 'FocusZone',
    componentName: 'FocusZone',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/FocusZone',
    examples: [
        {
            title: 'Non-uniform photos within bidirectional FocusZone',
            code: FocusZonePhotosExampleCode,
            codepenJS: FocusZonePhotosExampleCodepen,
            view: React.createElement(FocusZone_Photos_Example_1.FocusZonePhotosExample, null)
        },
        {
            title: 'Nesting FocusZones in list rows',
            code: FocusZoneListExampleCode,
            codepenJS: FocusZoneListExampleCodepen,
            view: React.createElement(FocusZone_List_Example_1.FocusZoneListExample, null)
        },
        {
            title: 'Disabled FocusZone',
            code: FocusZoneDisabledExampleCode,
            codepenJS: FocusZoneDisabledExampleCodepen,
            view: React.createElement(FocusZone_Disabled_Example_1.FocusZoneDisabledExample, null)
        },
        {
            title: 'Tabbable FocusZone',
            code: FocusZoneTabbableCode,
            codepenJS: FocusZoneTabbableCodepen,
            view: React.createElement(FocusZone_Tabbable_Example_1.FocusZoneTabbableExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/FocusZone/FocusZone.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/FocusZone/docs/FocusZoneOverview.md'),
    bestPractices: '',
    dos: '',
    donts: '',
    isHeaderVisible: true,
    isFeedbackVisible: true,
    allowNativeProps: true
};
