// office-ui-fabric-react/lib-amd/components/Modal/Modal.doc.js
define(["require", "exports", "react", "./examples/Modal.Basic.Example", "./examples/Modal.Modeless.Example", "./Modal.checklist"], function (require, exports, React, Modal_Basic_Example_1, Modal_Modeless_Example_1, Modal_checklist_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ModalBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Modal/examples/Modal.Basic.Example.tsx');
    var ModalModelessExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Modal/examples/Modal.Modeless.Example.tsx');
    exports.ModalPageProps = {
        title: 'Modal',
        componentName: 'Modal',
        componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Modal',
        componentStatus: Modal_checklist_1.ModalStatus,
        examples: [
            {
                title: 'Modal',
                code: ModalBasicExampleCode,
                view: React.createElement(Modal_Basic_Example_1.ModalBasicExample, null)
            },
            {
                title: 'Modeless Modal',
                code: ModalModelessExampleCode,
                view: React.createElement(Modal_Modeless_Example_1.ModalModelessExample, null)
            }
        ],
        propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Modal/Modal.types.ts')],
        overview: require('!raw-loader!office-ui-fabric-react/src/components/Modal/docs/ModalOverview.md'),
        bestPractices: '',
        dos: require('!raw-loader!office-ui-fabric-react/src/components/Modal/docs/ModalDos.md'),
        donts: require('!raw-loader!office-ui-fabric-react/src/components/Modal/docs/ModalDonts.md'),
        isHeaderVisible: true,
        isFeedbackVisible: true
    };
});
