// office-ui-fabric-react/lib/components/Nav/Nav.doc.js
import * as React from 'react';
import { NavBasicExample } from './examples/Nav.Basic.Example';
import { NavFabricDemoAppExample } from './examples/Nav.FabricDemoApp.Example';
import { NavNestedExample } from './examples/Nav.Nested.Example';
import { NavStatus } from './Nav.checklist';
import { NavCustomGroupHeadersExample } from 'office-ui-fabric-react/lib/components/Nav/examples/Nav.CustomGroupHeaders.Example';
var NavBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.Basic.Example.tsx');
var NavBasicExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.Basic.Example.tsx');
var NavFabricDemoAppExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.FabricDemoApp.Example.tsx');
var NavFabricDemoAppExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.FabricDemoApp.Example.tsx');
var NavNestedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.Nested.Example.tsx');
var NavNestedExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.Nested.Example.tsx');
var NavCustomGroupHeadersExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.CustomGroupHeaders.Example.tsx');
var NavCustomGroupHeadersExampleCodepen = require('!@uifabric/codepen-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.CustomGroupHeaders.Example.tsx');
export var NavPageProps = {
    title: 'Nav',
    componentName: 'Nav',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Nav',
    componentStatus: NavStatus,
    examples: [
        {
            title: 'Basic nav with sample links',
            code: NavBasicExampleCode,
            codepenJS: NavBasicExampleCodepen,
            view: React.createElement(NavBasicExample, null)
        },
        {
            title: 'Nav similar to the one in this demo app',
            code: NavFabricDemoAppExampleCode,
            codepenJS: NavFabricDemoAppExampleCodepen,
            view: React.createElement(NavFabricDemoAppExample, null)
        },
        {
            title: 'Nav with nested links',
            code: NavNestedExampleCode,
            codepenJS: NavNestedExampleCodepen,
            view: React.createElement(NavNestedExample, null)
        },
        {
            title: 'Nav with custom group header',
            code: NavCustomGroupHeadersExampleCode,
            codepenJS: NavCustomGroupHeadersExampleCodepen,
            view: React.createElement(NavCustomGroupHeadersExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Nav/Nav.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Nav/docs/NavOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Nav/docs/NavDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Nav/docs/NavDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
