// office-ui-fabric-react/lib/components/Nav/Nav.doc.js
import * as React from 'react';
import { NavBasicExample } from './examples/Nav.Basic.Example';
import { NavFabricDemoAppExample } from './examples/Nav.FabricDemoApp.Example';
import { NavNestedExample } from './examples/Nav.Nested.Example';
import { NavByKeysExample } from './examples/Nav.ByKeys.Example';
import { NavStatus } from './Nav.checklist';
import { NavCustomGroupHeadersExample } from 'office-ui-fabric-react/lib/components/Nav/examples/Nav.CustomGroupHeaders.Example';
import { NavDisabledItemsExample } from './examples/Nav.DisabledItems.Example';
var NavBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.Basic.Example.tsx');
var NavFabricDemoAppExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.FabricDemoApp.Example.tsx');
var NavNestedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.Nested.Example.tsx');
var NavByKeysExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.ByKeys.Example.tsx');
var NavCustomGroupHeadersExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.CustomGroupHeaders.Example.tsx');
var NavDisabledItemsExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Nav/examples/Nav.DisabledItems.Example.tsx');
export var NavPageProps = {
    title: 'Nav',
    componentName: 'Nav',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Nav',
    componentStatus: NavStatus,
    examples: [
        {
            title: 'Basic Nav bar with sample links',
            code: NavBasicExampleCode,
            view: React.createElement(NavBasicExample, null)
        },
        {
            title: 'Navigation menu used in this Fabric React demo app',
            code: NavFabricDemoAppExampleCode,
            view: React.createElement(NavFabricDemoAppExample, null)
        },
        {
            title: 'Nested navigation menu (without group header)',
            code: NavNestedExampleCode,
            view: React.createElement(NavNestedExample, null)
        },
        {
            title: 'Nav bar menu with group custom header',
            code: NavCustomGroupHeadersExampleCode,
            view: React.createElement(NavCustomGroupHeadersExample, null)
        },
        {
            title: 'Nav bar with disabled items',
            code: NavDisabledItemsExampleCode,
            view: React.createElement(NavDisabledItemsExample, null)
        },
        {
            title: 'Nav bar of links each with unique keys and empty urls',
            code: NavByKeysExampleCode,
            view: React.createElement(NavByKeysExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/Nav/Nav.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/Nav/docs/NavOverview.md'),
    bestPractices: '',
    dos: require('!raw-loader!office-ui-fabric-react/src/components/Nav/docs/NavDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/Nav/docs/NavDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
