// office-ui-fabric-react/lib/components/pickers/PeoplePicker/PeoplePicker.doc.js
import * as React from 'react';
import { PeoplePickerTypesExample } from './examples/PeoplePicker.Types.Example';
import { PeoplePickerStatus } from './PeoplePicker.checklist';
var PeoplePickerTypesExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/pickers/PeoplePicker/examples/PeoplePicker.Types.Example.tsx');
export var PeoplePickerPageProps = {
    title: 'PeoplePicker',
    componentName: 'PeoplePicker',
    componentUrl: 'https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/PeoplePicker',
    componentStatus: PeoplePickerStatus,
    examples: [
        {
            title: 'People Pickers',
            code: PeoplePickerTypesExampleCode,
            view: React.createElement(PeoplePickerTypesExample, null)
        }
    ],
    propertiesTablesSources: [require('!raw-loader!office-ui-fabric-react/src/components/pickers/BasePicker.types.ts')],
    overview: require('!raw-loader!office-ui-fabric-react/src/components/pickers/PeoplePicker/docs/PeoplePickerOverview.md'),
    bestPractices: require('!raw-loader!office-ui-fabric-react/src/components/pickers/PeoplePicker/docs/PeoplePickerBestPractices.md'),
    dos: require('!raw-loader!office-ui-fabric-react/src/components/pickers/PeoplePicker/docs/PeoplePickerDos.md'),
    donts: require('!raw-loader!office-ui-fabric-react/src/components/pickers/PeoplePicker/docs/PeoplePickerDonts.md'),
    isHeaderVisible: true,
    isFeedbackVisible: true
};
